// igniteui-angular/esm2015/lib/animations/scale/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { animate, animation, style } from '@angular/animations';
import { EaseOut } from '../easings';
const /** @type {?} */ base = [
    style({
        opacity: `{{startOpacity}}`,
        transform: `scale{{direction}}({{fromScale}})`,
        transformOrigin: `{{xPos}} {{yPos}}`
    }),
    animate(`{{duration}} {{delay}} {{easing}}`, style({
        opacity: `{{endOpacity}}`,
        transform: `scale{{direction}}({{toScale}})`,
        transformOrigin: `{{xPos}} {{yPos}}`
    }))
];
const /** @type {?} */ baseInParams = {
    delay: '0s',
    direction: '',
    duration: '350ms',
    easing: EaseOut.quad,
    endOpacity: 1,
    fromScale: .5,
    startOpacity: 0,
    toScale: 1,
    xPos: '50%',
    yPos: '50%'
};
const /** @type {?} */ baseOutParams = Object.assign({}, baseInParams, { easing: EaseOut.sine, endOpacity: 0, fromScale: 1, startOpacity: 1, toScale: .5 });
const /** @type {?} */ scaleInCenter = animation(base, { params: baseInParams });
const /** @type {?} */ scaleInBl = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '0', yPos: '100%' })
});
const /** @type {?} */ scaleInVerCenter = animation(base, {
    params: Object.assign({}, baseInParams, { direction: 'Y', fromScale: .4 })
});
const /** @type {?} */ scaleInTop = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '50%', yPos: '0' })
});
const /** @type {?} */ scaleInLeft = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '0', yPos: '50%' })
});
const /** @type {?} */ scaleInVerTop = animation(base, {
    params: Object.assign({}, baseInParams, { direction: 'Y', fromScale: .4, xPos: '100%', yPos: '0' })
});
const /** @type {?} */ scaleInTr = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '100%', yPos: '0' })
});
const /** @type {?} */ scaleInTl = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '0', yPos: '0' })
});
const /** @type {?} */ scaleInVerBottom = animation(base, {
    params: Object.assign({}, baseInParams, { direction: 'Y', fromScale: .4, xPos: '0', yPos: '100%' })
});
const /** @type {?} */ scaleInRight = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '100%', yPos: '50%' })
});
const /** @type {?} */ scaleInHorCenter = animation(base, {
    params: Object.assign({}, baseInParams, { direction: 'X', fromScale: .4 })
});
const /** @type {?} */ scaleInBr = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '100%', yPos: '100%' })
});
const /** @type {?} */ scaleInHorLeft = animation(base, {
    params: Object.assign({}, baseInParams, { direction: 'X', fromScale: .4, xPos: '0', yPos: '0' })
});
const /** @type {?} */ scaleInBottom = animation(base, {
    params: Object.assign({}, baseInParams, { xPos: '50%', yPos: '100%' })
});
const /** @type {?} */ scaleInHorRight = animation(base, {
    params: Object.assign({}, baseInParams, { direction: 'X', fromScale: .4, xPos: '100%', yPos: '100%' })
});
const /** @type {?} */ scaleOutCenter = animation(base, { params: baseOutParams });
const /** @type {?} */ scaleOutBl = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '0', yPos: '100%' })
});
const /** @type {?} */ scaleOutBr = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '100%', yPos: '100%' })
});
const /** @type {?} */ scaleOutVerCenter = animation(base, {
    params: Object.assign({}, baseOutParams, { direction: 'Y', toScale: .3 })
});
const /** @type {?} */ scaleOutVerTop = animation(base, {
    params: Object.assign({}, baseOutParams, { direction: 'Y', toScale: .3, xPos: '100%', yPos: '0' })
});
const /** @type {?} */ scaleOutVerBottom = animation(base, {
    params: Object.assign({}, baseOutParams, { direction: 'Y', toScale: .3, xPos: '0', yPos: '100%' })
});
const /** @type {?} */ scaleOutTop = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '50%', yPos: '0' })
});
const /** @type {?} */ scaleOutLeft = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '0', yPos: '50%' })
});
const /** @type {?} */ scaleOutTr = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '100%', yPos: '0' })
});
const /** @type {?} */ scaleOutTl = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '0', yPos: '0' })
});
const /** @type {?} */ scaleOutRight = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '100%', yPos: '50%' })
});
const /** @type {?} */ scaleOutBottom = animation(base, {
    params: Object.assign({}, baseOutParams, { xPos: '50%', yPos: '100%' })
});
const /** @type {?} */ scaleOutHorCenter = animation(base, {
    params: Object.assign({}, baseOutParams, { direction: 'X', toScale: .3 })
});
const /** @type {?} */ scaleOutHorLeft = animation(base, {
    params: Object.assign({}, baseOutParams, { direction: 'X', toScale: .3, xPos: '0', yPos: '0' })
});
const /** @type {?} */ scaleOutHorRight = animation(base, {
    params: Object.assign({}, baseOutParams, { direction: 'X', toScale: .3, xPos: '100%', yPos: '100%' })
});
export { scaleInTop, scaleInRight, scaleInBottom, scaleInLeft, scaleInCenter, scaleInTr, scaleInBr, scaleInBl, scaleInTl, scaleInVerTop, scaleInVerBottom, scaleInVerCenter, scaleInHorCenter, scaleInHorLeft, scaleInHorRight, scaleOutTop, scaleOutRight, scaleOutBottom, scaleOutLeft, scaleOutCenter, scaleOutTr, scaleOutBr, scaleOutBl, scaleOutTl, scaleOutVerTop, scaleOutVerBottom, scaleOutVerCenter, scaleOutHorCenter, scaleOutHorLeft, scaleOutHorRight };
