// igniteui-angular/esm2015/lib/services/csv/char-separated-value-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ExportUtilities } from '../exporter-common/export-utilities';
/**
 * @hidden
 */
export class CharSeparatedValueData {
    /**
     * @param {?} _data
     * @param {?} valueDelimiter
     */
    constructor(_data, valueDelimiter) {
        this._data = _data;
        this._headerRecord = '';
        this._dataRecords = '';
        this._eor = '\r\n';
        this._escapeCharacters = ['\r', '\n', '\r\n'];
        this._delimiterLength = 1;
        this._isSpecialData = false;
        this.setDelimiter(valueDelimiter);
    }
    /**
     * @return {?}
     */
    prepareData() {
        if (!this._data || this._data.length === 0) {
            return '';
        }
        const /** @type {?} */ keys = ExportUtilities.getKeysFromData(this._data);
        if (keys.length === 0) {
            return '';
        }
        this._isSpecialData = ExportUtilities.isSpecialData(this._data);
        this._escapeCharacters.push(this._delimiter);
        this._headerRecord = this.processHeaderRecord(keys, this._escapeCharacters);
        this._dataRecords = this.processDataRecords(this._data, keys, this._escapeCharacters);
        return this._headerRecord + this._dataRecords;
    }
    /**
     * @param {?} value
     * @param {?} escapeChars
     * @return {?}
     */
    processField(value, escapeChars) {
        let /** @type {?} */ safeValue = ExportUtilities.hasValue(value) ? String(value) : '';
        if (escapeChars.some((v) => safeValue.includes(v))) {
            safeValue = `"${safeValue}"`;
        }
        return safeValue + this._delimiter;
    }
    /**
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    processHeaderRecord(keys, escapeChars) {
        let /** @type {?} */ recordData = '';
        for (const /** @type {?} */ keyName of keys) {
            recordData += this.processField(keyName, this._escapeCharacters);
        }
        return recordData.slice(0, -this._delimiterLength) + this._eor;
    }
    /**
     * @param {?} record
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    processRecord(record, keys, escapeChars) {
        let /** @type {?} */ recordData = '';
        for (const /** @type {?} */ keyName of keys) {
            const /** @type {?} */ value = (record[keyName] !== undefined) ? record[keyName] : this._isSpecialData ? record : '';
            recordData += this.processField(value, this._escapeCharacters);
        }
        return recordData.slice(0, -this._delimiterLength) + this._eor;
    }
    /**
     * @param {?} currentData
     * @param {?} keys
     * @param {?} escapeChars
     * @return {?}
     */
    processDataRecords(currentData, keys, escapeChars) {
        let /** @type {?} */ dataRecords = '';
        for (const /** @type {?} */ row of currentData) {
            dataRecords += this.processRecord(row, keys, escapeChars);
        }
        return dataRecords;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setDelimiter(value) {
        this._delimiter = value;
        this._delimiterLength = value.length;
    }
}
function CharSeparatedValueData_tsickle_Closure_declarations() {
    /** @type {?} */
    CharSeparatedValueData.prototype._headerRecord;
    /** @type {?} */
    CharSeparatedValueData.prototype._dataRecords;
    /** @type {?} */
    CharSeparatedValueData.prototype._eor;
    /** @type {?} */
    CharSeparatedValueData.prototype._delimiter;
    /** @type {?} */
    CharSeparatedValueData.prototype._escapeCharacters;
    /** @type {?} */
    CharSeparatedValueData.prototype._delimiterLength;
    /** @type {?} */
    CharSeparatedValueData.prototype._isSpecialData;
    /** @type {?} */
    CharSeparatedValueData.prototype._data;
}
