// igniteui-angular/esm5/lib/services/transaction/igx-transaction.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { TransactionType } from './transaction';
import { IgxBaseTransactionService } from './base-transaction';
import { EventEmitter, Injectable } from '@angular/core';
import { isObject, mergeObjects, cloneValue } from '../../core/utils';
// unsupported: template constraints.
// unsupported: template constraints.
/**
 * @template T, S
 */
var IgxTransactionService = /** @class */ (function (_super) {
    tslib_1.__extends(IgxTransactionService, _super);
    function IgxTransactionService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._transactions = [];
        _this._redoStack = [];
        _this._undoStack = [];
        _this._states = new Map();
        _this.onStateUpdate = new EventEmitter();
        return _this;
    }
    Object.defineProperty(IgxTransactionService.prototype, "canUndo", {
        get: /**
         * @return {?}
         */
        function () {
            return this._undoStack.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxTransactionService.prototype, "canRedo", {
        get: /**
         * @return {?}
         */
        function () {
            return this._redoStack.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} transaction
     * @param {?=} recordRef
     * @return {?}
     */
    IgxTransactionService.prototype.add = /**
     * @param {?} transaction
     * @param {?=} recordRef
     * @return {?}
     */
    function (transaction, recordRef) {
        var /** @type {?} */ states = this._isPending ? this._pendingStates : this._states;
        this.verifyAddedTransaction(states, transaction, recordRef);
        this.addTransaction(transaction, states, recordRef);
    };
    /**
     * @param {?} transaction
     * @param {?} states
     * @param {?=} recordRef
     * @param {?=} useInUndo
     * @return {?}
     */
    IgxTransactionService.prototype.addTransaction = /**
     * @param {?} transaction
     * @param {?} states
     * @param {?=} recordRef
     * @param {?=} useInUndo
     * @return {?}
     */
    function (transaction, states, recordRef, useInUndo) {
        if (useInUndo === void 0) { useInUndo = true; }
        this.updateState(states, transaction, recordRef);
        var /** @type {?} */ transactions = this._isPending ? this._pendingTransactions : this._transactions;
        transactions.push(transaction);
        if (!this._isPending) {
            this._undoStack.push({ transaction: transaction, recordRef: recordRef, useInUndo: useInUndo });
            this._redoStack = [];
            this.onStateUpdate.emit();
        }
    };
    /**
     * @param {?=} id
     * @return {?}
     */
    IgxTransactionService.prototype.getTransactionLog = /**
     * @param {?=} id
     * @return {?}
     */
    function (id) {
        if (id) {
            return this._transactions.filter(function (t) { return t.id === id; });
        }
        return tslib_1.__spread(this._transactions);
    };
    /**
     * @param {?} mergeChanges
     * @return {?}
     */
    IgxTransactionService.prototype.getAggregatedChanges = /**
     * @param {?} mergeChanges
     * @return {?}
     */
    function (mergeChanges) {
        var _this = this;
        var /** @type {?} */ result = [];
        this._states.forEach(function (state, key) {
            var /** @type {?} */ value = mergeChanges ? _this.mergeValues(state.recordRef, state.value) : state.value;
            result.push(/** @type {?} */ ({ id: key, newValue: value, type: state.type }));
        });
        return result;
    };
    /**
     * @param {?} id
     * @return {?}
     */
    IgxTransactionService.prototype.getState = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this._states.get(id);
    };
    Object.defineProperty(IgxTransactionService.prototype, "enabled", {
        get: /**
         * @return {?}
         */
        function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} id
     * @param {?} mergeChanges
     * @return {?}
     */
    IgxTransactionService.prototype.getAggregatedValue = /**
     * @param {?} id
     * @param {?} mergeChanges
     * @return {?}
     */
    function (id, mergeChanges) {
        var /** @type {?} */ state = this._states.get(id);
        var /** @type {?} */ pendingState = _super.prototype.getState.call(this, id);
        //  if there is no state and there is no pending state return null
        if (!state && !pendingState) {
            return null;
        }
        var /** @type {?} */ pendingChange = _super.prototype.getAggregatedValue.call(this, id, false);
        var /** @type {?} */ change = state && state.value;
        var /** @type {?} */ aggregatedValue = this.mergeValues(change, pendingChange);
        if (mergeChanges) {
            var /** @type {?} */ originalValue = state ? state.recordRef : pendingState.recordRef;
            aggregatedValue = this.mergeValues(originalValue, aggregatedValue);
        }
        return aggregatedValue;
    };
    /**
     * @param {?} commit
     * @return {?}
     */
    IgxTransactionService.prototype.endPending = /**
     * @param {?} commit
     * @return {?}
     */
    function (commit) {
        var _this = this;
        this._isPending = false;
        if (commit) {
            var /** @type {?} */ i_1 = 0;
            this._pendingStates.forEach(function (s, k) {
                _this.addTransaction(/** @type {?} */ ({ id: k, newValue: s.value, type: s.type }), _this._states, s.recordRef, i_1 === 0);
                i_1++;
            });
        }
        _super.prototype.endPending.call(this, commit);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    IgxTransactionService.prototype.commit = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this._states.forEach(function (s) {
            var /** @type {?} */ index = data.findIndex(function (i) { return JSON.stringify(i) === JSON.stringify(s.recordRef); });
            switch (s.type) {
                case TransactionType.ADD:
                    data.push(s.value);
                    break;
                case TransactionType.DELETE:
                    if (0 <= index && index < data.length) {
                        data.splice(index, 1);
                    }
                    break;
                case TransactionType.UPDATE:
                    if (0 <= index && index < data.length) {
                        data[index] = _this.updateValue(s);
                    }
                    break;
            }
        });
        this.clear();
    };
    /**
     * @return {?}
     */
    IgxTransactionService.prototype.clear = /**
     * @return {?}
     */
    function () {
        this._transactions = [];
        this._states.clear();
        this._redoStack = [];
        this._undoStack = [];
        this.onStateUpdate.emit();
    };
    /**
     * @return {?}
     */
    IgxTransactionService.prototype.undo = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._undoStack.length <= 0) {
            return;
        }
        var /** @type {?} */ action;
        do {
            action = this._undoStack.pop();
            this._transactions.pop();
            this._redoStack.push(action);
        } while (!action.useInUndo);
        this._states.clear();
        this._undoStack.map(function (a) { return _this.updateState(_this._states, a.transaction, a.recordRef); });
        this.onStateUpdate.emit();
    };
    /**
     * @return {?}
     */
    IgxTransactionService.prototype.redo = /**
     * @return {?}
     */
    function () {
        if (this._redoStack.length > 0) {
            //  remove first item from redo stack (it should always has useInUndo === true)
            //  and then all next items until there are items and useInUndo === false.
            //  If there are no more items, or next item's useInUndo === true leave.
            var /** @type {?} */ undoItem = void 0;
            undoItem = this._redoStack.pop();
            this.updateState(this._states, undoItem.transaction, undoItem.recordRef);
            this._transactions.push(undoItem.transaction);
            this._undoStack.push(undoItem);
            while (this._redoStack[this._redoStack.length - 1] && !this._redoStack[this._redoStack.length - 1].useInUndo) {
                undoItem = this._redoStack.pop();
                this.updateState(this._states, undoItem.transaction, undoItem.recordRef);
                this._transactions.push(undoItem.transaction);
                this._undoStack.push(undoItem);
            }
            this.onStateUpdate.emit();
        }
    };
    /**
     * Verifies if the passed transaction is correct. If not throws an exception.
     * @param transaction Transaction to be verified
     */
    /**
     * Verifies if the passed transaction is correct. If not throws an exception.
     * @param {?} states
     * @param {?} transaction Transaction to be verified
     * @param {?=} recordRef
     * @return {?}
     */
    IgxTransactionService.prototype.verifyAddedTransaction = /**
     * Verifies if the passed transaction is correct. If not throws an exception.
     * @param {?} states
     * @param {?} transaction Transaction to be verified
     * @param {?=} recordRef
     * @return {?}
     */
    function (states, transaction, recordRef) {
        var /** @type {?} */ state = states.get(transaction.id);
        switch (transaction.type) {
            case TransactionType.ADD:
                if (state) {
                    //  cannot add same item twice
                    throw new Error("Cannot add this transaction. Transaction with id: " + transaction.id + " has been already added.");
                }
                break;
            case TransactionType.DELETE:
            case TransactionType.UPDATE:
                if (state && state.type === TransactionType.DELETE) {
                    //  cannot delete or update deleted items
                    throw new Error("Cannot add this transaction. Transaction with id: " + transaction.id + " has been already deleted.");
                }
                if (!state && !recordRef && !this._isPending) {
                    //  cannot initially add transaction or delete item with no recordRef
                    throw new Error("Cannot add this transaction. This is first transaction of type " + transaction.type + " " +
                        ("for id " + transaction.id + ". For first transaction of this type recordRef is mandatory."));
                }
                break;
        }
    };
    /**
     * Updates the provided states collection according to passed transaction and recordRef
     * @param states States collection to apply the update to
     * @param transaction Transaction to apply to the current state
     * @param recordRef Reference to the value of the record in data source, if any, where transaction should be applied
     */
    /**
     * Updates the provided states collection according to passed transaction and recordRef
     * @param {?} states States collection to apply the update to
     * @param {?} transaction Transaction to apply to the current state
     * @param {?=} recordRef Reference to the value of the record in data source, if any, where transaction should be applied
     * @return {?}
     */
    IgxTransactionService.prototype.updateState = /**
     * Updates the provided states collection according to passed transaction and recordRef
     * @param {?} states States collection to apply the update to
     * @param {?} transaction Transaction to apply to the current state
     * @param {?=} recordRef Reference to the value of the record in data source, if any, where transaction should be applied
     * @return {?}
     */
    function (states, transaction, recordRef) {
        var /** @type {?} */ state = states.get(transaction.id);
        //  if TransactionType is ADD simply add transaction to states;
        //  if TransactionType is DELETE:
        //    - if there is state with this id of type ADD remove it from the states;
        //    - if there is state with this id of type UPDATE change its type to DELETE;
        //    - if there is no state with this id add transaction to states;
        //  if TransactionType is UPDATE:
        //    - if there is state with this id of type ADD merge new value and state recordRef into state new value
        //    - if there is state with this id of type UPDATE merge new value into state new value
        //    - if there is state with this id and state type is DELETE change its type to UPDATE
        //    - if there is no state with this id add transaction to states;
        if (state) {
            switch (transaction.type) {
                case TransactionType.DELETE:
                    if (state.type === TransactionType.ADD) {
                        states.delete(transaction.id);
                    }
                    else if (state.type === TransactionType.UPDATE) {
                        state.value = transaction.newValue;
                        state.type = TransactionType.DELETE;
                    }
                    break;
                case TransactionType.UPDATE:
                    if (isObject(state.value)) {
                        if (state.type === TransactionType.ADD) {
                            state.value = this.mergeValues(state.value, transaction.newValue);
                        }
                        if (state.type === TransactionType.UPDATE) {
                            mergeObjects(state.value, transaction.newValue);
                        }
                    }
                    else {
                        state.value = transaction.newValue;
                    }
            }
        }
        else {
            state = /** @type {?} */ ({ value: cloneValue(transaction.newValue), recordRef: recordRef, type: transaction.type });
            states.set(transaction.id, state);
        }
        //  should not clean pending state. This will happen automatically on endPending call
        if (!this._isPending) {
            this.cleanState(transaction.id, states);
        }
    };
    /**
     * Compares the state with recordRef and clears all duplicated values. If any state ends as
     * empty object removes it from states.
     * @param state State to clean
     */
    /**
     * Compares the state with recordRef and clears all duplicated values. If any state ends as
     * empty object removes it from states.
     * @param {?} id
     * @param {?} states
     * @return {?}
     */
    IgxTransactionService.prototype.cleanState = /**
     * Compares the state with recordRef and clears all duplicated values. If any state ends as
     * empty object removes it from states.
     * @param {?} id
     * @param {?} states
     * @return {?}
     */
    function (id, states) {
        var /** @type {?} */ state = states.get(id);
        //  do nothing if
        //  there is no state, or
        //  there is no state value (e.g. DELETED transaction), or
        //  there is no recordRef (e.g. ADDED transaction)
        if (state && state.value && state.recordRef) {
            //  if state's value is object compare each key with the ones in recordRef
            //  if values in any key are the same delete it from state's value
            //  if state's value is not object, simply compare with recordRef and remove
            //  the state if they are equal
            if (isObject(state.recordRef)) {
                try {
                    for (var _a = tslib_1.__values(Object.keys(state.value)), _b = _a.next(); !_b.done; _b = _a.next()) {
                        var key = _b.value;
                        if (JSON.stringify(state.recordRef[key]) === JSON.stringify(state.value[key])) {
                            delete state.value[key];
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                //  if state's value is empty remove the state from the states, only if state is not DELETE type
                if (state.type !== TransactionType.DELETE && Object.keys(state.value).length === 0) {
                    states.delete(id);
                }
            }
            else {
                if (state.recordRef === state.value) {
                    states.delete(id);
                }
            }
        }
        var e_1, _c;
    };
    IgxTransactionService.decorators = [
        { type: Injectable },
    ];
    return IgxTransactionService;
}(IgxBaseTransactionService));
export { IgxTransactionService };
function IgxTransactionService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTransactionService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTransactionService.ctorParameters;
    /** @type {?} */
    IgxTransactionService.prototype._transactions;
    /** @type {?} */
    IgxTransactionService.prototype._redoStack;
    /** @type {?} */
    IgxTransactionService.prototype._undoStack;
    /** @type {?} */
    IgxTransactionService.prototype._states;
    /** @type {?} */
    IgxTransactionService.prototype.onStateUpdate;
}
