// igniteui-angular/esm5/lib/services/transaction/transaction.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
var TransactionType = {
    ADD: 'add',
    DELETE: 'delete',
    UPDATE: 'update',
};
export { TransactionType };
/**
 * @record
 */
export function Transaction() { }
function Transaction_tsickle_Closure_declarations() {
    /** @type {?} */
    Transaction.prototype.id;
    /** @type {?} */
    Transaction.prototype.type;
    /** @type {?} */
    Transaction.prototype.newValue;
}
/**
 * \@experimental \@hidden
 * @record
 */
export function HierarchicalTransaction() { }
function HierarchicalTransaction_tsickle_Closure_declarations() {
    /** @type {?} */
    HierarchicalTransaction.prototype.parentId;
}
/**
 * @record
 */
export function State() { }
function State_tsickle_Closure_declarations() {
    /** @type {?} */
    State.prototype.value;
    /** @type {?} */
    State.prototype.recordRef;
    /** @type {?} */
    State.prototype.type;
}
/**
 * \@experimental \@hidden
 * @record
 */
export function HierarchicalState() { }
function HierarchicalState_tsickle_Closure_declarations() {
    /** @type {?} */
    HierarchicalState.prototype.parentId;
}
/**
 * \@experimental \@hidden
 * @record
 */
export function HierarchicalTransactionNode() { }
function HierarchicalTransactionNode_tsickle_Closure_declarations() {
    /** @type {?} */
    HierarchicalTransactionNode.prototype.id;
    /** @type {?|undefined} */
    HierarchicalTransactionNode.prototype.parentId;
    /** @type {?} */
    HierarchicalTransactionNode.prototype.childNodes;
}
// unsupported: template constraints.
// unsupported: template constraints.
/**
 * @record
 * @template T, S
 */
export function TransactionService() { }
function TransactionService_tsickle_Closure_declarations() {
    /**
     * Returns whether transaction is enabled for this service
     * @type {?}
     */
    TransactionService.prototype.enabled;
    /**
     * Event fired when transaction state has changed - add transaction, commit all transactions, undo and redo.
     * @type {?|undefined}
     */
    TransactionService.prototype.onStateUpdate;
    /**
     * \@return if there are any transactions in the Undo stack
     * @type {?}
     */
    TransactionService.prototype.canUndo;
    /**
     * \@return if there are any transactions in the Redo stack
     * @type {?}
     */
    TransactionService.prototype.canRedo;
    /**
     * Adds provided  transaction with recordRef if any
     * \@param transaction Transaction to be added
     * \@param recordRef Reference to the value of the record in the data source related to the changed item
     * @type {?}
     */
    TransactionService.prototype.add;
    /**
     * Returns all recorded transactions in chronological order.
     * \@param id Optional record id to get transactions for
     * \@return All transaction in the service or for the specified record
     * @type {?}
     */
    TransactionService.prototype.getTransactionLog;
    /**
     * Remove the last transaction if any
     * @type {?}
     */
    TransactionService.prototype.undo;
    /**
     * Applies the last undone transaction if any
     * @type {?}
     */
    TransactionService.prototype.redo;
    /**
     * Returns aggregated changes from all transactions
     * \@param mergeChanges If set to true will merge each state's value over relate recordRef
     * and will record resulting value in the related transaction
     * \@return Collection of aggregated transactions for each changed record
     * @type {?}
     */
    TransactionService.prototype.getAggregatedChanges;
    /**
     * Returns the state of the record with provided id
     * \@param id The id of the record
     * \@return State of the record if any
     * @type {?}
     */
    TransactionService.prototype.getState;
    /**
     * Returns value of the required id including all uncommitted changes
     * \@param id The id of the record to return value for
     * \@param mergeChanges If set to true will merge state's value over relate recordRef
     * and will return merged value
     * \@return Value with changes or **null**
     * @type {?}
     */
    TransactionService.prototype.getAggregatedValue;
    /**
     * Applies all transactions over the provided data
     * \@param data Data source to update
     * @type {?}
     */
    TransactionService.prototype.commit;
    /**
     * Clears all transactions
     * @type {?}
     */
    TransactionService.prototype.clear;
    /**
     * Starts pending transactions. All transactions passed after call to startPending
     * will not be added to transaction log
     * @type {?}
     */
    TransactionService.prototype.startPending;
    /**
     * Clears all pending transactions and aggregated pending state. If commit is set to true
     * commits pending states as single transaction
     * \@param commit Should commit the pending states
     * @type {?}
     */
    TransactionService.prototype.endPending;
}
