// igniteui-angular/esm5/lib/dialog/dialog.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, HostBinding } from '@angular/core';
/**
 * @hidden
 */
var IgxDialogTitleDirective = /** @class */ (function () {
    function IgxDialogTitleDirective() {
        this.defaultStyle = true;
    }
    IgxDialogTitleDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'igx-dialog-title,[igxDialogTitle]'
                },] },
    ];
    /** @nocollapse */
    IgxDialogTitleDirective.propDecorators = {
        "defaultStyle": [{ type: HostBinding, args: ['class.igx-dialog__window-title',] },],
    };
    return IgxDialogTitleDirective;
}());
export { IgxDialogTitleDirective };
function IgxDialogTitleDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxDialogTitleDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxDialogTitleDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxDialogTitleDirective.propDecorators;
    /** @type {?} */
    IgxDialogTitleDirective.prototype.defaultStyle;
}
/**
 * @hidden
 */
var IgxDialogActionsDirective = /** @class */ (function () {
    function IgxDialogActionsDirective() {
        this.defaultClass = true;
    }
    IgxDialogActionsDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'igx-dialog-actions,[igxDialogActions]'
                },] },
    ];
    /** @nocollapse */
    IgxDialogActionsDirective.propDecorators = {
        "defaultClass": [{ type: HostBinding, args: ['class.igx-dialog__window-actions',] },],
    };
    return IgxDialogActionsDirective;
}());
export { IgxDialogActionsDirective };
function IgxDialogActionsDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxDialogActionsDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxDialogActionsDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxDialogActionsDirective.propDecorators;
    /** @type {?} */
    IgxDialogActionsDirective.prototype.defaultClass;
}
