// igniteui-angular/esm5/lib/expansion-panel/expansion-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectorRef, EventEmitter, ElementRef, HostBinding, Input, Output, ContentChild, forwardRef, } from '@angular/core';
import { AnimationBuilder, useAnimation } from '@angular/animations';
import { growVerOut, growVerIn } from '../animations/main';
import { IgxExpansionPanelBodyComponent } from './expansion-panel-body.component';
import { IgxExpansionPanelHeaderComponent } from './expansion-panel-header.component';
var /** @type {?} */ NEXT_ID = 0;
var IgxExpansionPanelComponent = /** @class */ (function () {
    function IgxExpansionPanelComponent(cdr, elementRef, builder) {
        this.cdr = cdr;
        this.elementRef = elementRef;
        this.builder = builder;
        this.animationSettings = {
            openAnimation: growVerIn,
            closeAnimation: growVerOut
        };
        /**
         * Sets/gets the `id` of the expansion panel component.
         * If not set, `id` will have value `"igx-expansion-panel-0"`;
         * ```html
         * <igx-expansion-panel id = "my-first-expansion-panel"></igx-expansion-panel>
         * ```
         * ```typescript
         * let panelId =  this.panel.id;
         * ```
         * \@memberof IgxExpansionPanelComponent
         */
        this.id = "igx-expansion-panel-" + NEXT_ID++;
        this.cssClass = 'igx-expansion-panel';
        this.collapsed = true;
        this.onCollapsed = new EventEmitter();
        this.onExpanded = new EventEmitter();
    }
    Object.defineProperty(IgxExpansionPanelComponent.prototype, "headerId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.header ? this.id + "-header" : '';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} cb
     * @return {?}
     */
    IgxExpansionPanelComponent.prototype.playOpenAnimation = /**
     * @param {?} cb
     * @return {?}
     */
    function (cb) {
        if (!this.body) {
            // if not body element is passed, there is nothing to animate
            return;
        }
        var /** @type {?} */ animation = useAnimation(this.animationSettings.openAnimation);
        var /** @type {?} */ animationBuilder = this.builder.build(animation);
        var /** @type {?} */ openAnimationPlayer = animationBuilder.create(this.body.element.nativeElement);
        openAnimationPlayer.onDone(function () {
            cb();
            openAnimationPlayer.reset();
        });
        openAnimationPlayer.play();
    };
    /**
     * @param {?} cb
     * @return {?}
     */
    IgxExpansionPanelComponent.prototype.playCloseAnimation = /**
     * @param {?} cb
     * @return {?}
     */
    function (cb) {
        if (!this.body) {
            // if not body element is passed, there is nothing to animate
            return;
        }
        var /** @type {?} */ animation = useAnimation(this.animationSettings.closeAnimation);
        var /** @type {?} */ animationBuilder = this.builder.build(animation);
        var /** @type {?} */ closeAnimationPlayer = animationBuilder.create(this.body.element.nativeElement);
        closeAnimationPlayer.onDone(function () {
            cb();
            closeAnimationPlayer.reset();
        });
        closeAnimationPlayer.play();
    };
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelComponent.prototype.collapse = /**
     * @param {?=} evt
     * @return {?}
     */
    function (evt) {
        var _this = this;
        this.playCloseAnimation(function () {
            _this.onCollapsed.emit({ event: evt, panel: _this });
            _this.collapsed = true;
        });
    };
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelComponent.prototype.expand = /**
     * @param {?=} evt
     * @return {?}
     */
    function (evt) {
        var _this = this;
        this.collapsed = false;
        this.cdr.detectChanges();
        this.playOpenAnimation(function () {
            _this.onExpanded.emit({ event: evt, panel: _this });
        });
    };
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelComponent.prototype.toggle = /**
     * @param {?=} evt
     * @return {?}
     */
    function (evt) {
        if (this.collapsed) {
            this.expand(evt);
        }
        else {
            this.collapse(evt);
        }
    };
    IgxExpansionPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'igx-expansion-panel',
                    template: "<ng-content select=\"igx-expansion-panel-header\"></ng-content>\n<ng-content *ngIf=\"!collapsed\" select=\"igx-expansion-panel-body\"></ng-content>\n"
                },] },
    ];
    /** @nocollapse */
    IgxExpansionPanelComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef, },
        { type: ElementRef, },
        { type: AnimationBuilder, },
    ]; };
    IgxExpansionPanelComponent.propDecorators = {
        "animationSettings": [{ type: Input },],
        "id": [{ type: HostBinding, args: ['attr.id',] }, { type: Input },],
        "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel',] },],
        "collapsed": [{ type: Input },],
        "onCollapsed": [{ type: Output },],
        "onExpanded": [{ type: Output },],
        "body": [{ type: ContentChild, args: [forwardRef(function () { return IgxExpansionPanelBodyComponent; }), { read: IgxExpansionPanelBodyComponent },] },],
        "header": [{ type: ContentChild, args: [forwardRef(function () { return IgxExpansionPanelHeaderComponent; }), { read: IgxExpansionPanelHeaderComponent },] },],
    };
    return IgxExpansionPanelComponent;
}());
export { IgxExpansionPanelComponent };
function IgxExpansionPanelComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelComponent.propDecorators;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.animationSettings;
    /**
     * Sets/gets the `id` of the expansion panel component.
     * If not set, `id` will have value `"igx-expansion-panel-0"`;
     * ```html
     * <igx-expansion-panel id = "my-first-expansion-panel"></igx-expansion-panel>
     * ```
     * ```typescript
     * let panelId =  this.panel.id;
     * ```
     * \@memberof IgxExpansionPanelComponent
     * @type {?}
     */
    IgxExpansionPanelComponent.prototype.id;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.cssClass;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.collapsed;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.onCollapsed;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.onExpanded;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.body;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.header;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.cdr;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.elementRef;
    /** @type {?} */
    IgxExpansionPanelComponent.prototype.builder;
}
