// igniteui-angular/esm5/lib/expansion-panel/expansion-panel-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectorRef, ElementRef, HostBinding, HostListener, Input, Host, EventEmitter, Output, ContentChild } from '@angular/core';
import { IgxExpansionPanelComponent } from './expansion-panel.component';
import { IgxExpansionPanelIconDirective } from './expansion-panel.directives';
/**
 * @record
 */
export function IExpansionPanelEventArgs() { }
function IExpansionPanelEventArgs_tsickle_Closure_declarations() {
    /** @type {?} */
    IExpansionPanelEventArgs.prototype.event;
    /** @type {?} */
    IExpansionPanelEventArgs.prototype.panel;
}
/** @enum {string} */
var ICON_POSITION = {
    LEFT: 'left',
    NONE: 'none',
    RIGHT: 'right',
};
export { ICON_POSITION };
var IgxExpansionPanelHeaderComponent = /** @class */ (function () {
    function IgxExpansionPanelHeaderComponent(panel, cdr, elementRef) {
        this.panel = panel;
        this.cdr = cdr;
        this.elementRef = elementRef;
        this._iconTemplate = false;
        this.id = '';
        this.lv = '3';
        this.role = 'heading';
        this.iconPosition = ICON_POSITION.LEFT;
        this.onInteraction = new EventEmitter();
        this.cssClass = 'igx-expansion-panel__header';
        this.disabled = false;
        this.id = this.panel.id + "-header";
    }
    Object.defineProperty(IgxExpansionPanelHeaderComponent.prototype, "iconTemplate", {
        get: /**
         * @return {?}
         */
        function () {
            return this._iconTemplate;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._iconTemplate = /** @type {?} */ (val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxExpansionPanelHeaderComponent.prototype, "controls", {
        get: /**
         * @return {?}
         */
        function () {
            return this.panel.id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxExpansionPanelHeaderComponent.prototype, "isExpanded", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.panel.collapsed;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?=} evt
     * @return {?}
     */
    IgxExpansionPanelHeaderComponent.prototype.onAction = /**
     * @param {?=} evt
     * @return {?}
     */
    function (evt) {
        if (this.disabled) {
            evt.stopPropagation();
            return;
        }
        this.onInteraction.emit({ event: evt, panel: this.panel });
        this.panel.toggle(evt);
        evt.preventDefault();
    };
    Object.defineProperty(IgxExpansionPanelHeaderComponent.prototype, "iconPositionClass", {
        get: /**
         * @return {?}
         */
        function () {
            switch (this.iconPosition) {
                case (ICON_POSITION.LEFT):
                    return "igx-expansion-panel__header-icon--start";
                case (ICON_POSITION.RIGHT):
                    return "igx-expansion-panel__header-icon--end";
                case (ICON_POSITION.NONE):
                    return "igx-expansion-panel__header-icon--none";
                default:
                    return '';
            }
        },
        enumerable: true,
        configurable: true
    });
    IgxExpansionPanelHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'igx-expansion-panel-header',
                    template: "<div class=\"igx-expansion-panel__header-inner\" tabindex=\"0\" role=\"button\" [attr.id]=\"id\"\n[attr.aria-disabled]=\"disabled\" [attr.aria-expanded]=\"isExpanded\" [attr.aria-controls]=\"controls\">\n    <div class=\"igx-expansion-panel__title-wrapper\">\n        <ng-content select=\"igx-expansion-panel-title\"></ng-content>\n        <ng-content select=\"igx-expansion-panel-description\"></ng-content>\n    </div>\n    <ng-content></ng-content>\n    <div [class]=\"iconPositionClass\">\n        <ng-content *ngIf=\"iconTemplate\" select=\"igx-expansion-panel-icon\"></ng-content>\n        <igx-icon *ngIf=\"!iconTemplate\" fontSet=\"material\">\n            {{panel.collapsed? 'expand_more':'expand_less'}}\n        </igx-icon>\n    </div>\n</div>\n"
                },] },
    ];
    /** @nocollapse */
    IgxExpansionPanelHeaderComponent.ctorParameters = function () { return [
        { type: IgxExpansionPanelComponent, decorators: [{ type: Host },] },
        { type: ChangeDetectorRef, },
        { type: ElementRef, },
    ]; };
    IgxExpansionPanelHeaderComponent.propDecorators = {
        "iconTemplate": [{ type: ContentChild, args: [IgxExpansionPanelIconDirective,] },],
        "lv": [{ type: HostBinding, args: ['attr.aria-level',] }, { type: Input },],
        "role": [{ type: HostBinding, args: ['attr.role',] }, { type: Input },],
        "iconPosition": [{ type: Input },],
        "onInteraction": [{ type: Output },],
        "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel__header',] },],
        "isExpanded": [{ type: HostBinding, args: ['class.igx-expansion-panel__header--expanded',] },],
        "disabled": [{ type: Input }, { type: HostBinding, args: ['class.igx-expansion-panel--disabled',] },],
        "onAction": [{ type: HostListener, args: ['keydown.Enter', ['$event'],] }, { type: HostListener, args: ['keydown.Space', ['$event'],] }, { type: HostListener, args: ['keydown.Spacebar', ['$event'],] }, { type: HostListener, args: ['click', ['$event'],] },],
    };
    return IgxExpansionPanelHeaderComponent;
}());
export { IgxExpansionPanelHeaderComponent };
function IgxExpansionPanelHeaderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelHeaderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelHeaderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelHeaderComponent.propDecorators;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype._iconTemplate;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.id;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.lv;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.role;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.iconPosition;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.onInteraction;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.cssClass;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.disabled;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.panel;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.cdr;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.elementRef;
}
