// igniteui-angular/esm5/lib/expansion-panel/expansion-panel-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding, Inject, forwardRef, ElementRef, Input, ChangeDetectorRef } from '@angular/core';
import { IgxExpansionPanelComponent } from './expansion-panel.component';
var IgxExpansionPanelBodyComponent = /** @class */ (function () {
    function IgxExpansionPanelBodyComponent(panel, element, cdr) {
        this.panel = panel;
        this.element = element;
        this.cdr = cdr;
        this._labelledBy = '';
        this._label = '';
        this.cssClass = "igx-expansion-panel__body";
        this._title = '';
        this.role = 'region';
    }
    Object.defineProperty(IgxExpansionPanelBodyComponent.prototype, "label", {
        get: /**
         * @return {?}
         */
        function () {
            return this._label || this.panel.id + '-region';
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._label = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxExpansionPanelBodyComponent.prototype, "labelledBy", {
        get: /**
         * @return {?}
         */
        function () {
            return this._labelledBy;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._labelledBy = val;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxExpansionPanelBodyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.labelledBy = this.panel.headerId;
        this.label = this.panel.id + '-region';
    };
    IgxExpansionPanelBodyComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'igx-expansion-panel-body',
                    template: "<ng-content></ng-content>"
                },] },
    ];
    /** @nocollapse */
    IgxExpansionPanelBodyComponent.ctorParameters = function () { return [
        { type: IgxExpansionPanelComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return IgxExpansionPanelComponent; }),] },] },
        { type: ElementRef, },
        { type: ChangeDetectorRef, },
    ]; };
    IgxExpansionPanelBodyComponent.propDecorators = {
        "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel__body',] },],
        "label": [{ type: Input }, { type: HostBinding, args: ['attr.aria-label',] },],
        "labelledBy": [{ type: Input }, { type: HostBinding, args: ['attr.aria-labelledby',] },],
        "role": [{ type: Input }, { type: HostBinding, args: ['attr.role',] },],
    };
    return IgxExpansionPanelBodyComponent;
}());
export { IgxExpansionPanelBodyComponent };
function IgxExpansionPanelBodyComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelBodyComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelBodyComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelBodyComponent.propDecorators;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype._labelledBy;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype._label;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.cssClass;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype._title;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.role;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.panel;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.element;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.cdr;
}
