// igniteui-angular/esm5/lib/services/overlay/position/connected-positioning-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { HorizontalAlignment, VerticalAlignment, getPointFromPositionsSettings } from './../utilities';
import { scaleInVerTop, scaleOutVerTop } from '../../../animations/main';
var ConnectedPositioningStrategy = /** @class */ (function () {
    function ConnectedPositioningStrategy(settings) {
        this._defaultSettings = {
            // default Point(0, 0) in getPointFromPositionsSettings
            target: null,
            horizontalDirection: HorizontalAlignment.Right,
            verticalDirection: VerticalAlignment.Bottom,
            horizontalStartPoint: HorizontalAlignment.Left,
            verticalStartPoint: VerticalAlignment.Bottom,
            openAnimation: scaleInVerTop,
            closeAnimation: scaleOutVerTop
        };
        this.settings = Object.assign({}, this._defaultSettings, settings);
    }
    // we no longer use the element inside the position() as its dimensions are cached in rect
    /**
     * @param {?} contentElement
     * @param {?} size
     * @param {?=} document
     * @param {?=} initialCall
     * @return {?}
     */
    ConnectedPositioningStrategy.prototype.position = /**
     * @param {?} contentElement
     * @param {?} size
     * @param {?=} document
     * @param {?=} initialCall
     * @return {?}
     */
    function (contentElement, size, document, initialCall) {
        var /** @type {?} */ startPoint = getPointFromPositionsSettings(this.settings, contentElement.parentElement);
        contentElement.style.top = startPoint.y + this.settings.verticalDirection * size.height + 'px';
        contentElement.style.left = startPoint.x + this.settings.horizontalDirection * size.width + 'px';
    };
    return ConnectedPositioningStrategy;
}());
export { ConnectedPositioningStrategy };
function ConnectedPositioningStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    ConnectedPositioningStrategy.prototype._defaultSettings;
    /** @type {?} */
    ConnectedPositioningStrategy.prototype.settings;
}
