// igniteui-angular/esm5/lib/grids/grid.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DOCUMENT, DatePipe, DecimalPipe } from '@angular/common';
import { ChangeDetectorRef, Directive, ElementRef, HostListener, Inject, Injectable, Input, NgZone, Output, Pipe, Renderer2, TemplateRef } from '@angular/core';
import { animationFrameScheduler, fromEvent, interval, Subject } from 'rxjs';
import { map, switchMap, takeUntil, throttle } from 'rxjs/operators';
import { IgxColumnComponent } from './column.component';
import { IgxDragDirective, IgxDropDirective } from '../directives/dragdrop/dragdrop.directive';
import { IgxGridForOfDirective } from '../directives/for-of/for_of.directive';
import { ConnectedPositioningStrategy } from '../services';
import { getPointFromPositionsSettings, VerticalAlignment } from '../services/overlay/utilities';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
var IgxColumnResizerDirective = /** @class */ (function () {
    function IgxColumnResizerDirective(element, document, zone) {
        var _this = this;
        this.element = element;
        this.document = document;
        this.zone = zone;
        this.restrictHResizeMin = Number.MIN_SAFE_INTEGER;
        this.restrictHResizeMax = Number.MAX_SAFE_INTEGER;
        this.resizeEndTimeout = 0;
        this.resizeEnd = new Subject();
        this.resizeStart = new Subject();
        this.resize = new Subject();
        this._destroy = new Subject();
        this.resizeStart.pipe(map(function (event) { return event.clientX; }), takeUntil(this._destroy), switchMap(function (offset) {
            return _this.resize.pipe(map(function (event) { return event.clientX - offset; }), takeUntil(_this.resizeEnd), takeUntil(_this._destroy));
        })).subscribe(function (pos) {
            var /** @type {?} */ left = _this._left + pos;
            _this.left = left < _this.restrictHResizeMin ? _this.restrictHResizeMin + 'px' : left + 'px';
            if (left > _this.restrictHResizeMax) {
                _this.left = _this.restrictHResizeMax + 'px';
            }
            else if (left > _this.restrictHResizeMin) {
                _this.left = left + 'px';
            }
        });
    }
    /**
     * @return {?}
     */
    IgxColumnResizerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.zone.runOutsideAngular(function () {
            fromEvent(_this.document.defaultView, 'mousedown').pipe(takeUntil(_this._destroy))
                .subscribe(function (res) { return _this.onMousedown(res); });
            fromEvent(_this.document.defaultView, 'mousemove').pipe(throttle(function () { return interval(0, animationFrameScheduler); }), takeUntil(_this._destroy)).subscribe(function (res) { return _this.onMousemove(res); });
            fromEvent(_this.document.defaultView, 'mouseup').pipe(takeUntil(_this._destroy))
                .subscribe(function (res) { return _this.onMouseup(res); });
        });
    };
    /**
     * @return {?}
     */
    IgxColumnResizerDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._destroy.next(true);
        this._destroy.unsubscribe();
    };
    Object.defineProperty(IgxColumnResizerDirective.prototype, "left", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            var _this = this;
            requestAnimationFrame(function () { return _this.element.nativeElement.style.left = val; });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnResizerDirective.prototype.onMouseup = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        setTimeout(function () {
            _this.resizeEnd.next(event);
            _this.resizeEnd.complete();
        }, this.resizeEndTimeout);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnResizerDirective.prototype.onMousedown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.resizeStart.next(event);
        event.preventDefault();
        var /** @type {?} */ elStyle = this.document.defaultView.getComputedStyle(this.element.nativeElement);
        this._left = Number.isNaN(parseInt(elStyle.left, 10)) ? 0 : parseInt(elStyle.left, 10);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnResizerDirective.prototype.onMousemove = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.resize.next(event);
        event.preventDefault();
    };
    IgxColumnResizerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxResizer]'
                },] },
    ];
    /** @nocollapse */
    IgxColumnResizerDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
        { type: NgZone, },
    ]; };
    IgxColumnResizerDirective.propDecorators = {
        "restrictHResizeMin": [{ type: Input },],
        "restrictHResizeMax": [{ type: Input },],
        "resizeEndTimeout": [{ type: Input },],
        "resizeEnd": [{ type: Output },],
        "resizeStart": [{ type: Output },],
        "resize": [{ type: Output },],
    };
    return IgxColumnResizerDirective;
}());
export { IgxColumnResizerDirective };
function IgxColumnResizerDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnResizerDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnResizerDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxColumnResizerDirective.propDecorators;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.restrictHResizeMin;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.restrictHResizeMax;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.resizeEndTimeout;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.resizeEnd;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.resizeStart;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.resize;
    /** @type {?} */
    IgxColumnResizerDirective.prototype._left;
    /** @type {?} */
    IgxColumnResizerDirective.prototype._destroy;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.element;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.document;
    /** @type {?} */
    IgxColumnResizerDirective.prototype.zone;
}
/**
 * @hidden
 */
var IgxCellTemplateDirective = /** @class */ (function () {
    function IgxCellTemplateDirective(template) {
        this.template = template;
    }
    IgxCellTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxCell]'
                },] },
    ];
    /** @nocollapse */
    IgxCellTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxCellTemplateDirective;
}());
export { IgxCellTemplateDirective };
function IgxCellTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCellTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCellTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxCellTemplateDirective.prototype.template;
}
/**
 * @hidden
 */
var IgxCellHeaderTemplateDirective = /** @class */ (function () {
    function IgxCellHeaderTemplateDirective(template) {
        this.template = template;
    }
    IgxCellHeaderTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxHeader]'
                },] },
    ];
    /** @nocollapse */
    IgxCellHeaderTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxCellHeaderTemplateDirective;
}());
export { IgxCellHeaderTemplateDirective };
function IgxCellHeaderTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCellHeaderTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCellHeaderTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxCellHeaderTemplateDirective.prototype.template;
}
/**
 * @hidden
 */
var IgxCellFooterTemplateDirective = /** @class */ (function () {
    function IgxCellFooterTemplateDirective(template) {
        this.template = template;
    }
    IgxCellFooterTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxFooter]'
                },] },
    ];
    /** @nocollapse */
    IgxCellFooterTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxCellFooterTemplateDirective;
}());
export { IgxCellFooterTemplateDirective };
function IgxCellFooterTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCellFooterTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCellFooterTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxCellFooterTemplateDirective.prototype.template;
}
/**
 * @hidden
 */
var IgxCellEditorTemplateDirective = /** @class */ (function () {
    function IgxCellEditorTemplateDirective(template) {
        this.template = template;
    }
    IgxCellEditorTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxCellEditor]'
                },] },
    ];
    /** @nocollapse */
    IgxCellEditorTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxCellEditorTemplateDirective;
}());
export { IgxCellEditorTemplateDirective };
function IgxCellEditorTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCellEditorTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCellEditorTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxCellEditorTemplateDirective.prototype.template;
}
/**
 * @hidden
 */
var IgxColumnMovingService = /** @class */ (function () {
    function IgxColumnMovingService() {
    }
    Object.defineProperty(IgxColumnMovingService.prototype, "column", {
        get: /**
         * @return {?}
         */
        function () {
            return this._column;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this._column = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingService.prototype, "icon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._icon;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this._icon = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    IgxColumnMovingService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */ IgxColumnMovingService.ngInjectableDef = i0.defineInjectable({ factory: function IgxColumnMovingService_Factory() { return new IgxColumnMovingService(); }, token: IgxColumnMovingService, providedIn: "root" });
    return IgxColumnMovingService;
}());
export { IgxColumnMovingService };
function IgxColumnMovingService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnMovingService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnMovingService.ctorParameters;
    /** @type {?} */
    IgxColumnMovingService.prototype._icon;
    /** @type {?} */
    IgxColumnMovingService.prototype._column;
    /** @type {?} */
    IgxColumnMovingService.prototype.cancelDrop;
    /** @type {?} */
    IgxColumnMovingService.prototype.selection;
}
/** @enum {number} */
var DropPosition = {
    BeforeDropTarget: 0,
    AfterDropTarget: 1,
    None: 2,
};
export { DropPosition };
DropPosition[DropPosition.BeforeDropTarget] = "BeforeDropTarget";
DropPosition[DropPosition.AfterDropTarget] = "AfterDropTarget";
DropPosition[DropPosition.None] = "None";
/**
 * @hidden
 */
var IgxColumnMovingDragDirective = /** @class */ (function (_super) {
    tslib_1.__extends(IgxColumnMovingDragDirective, _super);
    function IgxColumnMovingDragDirective(_element, _zone, _renderer, _cdr, cms) {
        var _this = _super.call(this, _cdr, _element, _zone, _renderer) || this;
        _this.cms = cms;
        _this._ghostImageClass = 'igx-grid__drag-ghost-image';
        _this._dragGhostImgIconClass = 'igx-grid__drag-ghost-image-icon';
        _this._dragGhostImgIconGroupClass = 'igx-grid__drag-ghost-image-icon-group';
        return _this;
    }
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "data", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._column = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "column", {
        get: /**
         * @return {?}
         */
        function () {
            return this._column;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "draggable", {
        get: /**
         * @return {?}
         */
        function () {
            return this.column && (this.column.movable || this.column.groupable);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDragDirective.prototype, "icon", {
        get: /**
         * @return {?}
         */
        function () {
            return this.cms.icon;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDragDirective.prototype.onEscape = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.cms.cancelDrop = true;
        this.onPointerUp(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDragDirective.prototype.onPointerDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.draggable || event.target.getAttribute('draggable') === 'false') {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        this._removeOnDestroy = false;
        this.cms.column = this.column;
        this.ghostImageClass = this._ghostImageClass;
        _super.prototype.onPointerDown.call(this, event);
        this.column.grid.isColumnMoving = true;
        this.column.grid.cdr.detectChanges();
        var /** @type {?} */ currSelection = this.column.grid.selection.first_item(this.column.gridID + '-cell');
        if (currSelection) {
            this.cms.selection = {
                column: this.column.grid.columnList.toArray()[currSelection.columnID],
                rowID: currSelection.rowID
            };
        }
        var /** @type {?} */ args = {
            source: this.column
        };
        this.column.grid.onColumnMovingStart.emit(args);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDragDirective.prototype.onPointerMove = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        _super.prototype.onPointerMove.call(this, event);
        if (this._dragStarted && this._dragGhost && !this.column.grid.draggedColumn) {
            this.column.grid.draggedColumn = this.column;
            this.column.grid.cdr.detectChanges();
        }
        if (this.column.grid.isColumnMoving) {
            var /** @type {?} */ args = {
                source: this.column,
                cancel: false
            };
            this.column.grid.onColumnMoving.emit(args);
            if (args.cancel) {
                this.onEscape(event);
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDragDirective.prototype.onPointerUp = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        // Run it explicitly inside the zone because sometimes onPointerUp executes after the code below.
        this.zone.run(function () {
            _super.prototype.onPointerUp.call(_this, event);
            _this.column.grid.isColumnMoving = false;
            _this.column.grid.draggedColumn = null;
            _this.column.grid.cdr.detectChanges();
        });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDragDirective.prototype.createDragGhost = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        _super.prototype.createDragGhost.call(this, event);
        var /** @type {?} */ pageX, /** @type {?} */ pageY;
        if (this.pointerEventsEnabled || !this.touchEventsEnabled) {
            pageX = event.pageX;
            pageY = event.pageY;
        }
        else {
            pageX = event.touches[0].pageX;
            pageY = event.touches[0].pageY;
        }
        this._dragGhost.style.height = null;
        this._dragGhost.style.minWidth = null;
        this._dragGhost.style.flexBasis = null;
        this._dragGhost.style.position = null;
        var /** @type {?} */ icon = document.createElement('i');
        var /** @type {?} */ text = document.createTextNode('block');
        icon.appendChild(text);
        icon.classList.add('material-icons');
        this.cms.icon = icon;
        if (!this.column.columnGroup) {
            this.renderer.addClass(icon, this._dragGhostImgIconClass);
            this._dragGhost.removeChild(this._dragGhost.children[2]);
            this._dragGhost.insertBefore(icon, this._dragGhost.children[1]);
            this.left = this._dragStartX = pageX - ((this._dragGhost.getBoundingClientRect().width / 3) * 2);
            this.top = this._dragStartY = pageY - ((this._dragGhost.getBoundingClientRect().height / 3) * 2);
        }
        else {
            this._dragGhost.removeChild(this._dragGhost.children[2]);
            this._dragGhost.removeChild(this._dragGhost.firstElementChild);
            this._dragGhost.removeChild(this._dragGhost.lastElementChild);
            this._dragGhost.insertBefore(icon, this._dragGhost.firstElementChild);
            this.renderer.addClass(icon, this._dragGhostImgIconGroupClass);
            this._dragGhost.children[1].style.paddingLeft = '0px';
            this.left = this._dragStartX = pageX - ((this._dragGhost.getBoundingClientRect().width / 3) * 2);
            this.top = this._dragStartY = pageY - ((this._dragGhost.getBoundingClientRect().height / 3) * 2);
        }
    };
    IgxColumnMovingDragDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxColumnMovingDrag]'
                },] },
    ];
    /** @nocollapse */
    IgxColumnMovingDragDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: NgZone, },
        { type: Renderer2, },
        { type: ChangeDetectorRef, },
        { type: IgxColumnMovingService, },
    ]; };
    IgxColumnMovingDragDirective.propDecorators = {
        "data": [{ type: Input, args: ['igxColumnMovingDrag',] },],
        "onEscape": [{ type: HostListener, args: ['document:keydown.escape', ['$event'],] },],
    };
    return IgxColumnMovingDragDirective;
}(IgxDragDirective));
export { IgxColumnMovingDragDirective };
function IgxColumnMovingDragDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnMovingDragDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnMovingDragDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxColumnMovingDragDirective.propDecorators;
    /** @type {?} */
    IgxColumnMovingDragDirective.prototype._column;
    /** @type {?} */
    IgxColumnMovingDragDirective.prototype._ghostImageClass;
    /** @type {?} */
    IgxColumnMovingDragDirective.prototype._dragGhostImgIconClass;
    /** @type {?} */
    IgxColumnMovingDragDirective.prototype._dragGhostImgIconGroupClass;
    /** @type {?} */
    IgxColumnMovingDragDirective.prototype.cms;
}
/**
 * @hidden
 */
var IgxColumnMovingDropDirective = /** @class */ (function (_super) {
    tslib_1.__extends(IgxColumnMovingDropDirective, _super);
    function IgxColumnMovingDropDirective(elementRef, renderer, zone, cms) {
        var _this = _super.call(this, elementRef, renderer, zone) || this;
        _this.elementRef = elementRef;
        _this.renderer = renderer;
        _this.zone = zone;
        _this.cms = cms;
        _this._dropIndicator = null;
        _this._lastDropIndicator = null;
        _this._dragLeave = new Subject();
        _this._dropIndicatorClass = 'igx-grid__th-drop-indicator--active';
        return _this;
    }
    Object.defineProperty(IgxColumnMovingDropDirective.prototype, "data", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val instanceof IgxColumnComponent) {
                this._column = val;
            }
            if (val instanceof IgxGridForOfDirective) {
                this._hVirtDir = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDropDirective.prototype, "column", {
        get: /**
         * @return {?}
         */
        function () {
            return this._column;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDropDirective.prototype, "isDropTarget", {
        get: /**
         * @return {?}
         */
        function () {
            return this._column && this._column.grid.hasMovableColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingDropDirective.prototype, "horizontalScroll", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._hVirtDir) {
                return this._hVirtDir;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxColumnMovingDropDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._dragLeave.next(true);
        this._dragLeave.complete();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDropDirective.prototype.onDragOver = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.isDropTarget &&
            this.cms.column !== this.column &&
            this.cms.column.level === this.column.level &&
            this.cms.column.parent === this.column.parent) {
            if (this._lastDropIndicator) {
                this.renderer.removeClass(this._dropIndicator, this._dropIndicatorClass);
            }
            var /** @type {?} */ pos = this.elementRef.nativeElement.getBoundingClientRect().left + parseFloat(this.column.width) / 2;
            if (event.detail.pageX < pos) {
                this._dropPos = DropPosition.BeforeDropTarget;
                this._lastDropIndicator = this._dropIndicator = this.elementRef.nativeElement.firstElementChild;
            }
            else {
                this._dropPos = DropPosition.AfterDropTarget;
                this._lastDropIndicator = this._dropIndicator = this.elementRef.nativeElement.lastElementChild;
            }
            if (this.cms.icon.innerText !== 'block') {
                this.renderer.addClass(this._dropIndicator, this._dropIndicatorClass);
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDropDirective.prototype.onDragEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        var /** @type {?} */ drag = event.detail.owner;
        if (!(drag instanceof IgxColumnMovingDragDirective)) {
            return;
        }
        if (this.isDropTarget &&
            this.cms.column !== this.column &&
            this.cms.column.level === this.column.level &&
            this.cms.column.parent === this.column.parent) {
            if (!this.column.pinned || (this.column.pinned && this.cms.column.pinned)) {
                this.cms.icon.innerText = 'swap_horiz';
            }
            if (!this.cms.column.pinned && this.column.pinned) {
                var /** @type {?} */ nextPinnedWidth = this.column.grid.getPinnedWidth(true) + parseFloat(this.cms.column.width);
                if (nextPinnedWidth <= this.column.grid.calcPinnedContainerMaxWidth) {
                    this.cms.icon.innerText = 'lock';
                }
                else {
                    this.cms.icon.innerText = 'block';
                }
            }
        }
        else {
            this.cms.icon.innerText = 'block';
        }
        if (this.horizontalScroll) {
            this.cms.icon.innerText = event.target.id === 'right' ? 'arrow_forward' : 'arrow_back';
            interval(100).pipe(takeUntil(this._dragLeave)).subscribe(function (val) {
                event.target.id === 'right' ? _this.horizontalScroll.getHorizontalScroll().scrollLeft += 15 :
                    _this.horizontalScroll.getHorizontalScroll().scrollLeft -= 15;
            });
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDropDirective.prototype.onDragLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var /** @type {?} */ drag = event.detail.owner;
        if (!(drag instanceof IgxColumnMovingDragDirective)) {
            return;
        }
        this.cms.icon.innerText = 'block';
        if (this._dropIndicator) {
            this.renderer.removeClass(this._dropIndicator, this._dropIndicatorClass);
        }
        if (this.horizontalScroll) {
            this._dragLeave.next(true);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxColumnMovingDropDirective.prototype.onDragDrop = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ drag = event.detail.owner;
        if (!(drag instanceof IgxColumnMovingDragDirective)) {
            return;
        }
        if (this.horizontalScroll) {
            this._dragLeave.next(true);
        }
        if (this.isDropTarget) {
            var /** @type {?} */ args = {
                source: this.cms.column,
                target: this.column
            };
            var /** @type {?} */ nextPinnedWidth = void 0;
            if (this.column.pinned && !this.cms.column.pinned) {
                nextPinnedWidth = this.column.grid.getPinnedWidth(true) + parseFloat(this.cms.column.width);
            }
            if ((nextPinnedWidth && nextPinnedWidth > this.column.grid.calcPinnedContainerMaxWidth) ||
                this.column.level !== this.cms.column.level ||
                this.column.parent !== this.cms.column.parent ||
                this.cms.cancelDrop) {
                this.cms.cancelDrop = false;
                this.column.grid.onColumnMovingEnd.emit(args);
                return;
            }
            this.column.grid.moveColumn(this.cms.column, this.column, this._dropPos);
            if (this.cms.selection && this.cms.selection.column) {
                var /** @type {?} */ colID = this.column.grid.columnList.toArray().indexOf(this.cms.selection.column);
                this.column.grid.selection.set(this.column.gridID + '-cell', new Set([{
                        rowID: this.cms.selection.rowID,
                        columnID: colID
                    }]));
                var /** @type {?} */ cell = this.column.grid.getCellByKey(this.cms.selection.rowID, this.cms.selection.column.field);
                if (cell) {
                    cell.nativeElement.focus();
                }
                this.cms.selection = null;
            }
            this.column.grid.draggedColumn = null;
            this.column.grid.cdr.detectChanges();
        }
    };
    IgxColumnMovingDropDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxColumnMovingDrop]'
                },] },
    ];
    /** @nocollapse */
    IgxColumnMovingDropDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: NgZone, },
        { type: IgxColumnMovingService, },
    ]; };
    IgxColumnMovingDropDirective.propDecorators = {
        "data": [{ type: Input, args: ['igxColumnMovingDrop',] },],
    };
    return IgxColumnMovingDropDirective;
}(IgxDropDirective));
export { IgxColumnMovingDropDirective };
function IgxColumnMovingDropDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnMovingDropDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnMovingDropDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxColumnMovingDropDirective.propDecorators;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._dropPos;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._dropIndicator;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._lastDropIndicator;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._column;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._hVirtDir;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._dragLeave;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype._dropIndicatorClass;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype.elementRef;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype.renderer;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype.zone;
    /** @type {?} */
    IgxColumnMovingDropDirective.prototype.cms;
}
/**
 * @hidden
 */
var IgxDatePipeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(IgxDatePipeComponent, _super);
    function IgxDatePipeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    IgxDatePipeComponent.prototype.transform = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value && value instanceof Date) {
            return _super.prototype.transform.call(this, value);
        }
        else {
            return value;
        }
    };
    IgxDatePipeComponent.decorators = [
        { type: Pipe, args: [{
                    name: 'igxdate'
                },] },
    ];
    return IgxDatePipeComponent;
}(DatePipe));
export { IgxDatePipeComponent };
function IgxDatePipeComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxDatePipeComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxDatePipeComponent.ctorParameters;
}
/**
 * @hidden
 */
var IgxDecimalPipeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(IgxDecimalPipeComponent, _super);
    function IgxDecimalPipeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    IgxDecimalPipeComponent.prototype.transform = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value && typeof value === 'number') {
            return _super.prototype.transform.call(this, value);
        }
        else {
            return value;
        }
    };
    IgxDecimalPipeComponent.decorators = [
        { type: Pipe, args: [{
                    name: 'igxdecimal'
                },] },
    ];
    return IgxDecimalPipeComponent;
}(DecimalPipe));
export { IgxDecimalPipeComponent };
function IgxDecimalPipeComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxDecimalPipeComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxDecimalPipeComponent.ctorParameters;
}
/**
 * @record
 */
export function ContainerPositionSettings() { }
function ContainerPositionSettings_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ContainerPositionSettings.prototype.container;
}
var ContainerPositioningStrategy = /** @class */ (function (_super) {
    tslib_1.__extends(ContainerPositioningStrategy, _super);
    function ContainerPositioningStrategy() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isTop = false;
        _this.isTopInitialPosition = null;
        return _this;
    }
    /**
     * @param {?} contentElement
     * @param {?} size
     * @param {?=} document
     * @param {?=} initialCall
     * @return {?}
     */
    ContainerPositioningStrategy.prototype.position = /**
     * @param {?} contentElement
     * @param {?} size
     * @param {?=} document
     * @param {?=} initialCall
     * @return {?}
     */
    function (contentElement, size, document, initialCall) {
        var /** @type {?} */ container = this.settings.container; // grid.tbody
        var /** @type {?} */ target = /** @type {?} */ (this.settings.target); // current grid.row
        _super.prototype.position.call(this, contentElement, { width: target.clientWidth, height: target.clientHeight }, document, initialCall);
        // Position of the overlay depends on the available space in the grid.
        // If the bottom space is not enough then the the row overlay will show at the top of the row.
        // Once shown, either top or bottom, then this position stays until the overlay is closed (isTopInitialPosition property),
        // which means that when scrolling then overlay may hide, while the row is still visible (UX requirement).
        this.isTop = this.isTopInitialPosition !== null ?
            this.isTopInitialPosition :
            container.clientHeight <
                target.offsetTop + target.getBoundingClientRect().height + contentElement.getBoundingClientRect().height;
        this.settings.verticalStartPoint = this.isTop ? VerticalAlignment.Top : VerticalAlignment.Bottom;
        var /** @type {?} */ startPoint = getPointFromPositionsSettings(this.settings, contentElement.parentElement);
        contentElement.style.top = startPoint.y + (this.isTop ? VerticalAlignment.Top : VerticalAlignment.Bottom) * size.height + 'px';
        contentElement.style.width = target.clientWidth + 'px';
    };
    return ContainerPositioningStrategy;
}(ConnectedPositioningStrategy));
export { ContainerPositioningStrategy };
function ContainerPositioningStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    ContainerPositioningStrategy.prototype.isTop;
    /** @type {?} */
    ContainerPositioningStrategy.prototype.isTopInitialPosition;
    /** @type {?} */
    ContainerPositioningStrategy.prototype.settings;
}
