// igniteui-angular/esm5/lib/grids/grid/grid.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Renderer2, NgZone, HostBinding, TemplateRef } from '@angular/core';
import { IgxDropDirective } from '../../directives/dragdrop/dragdrop.directive';
import { IgxColumnMovingDragDirective } from '../grid.common';
import { SortingDirection } from '../../data-operations/sorting-expression.interface';
/**
 * @hidden
 */
var IgxGroupByRowTemplateDirective = /** @class */ (function () {
    function IgxGroupByRowTemplateDirective(template) {
        this.template = template;
    }
    IgxGroupByRowTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxGroupByRow]'
                },] },
    ];
    /** @nocollapse */
    IgxGroupByRowTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxGroupByRowTemplateDirective;
}());
export { IgxGroupByRowTemplateDirective };
function IgxGroupByRowTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxGroupByRowTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxGroupByRowTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxGroupByRowTemplateDirective.prototype.template;
}
/**
 * @hidden
 */
var IgxGroupAreaDropDirective = /** @class */ (function (_super) {
    tslib_1.__extends(IgxGroupAreaDropDirective, _super);
    function IgxGroupAreaDropDirective(elementRef, renderer, zone) {
        var _this = _super.call(this, elementRef, renderer, zone) || this;
        _this.elementRef = elementRef;
        _this.renderer = renderer;
        _this.zone = zone;
        _this.hovered = false;
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    IgxGroupAreaDropDirective.prototype.onDragEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var /** @type {?} */ drag = event.detail.owner;
        var /** @type {?} */ column = drag.column;
        var /** @type {?} */ grid = /** @type {?} */ (column.grid);
        var /** @type {?} */ isGrouped = grid.groupingExpressions.findIndex(function (item) { return item.fieldName === column.field; }) !== -1;
        if (column.groupable && !isGrouped) {
            drag.icon.innerText = 'group_work';
            this.hovered = true;
        }
        else {
            drag.icon.innerText = 'block';
            this.hovered = false;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxGroupAreaDropDirective.prototype.onDragLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.detail.owner.icon.innerText = 'block';
        this.hovered = false;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    IgxGroupAreaDropDirective.prototype.onDragDrop = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var /** @type {?} */ drag = event.detail.owner;
        if (drag instanceof IgxColumnMovingDragDirective) {
            var /** @type {?} */ column_1 = drag.column;
            var /** @type {?} */ grid = /** @type {?} */ (column_1.grid);
            var /** @type {?} */ isGrouped = grid.groupingExpressions.findIndex(function (item) { return item.fieldName === column_1.field; }) !== -1;
            if (column_1.groupable && !isGrouped) {
                grid.groupBy({ fieldName: column_1.field, dir: SortingDirection.Asc, ignoreCase: column_1.sortingIgnoreCase,
                    strategy: column_1.sortStrategy });
            }
        }
    };
    IgxGroupAreaDropDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxGroupAreaDrop]'
                },] },
    ];
    /** @nocollapse */
    IgxGroupAreaDropDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
        { type: NgZone, },
    ]; };
    IgxGroupAreaDropDirective.propDecorators = {
        "hovered": [{ type: HostBinding, args: ['class.igx-drop-area--hover',] },],
    };
    return IgxGroupAreaDropDirective;
}(IgxDropDirective));
export { IgxGroupAreaDropDirective };
function IgxGroupAreaDropDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxGroupAreaDropDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxGroupAreaDropDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxGroupAreaDropDirective.propDecorators;
    /** @type {?} */
    IgxGroupAreaDropDirective.prototype.hovered;
    /** @type {?} */
    IgxGroupAreaDropDirective.prototype.elementRef;
    /** @type {?} */
    IgxGroupAreaDropDirective.prototype.renderer;
    /** @type {?} */
    IgxGroupAreaDropDirective.prototype.zone;
}
