// igniteui-angular/esm5/lib/directives/prefix/prefix.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, Directive } from '@angular/core';
/**
 * @hidden
 */
var IgxPrefixDirective = /** @class */ (function () {
    function IgxPrefixDirective() {
    }
    IgxPrefixDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'igx-prefix,[igxPrefix]'
                },] },
    ];
    return IgxPrefixDirective;
}());
export { IgxPrefixDirective };
function IgxPrefixDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxPrefixDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxPrefixDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxPrefixModule = /** @class */ (function () {
    function IgxPrefixModule() {
    }
    IgxPrefixModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IgxPrefixDirective],
                    exports: [IgxPrefixDirective]
                },] },
    ];
    return IgxPrefixModule;
}());
export { IgxPrefixModule };
function IgxPrefixModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxPrefixModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxPrefixModule.ctorParameters;
}
