// igniteui-angular/esm2015/lib/navigation-drawer/navigation-drawer.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, HostBinding, Input, TemplateRef } from '@angular/core';
export class IgxNavDrawerItemDirective {
    constructor() {
        /**
         * @hidden
         */
        this.active = false;
        /**
         * @hidden
         */
        this.isHeader = false;
        /**
         * @hidden
         */
        this.activeClass = 'igx-nav-drawer__item--active';
    }
    /**
     * @hidden
     * @return {?}
     */
    get defaultCSS() {
        return !this.active && !this.isHeader;
    }
    /**
     * @hidden
     * @return {?}
     */
    get currentCSS() {
        return this.active && !this.isHeader;
    }
    /**
     * @hidden
     * @return {?}
     */
    get headerCSS() {
        return this.isHeader;
    }
}
IgxNavDrawerItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxDrawerItem]',
                exportAs: 'igxDrawerItem'
            },] },
];
/** @nocollapse */
IgxNavDrawerItemDirective.propDecorators = {
    "active": [{ type: Input, args: ['active',] },],
    "isHeader": [{ type: Input, args: ['isHeader',] },],
    "defaultCSS": [{ type: HostBinding, args: ['class.igx-nav-drawer__item',] },],
    "currentCSS": [{ type: HostBinding, args: ['class.igx-nav-drawer__item--active',] },],
    "headerCSS": [{ type: HostBinding, args: ['class.igx-nav-drawer__item--header',] },],
};
function IgxNavDrawerItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavDrawerItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavDrawerItemDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxNavDrawerItemDirective.propDecorators;
    /**
     * @hidden
     * @type {?}
     */
    IgxNavDrawerItemDirective.prototype.active;
    /**
     * @hidden
     * @type {?}
     */
    IgxNavDrawerItemDirective.prototype.isHeader;
    /**
     * @hidden
     * @type {?}
     */
    IgxNavDrawerItemDirective.prototype.activeClass;
}
export class IgxNavDrawerTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
IgxNavDrawerTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxDrawer]'
            },] },
];
/** @nocollapse */
IgxNavDrawerTemplateDirective.ctorParameters = () => [
    { type: TemplateRef, },
];
function IgxNavDrawerTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavDrawerTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavDrawerTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxNavDrawerTemplateDirective.prototype.template;
}
export class IgxNavDrawerMiniTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
IgxNavDrawerMiniTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxDrawerMini]'
            },] },
];
/** @nocollapse */
IgxNavDrawerMiniTemplateDirective.ctorParameters = () => [
    { type: TemplateRef, },
];
function IgxNavDrawerMiniTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavDrawerMiniTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavDrawerMiniTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxNavDrawerMiniTemplateDirective.prototype.template;
}
