// igniteui-angular/esm2015/lib/services/overlay/scroll/absolute-scroll-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class AbsoluteScrollStrategy {
    /**
     * @param {?=} scrollContainer
     */
    constructor(scrollContainer) {
        this._initialized = false;
        this.onScroll = () => {
            this._overlayService.reposition(this._id);
        };
        this._scrollContainer = scrollContainer;
    }
    /**
     * @param {?} document
     * @param {?} overlayService
     * @param {?} id
     * @return {?}
     */
    initialize(document, overlayService, id) {
        if (this._initialized) {
            return;
        }
        this._overlayService = overlayService;
        this._id = id;
        this._document = document;
        this._initialized = true;
    }
    /**
     * @return {?}
     */
    attach() {
        if (this._scrollContainer) {
            this._scrollContainer.addEventListener('scroll', this.onScroll, true);
        }
        else {
            this._document.addEventListener('scroll', this.onScroll, true);
        }
    }
    /**
     * @return {?}
     */
    detach() {
        if (this._scrollContainer) {
            this._scrollContainer.removeEventListener('scroll', this.onScroll, true);
        }
        else {
            this._document.removeEventListener('scroll', this.onScroll, true);
        }
        this._initialized = false;
    }
}
function AbsoluteScrollStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    AbsoluteScrollStrategy.prototype._initialized;
    /** @type {?} */
    AbsoluteScrollStrategy.prototype._document;
    /** @type {?} */
    AbsoluteScrollStrategy.prototype._overlayService;
    /** @type {?} */
    AbsoluteScrollStrategy.prototype._id;
    /** @type {?} */
    AbsoluteScrollStrategy.prototype._scrollContainer;
    /** @type {?} */
    AbsoluteScrollStrategy.prototype.onScroll;
}
