// igniteui-angular/esm2015/lib/services/overlay/scroll/close-scroll-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class CloseScrollStrategy {
    /**
     * @param {?=} scrollContainer
     */
    constructor(scrollContainer) {
        this._initialized = false;
        this.onScroll = (ev) => {
            if (!this._sourceElement) {
                return;
            }
            this.cumulativeScrollTop += this._sourceElement.scrollTop;
            this.cumulativeScrollLeft += this._sourceElement.scrollLeft;
            if (Math.abs(this.cumulativeScrollTop - this.initialScrollTop) > this._threshold ||
                Math.abs(this.cumulativeScrollLeft - this.initialScrollLeft) > this._threshold) {
                this._document.removeEventListener('scroll', this.onScroll, true);
                this._overlayService.hide(this._id);
            }
        };
        this._scrollContainer = scrollContainer;
        this._threshold = 10;
        this.cumulativeScrollTop = 0;
        this.cumulativeScrollLeft = 0;
    }
    /**
     * @param {?} document
     * @param {?} overlayService
     * @param {?} id
     * @return {?}
     */
    initialize(document, overlayService, id) {
        if (this._initialized) {
            return;
        }
        this._overlayService = overlayService;
        this._id = id;
        this._document = document;
        this._initialized = true;
    }
    /**
     * @return {?}
     */
    attach() {
        if (this._scrollContainer) {
            this._scrollContainer.addEventListener('scroll', this.onScroll);
            this._sourceElement = this._scrollContainer;
        }
        else {
            this._document.addEventListener('scroll', this.onScroll);
            if (document.documentElement.scrollHeight > document.documentElement.clientHeight) {
                this._sourceElement = /** @type {?} */ (document.documentElement);
            }
            else if (document.body.scrollHeight > document.body.clientHeight) {
                this._sourceElement = /** @type {?} */ (document.body);
            }
        }
        if (!this._sourceElement) {
            return;
        }
        this.cumulativeScrollTop = 0;
        this.cumulativeScrollLeft = 0;
        this.initialScrollTop = this._sourceElement.scrollTop;
        this.initialScrollLeft = this._sourceElement.scrollLeft;
    }
    /**
     * @return {?}
     */
    detach() {
        // TODO: check why event listener removes only on first call and remains on each next!!!
        if (this._scrollContainer) {
            this._scrollContainer.removeEventListener('scroll', this.onScroll);
        }
        else {
            this._document.removeEventListener('scroll', this.onScroll);
        }
        this._sourceElement = null;
        this.cumulativeScrollTop = 0;
        this.cumulativeScrollLeft = 0;
        this.initialScrollTop = 0;
        this.initialScrollLeft = 0;
        this._initialized = false;
    }
}
function CloseScrollStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    CloseScrollStrategy.prototype._document;
    /** @type {?} */
    CloseScrollStrategy.prototype._overlayService;
    /** @type {?} */
    CloseScrollStrategy.prototype._id;
    /** @type {?} */
    CloseScrollStrategy.prototype.initialScrollTop;
    /** @type {?} */
    CloseScrollStrategy.prototype.initialScrollLeft;
    /** @type {?} */
    CloseScrollStrategy.prototype.cumulativeScrollTop;
    /** @type {?} */
    CloseScrollStrategy.prototype.cumulativeScrollLeft;
    /** @type {?} */
    CloseScrollStrategy.prototype._threshold;
    /** @type {?} */
    CloseScrollStrategy.prototype._initialized;
    /** @type {?} */
    CloseScrollStrategy.prototype._sourceElement;
    /** @type {?} */
    CloseScrollStrategy.prototype._scrollContainer;
    /** @type {?} */
    CloseScrollStrategy.prototype.onScroll;
}
