// igniteui-angular/esm5/lib/core/displayDensity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { InjectionToken, Input } from '@angular/core';
/** @enum {string} */
var DisplayDensity = {
    comfortable: 'comfortable',
    cosy: 'cosy',
    compact: 'compact',
};
export { DisplayDensity };
/**
 * Describes the object used to configure the DisplayDensity in Angular DI.
 * @record
 */
export function IDisplayDensity() { }
function IDisplayDensity_tsickle_Closure_declarations() {
    /** @type {?} */
    IDisplayDensity.prototype.displayDensity;
}
/**
 * Defines the DisplayDensity DI token.
 */
export var /** @type {?} */ DisplayDensityToken = new InjectionToken('DisplayDensity');
/**
 * Base class containing all logic required for implementing DisplayDensity.
 */
var DisplayDensityBase = /** @class */ (function () {
    function DisplayDensityBase(displayDensityOptions) {
        this.displayDensityOptions = displayDensityOptions;
    }
    Object.defineProperty(DisplayDensityBase.prototype, "displayDensity", {
        get: /**
         * Returns the theme of the component.
         * The default theme is `comfortable`.
         * Available options are `comfortable`, `cosy`, `compact`.
         * ```typescript
         * let componentTheme = this.component.displayDensity;
         * ```
         * @return {?}
         */
        function () {
            return this._displayDensity;
        },
        set: /**
         * Sets the theme of the component.
         * @param {?} val
         * @return {?}
         */
        function (val) {
            switch (val) {
                case 'compact':
                    this._displayDensity = "compact" /* compact */;
                    break;
                case 'cosy':
                    this._displayDensity = "cosy" /* cosy */;
                    break;
                case 'comfortable':
                default:
                    this._displayDensity = "comfortable" /* comfortable */;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     *@hidden
     */
    /**
     * @hidden
     * @return {?}
     */
    DisplayDensityBase.prototype.isCosy = /**
     * @hidden
     * @return {?}
     */
    function () {
        return this._displayDensity === "cosy" /* cosy */ ||
            (!this._displayDensity && this.displayDensityOptions && this.displayDensityOptions.displayDensity === "cosy" /* cosy */);
    };
    /**
     *@hidden
     */
    /**
     * @hidden
     * @return {?}
     */
    DisplayDensityBase.prototype.isComfortable = /**
     * @hidden
     * @return {?}
     */
    function () {
        return this._displayDensity === "comfortable" /* comfortable */ ||
            (!this._displayDensity && (!this.displayDensityOptions ||
                this.displayDensityOptions.displayDensity === "comfortable" /* comfortable */));
    };
    /**
     *@hidden
     */
    /**
     * @hidden
     * @return {?}
     */
    DisplayDensityBase.prototype.isCompact = /**
     * @hidden
     * @return {?}
     */
    function () {
        return this._displayDensity === "compact" /* compact */ ||
            (!this._displayDensity && this.displayDensityOptions && this.displayDensityOptions.displayDensity === "compact" /* compact */);
    };
    DisplayDensityBase.propDecorators = {
        "displayDensity": [{ type: Input },],
    };
    return DisplayDensityBase;
}());
export { DisplayDensityBase };
function DisplayDensityBase_tsickle_Closure_declarations() {
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    DisplayDensityBase.propDecorators;
    /** @type {?} */
    DisplayDensityBase.prototype._displayDensity;
    /** @type {?} */
    DisplayDensityBase.prototype.displayDensityOptions;
}
