// igniteui-angular/esm5/lib/directives/for-of/display.container.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectorRef, Component, HostBinding, ViewChild, ViewContainerRef } from '@angular/core';
var DisplayContainerComponent = /** @class */ (function () {
    function DisplayContainerComponent(cdr, _viewContainer) {
        this.cdr = cdr;
        this._viewContainer = _viewContainer;
        this.cssClass = 'igx-display-container';
        this.notVirtual = true;
    }
    DisplayContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'igx-display-container',
                    template: "\n        <ng-template\n            #display_container\n            igxScrollInertia\n            [IgxScrollInertiaScrollContainer]=\"scrollContainer\"\n            [IgxScrollInertiaDirection]=\"scrollDirection\">\n        </ng-template>\n    "
                },] },
    ];
    /** @nocollapse */
    DisplayContainerComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef, },
        { type: ViewContainerRef, },
    ]; };
    DisplayContainerComponent.propDecorators = {
        "_vcr": [{ type: ViewChild, args: ['display_container', { read: ViewContainerRef },] },],
        "cssClass": [{ type: HostBinding, args: ['class',] },],
        "notVirtual": [{ type: HostBinding, args: ['class.igx-display-container--inactive',] },],
    };
    return DisplayContainerComponent;
}());
export { DisplayContainerComponent };
function DisplayContainerComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DisplayContainerComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DisplayContainerComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    DisplayContainerComponent.propDecorators;
    /** @type {?} */
    DisplayContainerComponent.prototype._vcr;
    /** @type {?} */
    DisplayContainerComponent.prototype.scrollDirection;
    /** @type {?} */
    DisplayContainerComponent.prototype.cssClass;
    /** @type {?} */
    DisplayContainerComponent.prototype.notVirtual;
    /** @type {?} */
    DisplayContainerComponent.prototype.scrollContainer;
    /** @type {?} */
    DisplayContainerComponent.prototype.cdr;
    /** @type {?} */
    DisplayContainerComponent.prototype._viewContainer;
}
