// igniteui-angular/esm5/lib/directives/for-of/virtual.helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, HostBinding, Input, ViewChild, ViewContainerRef, ChangeDetectorRef } from '@angular/core';
var VirtualHelperComponent = /** @class */ (function () {
    function VirtualHelperComponent(elementRef, cdr) {
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.cssClasses = 'igx-vhelper--vertical';
    }
    /**
     * @return {?}
     */
    VirtualHelperComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroyed = true;
    };
    VirtualHelperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'igx-virtual-helper',
                    template: '<div #container class="igx-vhelper__placeholder-content" [style.height.px]="height"></div>'
                },] },
    ];
    /** @nocollapse */
    VirtualHelperComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ChangeDetectorRef, },
    ]; };
    VirtualHelperComponent.propDecorators = {
        "_vcr": [{ type: ViewChild, args: ['container', { read: ViewContainerRef },] },],
        "itemsLength": [{ type: Input },],
        "cssClasses": [{ type: HostBinding, args: ['class',] },],
    };
    return VirtualHelperComponent;
}());
export { VirtualHelperComponent };
function VirtualHelperComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    VirtualHelperComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    VirtualHelperComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    VirtualHelperComponent.propDecorators;
    /** @type {?} */
    VirtualHelperComponent.prototype._vcr;
    /** @type {?} */
    VirtualHelperComponent.prototype.itemsLength;
    /** @type {?} */
    VirtualHelperComponent.prototype.height;
    /** @type {?} */
    VirtualHelperComponent.prototype.cssClasses;
    /** @type {?} */
    VirtualHelperComponent.prototype.destroyed;
    /** @type {?} */
    VirtualHelperComponent.prototype.elementRef;
    /** @type {?} */
    VirtualHelperComponent.prototype.cdr;
}
