// igniteui-angular/esm2015/lib/services/overlay/utilities.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {number} */
const HorizontalAlignment = {
    Left: -1,
    Center: -0.5,
    Right: 0,
};
export { HorizontalAlignment };
HorizontalAlignment[HorizontalAlignment.Left] = "Left";
HorizontalAlignment[HorizontalAlignment.Center] = "Center";
HorizontalAlignment[HorizontalAlignment.Right] = "Right";
/** @enum {number} */
const VerticalAlignment = {
    Top: -1,
    Middle: -0.5,
    Bottom: 0,
};
export { VerticalAlignment };
VerticalAlignment[VerticalAlignment.Top] = "Top";
VerticalAlignment[VerticalAlignment.Middle] = "Middle";
VerticalAlignment[VerticalAlignment.Bottom] = "Bottom";
export class Point {
    /**
     * @param {?} x
     * @param {?} y
     */
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}
function Point_tsickle_Closure_declarations() {
    /** @type {?} */
    Point.prototype.x;
    /** @type {?} */
    Point.prototype.y;
}
/**
 * @record
 */
export function PositionSettings() { }
function PositionSettings_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    PositionSettings.prototype.target;
    /** @type {?|undefined} */
    PositionSettings.prototype.horizontalDirection;
    /** @type {?|undefined} */
    PositionSettings.prototype.verticalDirection;
    /** @type {?|undefined} */
    PositionSettings.prototype.horizontalStartPoint;
    /** @type {?|undefined} */
    PositionSettings.prototype.verticalStartPoint;
    /** @type {?|undefined} */
    PositionSettings.prototype.openAnimation;
    /** @type {?|undefined} */
    PositionSettings.prototype.closeAnimation;
}
/**
 * @record
 */
export function OverlaySettings() { }
function OverlaySettings_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    OverlaySettings.prototype.positionStrategy;
    /** @type {?|undefined} */
    OverlaySettings.prototype.scrollStrategy;
    /** @type {?|undefined} */
    OverlaySettings.prototype.modal;
    /** @type {?|undefined} */
    OverlaySettings.prototype.closeOnOutsideClick;
    /** @type {?|undefined} */
    OverlaySettings.prototype.outlet;
}
/**
 * @record
 */
export function OverlayEventArgs() { }
function OverlayEventArgs_tsickle_Closure_declarations() {
    /**
     * Id of the overlay as returned by the `show()` method
     * @type {?}
     */
    OverlayEventArgs.prototype.id;
    /**
     * Available when `Type<T>` is provided to the `show()` method and allows access to the created Component instance
     * @type {?|undefined}
     */
    OverlayEventArgs.prototype.componentRef;
}
/**
 * @record
 */
export function OverlayCancelableEventArgs() { }
function OverlayCancelableEventArgs_tsickle_Closure_declarations() {
}
/**
 * @record
 */
export function OverlayAnimationEventArgs() { }
function OverlayAnimationEventArgs_tsickle_Closure_declarations() {
    /**
     * Id of the overlay as returned by the `show()` method
     * @type {?}
     */
    OverlayAnimationEventArgs.prototype.id;
    /**
     * Animation player that will play the animation
     * @type {?}
     */
    OverlayAnimationEventArgs.prototype.animationPlayer;
    /**
     * Type of animation to be played. It should be either 'open' or 'close'
     * @type {?}
     */
    OverlayAnimationEventArgs.prototype.animationType;
}
/**
 * @hidden
 * @param {?} settings
 * @param {?} overlayWrapper
 * @return {?}
 */
export function getPointFromPositionsSettings(settings, overlayWrapper) {
    let /** @type {?} */ result = new Point(0, 0);
    if (settings.target instanceof HTMLElement) {
        const /** @type {?} */ rect = (/** @type {?} */ (settings.target)).getBoundingClientRect();
        result.x = rect.right + rect.width * settings.horizontalStartPoint;
        result.y = rect.bottom + rect.height * settings.verticalStartPoint;
    }
    else if (settings.target instanceof Point) {
        result = settings.target;
    }
    //  if for some reason overlayWrapper is not at 0,0 position, e.g. overlay is in outlet
    //  which is in element with transform,perspective or filter set, we should translate the result
    //  accordingly
    if (overlayWrapper) {
        const /** @type {?} */ overlayWrapperPosition = overlayWrapper.getBoundingClientRect();
        result.x -= overlayWrapperPosition.left;
        result.y -= overlayWrapperPosition.top;
    }
    return result;
}
/**
 * @hidden
 * @record
 */
export function OverlayInfo() { }
function OverlayInfo_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    OverlayInfo.prototype.id;
    /** @type {?|undefined} */
    OverlayInfo.prototype.elementRef;
    /** @type {?|undefined} */
    OverlayInfo.prototype.componentRef;
    /** @type {?|undefined} */
    OverlayInfo.prototype.settings;
    /** @type {?|undefined} */
    OverlayInfo.prototype.initialSize;
    /** @type {?|undefined} */
    OverlayInfo.prototype.hook;
    /** @type {?|undefined} */
    OverlayInfo.prototype.openAnimationPlayer;
    /** @type {?|undefined} */
    OverlayInfo.prototype.closeAnimationPlayer;
    /** @type {?|undefined} */
    OverlayInfo.prototype.openAnimationInnerPlayer;
    /** @type {?|undefined} */
    OverlayInfo.prototype.closeAnimationInnerPlayer;
}
