// igniteui-angular/esm5/lib/animations/grow/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { animate, animation, style } from '@angular/animations';
import { EaseOut, EaseIn } from '../easings';
var /** @type {?} */ base = [
    style({
        opacity: "{{ startOpacity }}",
        height: "{{ startHeight }}"
    }),
    animate("{{duration}} {{delay}} {{easing}}", style({
        opacity: "{{ endOpacity }}",
        height: "{{ endHeight }}"
    }))
];
var /** @type {?} */ baseParams = {
    delay: '0s',
    duration: '350ms',
    easing: EaseIn.quad,
    startOpacity: 0,
    endOpacity: 1,
    startHeight: '',
    endHeight: ''
};
var /** @type {?} */ growVerIn = animation(base, {
    params: tslib_1.__assign({}, baseParams, { easing: EaseOut.quad, startOpacity: 0, endOpacity: 1, startHeight: '0px', endHeight: '*' })
});
var /** @type {?} */ growVerOut = animation(base, {
    params: tslib_1.__assign({}, baseParams, { easing: EaseOut.quad, startOpacity: 1, endOpacity: 0, startHeight: '*', endHeight: '0px' })
});
export { growVerIn, growVerOut };
