// igniteui-angular/esm5/lib/animations/slide/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { animate, animation, style } from '@angular/animations';
import { EaseIn, EaseOut } from '../easings';
var /** @type {?} */ base = [
    style({
        opacity: "{{startOpacity}}",
        transform: "{{fromPosition}}"
    }),
    animate("{{duration}} {{delay}} {{easing}}", style({
        opacity: "{{endOpacity}}",
        transform: "{{toPosition}}"
    }))
];
var /** @type {?} */ baseInParams = {
    delay: '0s',
    duration: '350ms',
    easing: EaseOut.quad,
    endOpacity: 1,
    fromPosition: 'translateY(-500px)',
    startOpacity: 0,
    toPosition: 'translateY(0)'
};
var /** @type {?} */ baseOutParams = {
    delay: '0s',
    duration: '350ms',
    easing: EaseIn.quad,
    endOpacity: 0,
    fromPosition: 'translateY(0)',
    startOpacity: 1,
    toPosition: 'translateY(-500px)'
};
var /** @type {?} */ slideInTop = animation(base, { params: baseInParams });
var /** @type {?} */ slideInLeft = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseOut.quad,
        endOpacity: 1,
        fromPosition: 'translateX(-500px)',
        startOpacity: 0,
        toPosition: 'translateY(0)'
    }
});
var /** @type {?} */ slideInRight = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseOut.quad,
        endOpacity: 1,
        fromPosition: 'translateX(500px)',
        startOpacity: 0,
        toPosition: 'translateY(0)'
    }
});
var /** @type {?} */ slideInBottom = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseOut.quad,
        endOpacity: 1,
        fromPosition: 'translateY(500px)',
        startOpacity: 0,
        toPosition: 'translateY(0)'
    }
});
var /** @type {?} */ slideInTr = animation(base, {
    params: tslib_1.__assign({}, baseInParams, { fromPosition: 'translateY(-500px) translateX(500px)', toPosition: 'translateY(0) translateX(0)' })
});
var /** @type {?} */ slideInTl = animation(base, {
    params: tslib_1.__assign({}, baseInParams, { fromPosition: 'translateY(-500px) translateX(-500px)', toPosition: 'translateY(0) translateX(0)' })
});
var /** @type {?} */ slideInBr = animation(base, {
    params: tslib_1.__assign({}, baseInParams, { fromPosition: 'translateY(500px) translateX(500px)', toPosition: 'translateY(0) translateX(0)' })
});
var /** @type {?} */ slideInBl = animation(base, {
    params: tslib_1.__assign({}, baseInParams, { fromPosition: 'translateY(500px) translateX(-500px)', toPosition: 'translateY(0) translateX(0)' })
});
var /** @type {?} */ slideOutTop = animation(base, { params: baseOutParams });
var /** @type {?} */ slideOutRight = animation(base, {
    params: tslib_1.__assign({}, baseOutParams, { toPosition: 'translateX(500px)' })
});
var /** @type {?} */ slideOutBottom = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseIn.quad,
        endOpacity: 0,
        fromPosition: 'translateY(0)',
        startOpacity: 1,
        toPosition: 'translateY(500px)'
    }
});
var /** @type {?} */ slideOutLeft = animation(base, {
    params: tslib_1.__assign({}, baseOutParams, { toPosition: 'translateX(-500px)' })
});
var /** @type {?} */ slideOutTr = animation(base, {
    params: tslib_1.__assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(-500px) translateX(500px)' })
});
var /** @type {?} */ slideOutBr = animation(base, {
    params: tslib_1.__assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(500px) translateX(500px)' })
});
var /** @type {?} */ slideOutBl = animation(base, {
    params: tslib_1.__assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(500px) translateX(-500px)' })
});
var /** @type {?} */ slideOutTl = animation(base, {
    params: tslib_1.__assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(-500px) translateX(-500px)' })
});
export { slideInTop, slideInRight, slideInBottom, slideInLeft, slideInTr, slideInBr, slideInBl, slideInTl, slideOutTop, slideOutBottom, slideOutRight, slideOutLeft, slideOutTr, slideOutBr, slideOutBl, slideOutTl };
