// igniteui-angular/esm2015/lib/card/card.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Directive, HostBinding, Input, NgModule } from '@angular/core';
import { IgxButtonModule } from '../directives/button/button.directive';
let /** @type {?} */ NEXT_ID = 0;
/**
 * IgxCardHeader is container for the card header
 */
export class IgxCardHeaderDirective {
}
IgxCardHeaderDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-card-header'
            },] },
];
function IgxCardHeaderDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCardHeaderDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCardHeaderDirective.ctorParameters;
}
/**
 * IgxCardContent is container for the card content
 */
export class IgxCardContentDirective {
}
IgxCardContentDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-card-content'
            },] },
];
function IgxCardContentDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCardContentDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCardContentDirective.ctorParameters;
}
/**
 * IgxCardActions is container for the card actions
 */
export class IgxCardActionsDirective {
}
IgxCardActionsDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-card-actions'
            },] },
];
function IgxCardActionsDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCardActionsDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCardActionsDirective.ctorParameters;
}
/**
 * IgxCardFooter is container for the card footer
 */
export class IgxCardFooterDirective {
    constructor() {
        this.role = 'footer';
    }
}
IgxCardFooterDirective.decorators = [
    { type: Directive, args: [{
                /*     host: {
                            role: "footer"
                        }, */
                // tslint:disable-next-line:directive-selector
                selector: 'igx-card-footer'
            },] },
];
/** @nocollapse */
IgxCardFooterDirective.propDecorators = {
    "role": [{ type: Input },],
};
function IgxCardFooterDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCardFooterDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCardFooterDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxCardFooterDirective.propDecorators;
    /** @type {?} */
    IgxCardFooterDirective.prototype.role;
}
/**
 * **Ignite UI for Angular Card** -
 * [Documentation](https://www.infragistics.com/products/ignite-ui-angular/angular/components/card.html)
 *
 * The Ignite UI Card serves as a container that allows custom content to be organized in an appealing way.  There are
 * four sections in a card that you can use to organize your content.  These are header, footer, content & actions.
 *
 * Example:
 * ```html
 * <igx-card>
 *   <igx-card-header>
 *     <div>
 *       <h3 class="igx-card-header__title--small">{{title}}</h3>
 *       <h5 class="igx-card-header__subtitle">{{subtitle}}</h5>
 *     </div>
 *   </igx-card-header>
 *   <igx-card-actions>
 *     <div>
 *       <button igxButton igxRipple>Share</button>
 *       <button igxButton igxRipple>Play Album</button>
 *     </div>
 *   </igx-card-actions>
 * </igx-card>
 * ```
 */
export class IgxCardComponent {
    constructor() {
        /**
         * Sets/gets the `id` of the card.
         * If not set, `id` will have value `"igx-card-0"`;
         * ```html
         * <igx-card id = "my-first-card"></igx-card>
         * ```
         * ```typescript
         * let cardId =  this.card.id;
         * ```
         * \@memberof IgxCardComponent
         */
        this.id = `igx-card-${NEXT_ID++}`;
    }
}
IgxCardComponent.decorators = [
    { type: Component, args: [{
                selector: 'igx-card',
                template: `<div class="igx-card" role="group">
    <ng-content></ng-content>
</div>`
            },] },
];
/** @nocollapse */
IgxCardComponent.propDecorators = {
    "id": [{ type: HostBinding, args: ['attr.id',] }, { type: Input },],
};
function IgxCardComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCardComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCardComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxCardComponent.propDecorators;
    /**
     * Sets/gets the `id` of the card.
     * If not set, `id` will have value `"igx-card-0"`;
     * ```html
     * <igx-card id = "my-first-card"></igx-card>
     * ```
     * ```typescript
     * let cardId =  this.card.id;
     * ```
     * \@memberof IgxCardComponent
     * @type {?}
     */
    IgxCardComponent.prototype.id;
}
/**
 * The `IgxCardModule` provides the {\@link IgxCardComponent} inside your application.
 */
export class IgxCardModule {
}
IgxCardModule.decorators = [
    { type: NgModule, args: [{
                declarations: [IgxCardComponent, IgxCardHeaderDirective,
                    IgxCardContentDirective, IgxCardActionsDirective, IgxCardFooterDirective],
                exports: [IgxCardComponent, IgxCardHeaderDirective,
                    IgxCardContentDirective, IgxCardActionsDirective, IgxCardFooterDirective],
                imports: [IgxButtonModule]
            },] },
];
function IgxCardModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCardModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCardModule.ctorParameters;
}
