// igniteui-angular/esm2015/lib/directives/input/input.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, forwardRef, HostBinding, HostListener, Inject, Input, Optional, Self } from '@angular/core';
import { FormControlName, NgModel } from '@angular/forms';
import { IgxInputGroupComponent } from '../../input-group/input-group.component';
const /** @type {?} */ nativeValidationAttributes = ['required', 'pattern', 'minlength', 'maxlength', 'min', 'max', 'step'];
/** @enum {number} */
const IgxInputState = {
    INITIAL: 0,
    VALID: 1,
    INVALID: 2,
};
export { IgxInputState };
IgxInputState[IgxInputState.INITIAL] = "INITIAL";
IgxInputState[IgxInputState.VALID] = "VALID";
IgxInputState[IgxInputState.INVALID] = "INVALID";
export class IgxInputDirective {
    /**
     * @param {?} inputGroup
     * @param {?} ngModel
     * @param {?} formControl
     * @param {?} element
     * @param {?} cdr
     */
    constructor(inputGroup, ngModel, formControl, element, cdr) {
        this.inputGroup = inputGroup;
        this.ngModel = ngModel;
        this.formControl = formControl;
        this.element = element;
        this.cdr = cdr;
        this._valid = IgxInputState.INITIAL;
        /**
         * Sets/gets whether the `"igx-input-group__input"` class is added to the host element.
         * Default value is `false`.
         * ```typescript
         * this.igxInput.isInput = true;
         * ```
         * ```typescript
         * let isCLassAdded = this.igxInput.isInput;
         * ```
         * \@memberof IgxInputDirective
         */
        this.isInput = false;
        /**
         * Sets/gets whether the `"class.igx-input-group__textarea"` class is added to the host element.
         * Default value is `false`.
         * ```typescript
         * this.igxInput.isTextArea = true;
         * ```
         * ```typescript
         * let isCLassAdded = this.igxInput.isTextArea;
         * ```
         * \@memberof IgxInputDirective
         */
        this.isTextArea = false;
    }
    /**
     * @return {?}
     */
    get ngControl() {
        return this.ngModel ? this.ngModel : this.formControl;
    }
    /**
     * Sets the `value` property.
     * ```html
     * <input-group>
     *  <input igxInput #igxInput [value]="'IgxInput Value'">
     * </input-group>
     * ```
     * \@memberof IgxInputDirective
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        this.nativeElement.value = value;
    }
    /**
     * Gets the `value` propery.
     * ```typescript
     * \@ViewChild('igxInput', {read: IgxInputDirective})
     *  public igxInput: IgxInputDirective;
     * let inputValue = this.igxInput.value;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get value() {
        return this.nativeElement.value;
    }
    /**
     * Sets the `disabled` property.
     * ```html
     * <input-group>
     *  <input igxInput #igxInput [disabled]="true">
     * </input-group>
     * ```
     * \@memberof IgxInputDirective
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this.nativeElement.disabled = value;
        this.inputGroup.disabled = value;
    }
    /**
     * Gets the `disabled` property
     * ```typescript
     * \@ViewChild('igxInput', {read: IgxInputDirective})
     *  public igxInput: IgxInputDirective;
     * let isDisabled = this.igxInput.disabled;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get disabled() {
        return this.nativeElement.hasAttribute('disabled');
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onFocus(event) {
        this.inputGroup.isFocused = true;
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onBlur(event) {
        this.inputGroup.isFocused = false;
        this._valid = IgxInputState.INITIAL;
        if (this.ngControl) {
            if (!this.ngControl.valid) {
                this._valid = IgxInputState.INVALID;
            }
        }
        else if (this._hasValidators() && !this.nativeElement.checkValidity()) {
            this._valid = IgxInputState.INVALID;
        }
    }
    /**
     * @hidden
     * @return {?}
     */
    onInput() {
        if (!this.ngControl && this._hasValidators()) {
            this._valid = this.nativeElement.checkValidity() ? IgxInputState.VALID : IgxInputState.INVALID;
        }
    }
    /**
     * @hidden
     * @return {?}
     */
    ngAfterViewInit() {
        this.inputGroup.hasPlaceholder = this.nativeElement.hasAttribute('placeholder');
        this.inputGroup.disabled = this.inputGroup.disabled || this.nativeElement.hasAttribute('disabled');
        this.inputGroup.isRequired = this.nativeElement.hasAttribute('required');
        // Also check the control's validators for required
        if (!this.inputGroup.isRequired && this.ngControl && this.ngControl.control.validator) {
            const /** @type {?} */ validation = this.ngControl.control.validator(/** @type {?} */ ({}));
            this.inputGroup.isRequired = validation && validation["required"];
        }
        const /** @type {?} */ elTag = this.nativeElement.tagName.toLowerCase();
        if (elTag === 'textarea') {
            this.isTextArea = true;
        }
        else {
            this.isInput = true;
        }
        if (this.ngControl) {
            this._statusChanges$ = this.ngControl.statusChanges.subscribe(this.onStatusChanged.bind(this));
        }
        this.cdr.detectChanges();
    }
    /**
     * @hidden
     * @return {?}
     */
    ngOnDestroy() {
        if (this._statusChanges$) {
            this._statusChanges$.unsubscribe();
        }
    }
    /**
     * Sets a focus on the igxInput.
     * ```typescript
     * this.igxInput.focus();
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    focus() {
        this.nativeElement.focus();
    }
    /**
     * Gets the `nativeElement` of the igxInput.
     * ```typescript
     * let igxInputNativeElement = this.igxInput.nativeElement;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get nativeElement() {
        return this.element.nativeElement;
    }
    /**
     * @hidden
     * @return {?}
     */
    onStatusChanged() {
        if (this.ngControl.control.validator || this.ngControl.control.asyncValidator) {
            if (this.ngControl.control.touched || this.ngControl.control.dirty) {
                if (this.inputGroup.isFocused) {
                    // the user is still typing in the control
                    this._valid = this.ngControl.valid ? IgxInputState.VALID : IgxInputState.INVALID;
                }
                else {
                    // the user had touched the control previosly but now the value is changing due to changes in the form
                    this._valid = this.ngControl.valid ? IgxInputState.INITIAL : IgxInputState.INVALID;
                }
            }
            else if (this._valid !== IgxInputState.INITIAL) {
                this._valid = this.ngControl.valid ? IgxInputState.INITIAL : IgxInputState.INVALID;
            }
            else if (this._valid === IgxInputState.INITIAL && this.ngControl.value !== undefined && this.ngControl.invalid) {
                this._valid = IgxInputState.INVALID;
            }
        }
    }
    /**
     * Gets whether the igxInput is required.
     * ```typescript
     * let isRequired = this.igxInput.required;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get required() {
        return this.nativeElement.hasAttribute('required');
    }
    /**
     * Gets whether the igxInput has a placeholder.
     * ```typescript
     * let hasPlaceholder = this.igxInput.hasPlaceholder;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get hasPlaceholder() {
        return this.nativeElement.hasAttribute('placeholder');
    }
    /**
     * Gets the placeholder element of the igxInput.
     * ```typescript
     * let igxInputPlaceholder = this.igxInput.placeholder;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get placeholder() {
        return this.nativeElement.placeholder;
    }
    /**
     * @return {?}
     */
    _hasValidators() {
        for (const /** @type {?} */ nativeValidationAttribute of nativeValidationAttributes) {
            if (this.nativeElement.hasAttribute(nativeValidationAttribute)) {
                return true;
            }
        }
        return !!this.ngControl && (!!this.ngControl.control.validator || !!this.ngControl.control.asyncValidator);
    }
    /**
     * Gets whether the igxInput is focused.
     * ```typescript
     * let isFocused = this.igxInput.focused;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get focused() {
        return this.inputGroup.isFocused;
    }
    /**
     * Gets the state of the igxInput.
     * ```typescript
     * let igxInputState = this.igxInput.valid;
     * ```
     * \@memberof IgxInputDirective
     * @return {?}
     */
    get valid() {
        return this._valid;
    }
    /**
     * Sets the state of the igxInput.
     * ```typescript
     * this.igxInput.valid = IgxInputState.INVALID;
     * ```
     * \@memberof IgxInputDirective
     * @param {?} value
     * @return {?}
     */
    set valid(value) {
        this._valid = value;
    }
}
IgxInputDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxInput]'
            },] },
];
/** @nocollapse */
IgxInputDirective.ctorParameters = () => [
    { type: IgxInputGroupComponent, decorators: [{ type: Inject, args: [forwardRef(() => IgxInputGroupComponent),] },] },
    { type: NgModel, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgModel,] },] },
    { type: FormControlName, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [FormControlName,] },] },
    { type: ElementRef, },
    { type: ChangeDetectorRef, },
];
IgxInputDirective.propDecorators = {
    "value": [{ type: Input, args: ['value',] },],
    "disabled": [{ type: Input },],
    "isInput": [{ type: HostBinding, args: ['class.igx-input-group__input',] },],
    "isTextArea": [{ type: HostBinding, args: ['class.igx-input-group__textarea',] },],
    "onFocus": [{ type: HostListener, args: ['focus', ['$event'],] },],
    "onBlur": [{ type: HostListener, args: ['blur', ['$event'],] },],
    "onInput": [{ type: HostListener, args: ['input',] },],
};
function IgxInputDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxInputDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxInputDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxInputDirective.propDecorators;
    /** @type {?} */
    IgxInputDirective.prototype._valid;
    /** @type {?} */
    IgxInputDirective.prototype._statusChanges$;
    /**
     * Sets/gets whether the `"igx-input-group__input"` class is added to the host element.
     * Default value is `false`.
     * ```typescript
     * this.igxInput.isInput = true;
     * ```
     * ```typescript
     * let isCLassAdded = this.igxInput.isInput;
     * ```
     * \@memberof IgxInputDirective
     * @type {?}
     */
    IgxInputDirective.prototype.isInput;
    /**
     * Sets/gets whether the `"class.igx-input-group__textarea"` class is added to the host element.
     * Default value is `false`.
     * ```typescript
     * this.igxInput.isTextArea = true;
     * ```
     * ```typescript
     * let isCLassAdded = this.igxInput.isTextArea;
     * ```
     * \@memberof IgxInputDirective
     * @type {?}
     */
    IgxInputDirective.prototype.isTextArea;
    /** @type {?} */
    IgxInputDirective.prototype.inputGroup;
    /** @type {?} */
    IgxInputDirective.prototype.ngModel;
    /** @type {?} */
    IgxInputDirective.prototype.formControl;
    /** @type {?} */
    IgxInputDirective.prototype.element;
    /** @type {?} */
    IgxInputDirective.prototype.cdr;
}
