// igniteui-angular/esm2015/lib/grids/cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, HostBinding, HostListener, Input, TemplateRef, ViewChild } from '@angular/core';
import { IgxSelectionAPIService } from '../core/selection';
import { IgxTextHighlightDirective } from '../directives/text-highlight/text-highlight.directive';
import { GridBaseAPIService } from './api.service';
import { IgxColumnComponent } from './column.component';
import { isNavigationKey, valToPxlsUsingRange } from '../core/utils';
/**
 * Providing reference to `IgxGridCellComponent`:
 * ```typescript
 * \@ViewChild('grid', { read: IgxGridComponent })
 *  public grid: IgxGridComponent;
 * ```
 * ```typescript
 *  let column = this.grid.columnList.first;
 * ```
 * ```typescript
 *  let cell = column.cells[0];
 * ```
 */
export class IgxGridCellComponent {
    /**
     * @param {?} gridAPI
     * @param {?} selection
     * @param {?} cdr
     * @param {?} element
     */
    constructor(gridAPI, selection, cdr, element) {
        this.gridAPI = gridAPI;
        this.selection = selection;
        this.cdr = cdr;
        this.element = element;
        /**
         * Sets/gets the highlight class of the cell.
         * Default value is `"igx-highlight"`.
         * ```typescript
         * let highlightClass = this.cell.highlightClass;
         * ```
         * ```typescript
         * this.cell.highlightClass = 'igx-cell-highlight';
         * ```
         * \@memberof IgxGridCellComponent
         */
        this.highlightClass = 'igx-highlight';
        /**
         * Sets/gets the active highlight class class of the cell.
         * Default value is `"igx-highlight__active"`.
         * ```typescript
         * let activeHighlightClass = this.cell.activeHighlightClass;
         * ```
         * ```typescript
         * this.cell.activeHighlightClass = 'igx-cell-highlight_active';
         * ```
         * \@memberof IgxGridCellComponent
         */
        this.activeHighlightClass = 'igx-highlight__active';
        /**
         * Sets/get the `tabindex` property of the cell.
         * Default value is `0`.
         * ```typescript
         * this.cell.tabindex = 1;
         * ```
         * ```typescript
         * let cellTabIndex = this.cell.tabindex;
         * ```
         * \@memberof IgxGridCellComponent
         */
        this.tabindex = 0;
        /**
         * Sets/get the `role` property of the cell.
         * Default value is `"gridcell"`.
         * ```typescript
         * this.cell.role = 'grid-cell';
         * ```
         * ```typescript
         * let cellRole = this.cell.role;
         * ```
         * \@memberof IgxGridCellComponent
         */
        this.role = 'gridcell';
        this.focused = false;
        this.isSelected = false;
        this.previousCellEditMode = false;
    }
    /**
     * Gets the cell formatter.
     * ```typescript
     * let cellForamatter = this.cell.formatter;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get formatter() {
        return this.column.formatter;
    }
    /**
     * Gets the cell template context object.
     * ```typescript
     *  let context = this.cell.context();
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get context() {
        return {
            $implicit: this.value,
            cell: this
        };
    }
    /**
     * Gets the cell template.
     * ```typescript
     * let template = this.cell.template;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get template() {
        if (this.inEditMode) {
            const /** @type {?} */ inlineEditorTemplate = this.column.inlineEditorTemplate;
            return inlineEditorTemplate ? inlineEditorTemplate : this.inlineEditorTemplate;
        }
        if (this.cellTemplate) {
            return this.cellTemplate;
        }
        return this.defaultCellTemplate;
    }
    /**
     * Gets the `id` of the grid in which the cell is stored.
     * ```typescript
     * let gridId = this.cell.gridID;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get gridID() {
        return this.row.gridID;
    }
    /**
     * Gets the grid of the cell.
     * ```typescript
     * let grid = this.cell.grid;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get grid() {
        return this.gridAPI.get(this.gridID);
    }
    /**
     * Gets the `index` of the row where the cell is stored.
     * ```typescript
     * let rowIndex = this.cell.rowIndex;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get rowIndex() {
        return this.row.index;
    }
    /**
     * Gets the `index` of the cell column.
     * ```typescript
     * let columnIndex = this.cell.columnIndex;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get columnIndex() {
        return this.column.index;
    }
    /**
     * Gets the visible `index` of the in which the cell is stored.
     * ```typescript
     * let visibleColumnIndex = this.cell.visibleColumnIndex;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get visibleColumnIndex() {
        return this.column.visibleIndex;
    }
    /**
     * Gets the `index` of the unpinned column in which the cell is stored.
     * ```typescript
     * let unpinnedColumnIndex = this.cell.ununpinnedColumnIndex;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get unpinnedColumnIndex() {
        return this.grid.unpinnedColumns.filter(c => !c.columnGroup).indexOf(this.column);
    }
    /**
     * Gets the ID of the cell.
     * ```typescript
     * let cellID = this.cell.cellID;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get cellID() {
        const /** @type {?} */ primaryKey = this.grid.primaryKey;
        const /** @type {?} */ rowID = primaryKey ? this.row.rowData[primaryKey] : this.row.rowData;
        return { rowID, columnID: this.columnIndex, rowIndex: this.rowIndex };
    }
    /**
     * Returns a reference to the nativeElement of the cell.
     * ```typescript
     * let cellNativeElement = this.cell.nativeElement;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get nativeElement() {
        return this.element.nativeElement;
    }
    /**
     * Gets whether the cell is in edit mode.
     * ```typescript
     * let isCellInEditMode = this.cell.inEditMode;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get inEditMode() {
        const /** @type {?} */ editableCell = this.gridAPI.get_cell_inEditMode(this.gridID);
        if (editableCell) {
            return this.cellID.rowID === editableCell.cellID.rowID &&
                this.cellID.columnID === editableCell.cellID.columnID;
        }
        else {
            return false;
        }
    }
    /**
     * Enables/disables the edit mode of the cell
     * ```typescript
     * this.cell.inEditMode = true;
     * ```
     * \@memberof IgxGridCellComponent
     * @param {?} value
     * @return {?}
     */
    set inEditMode(value) {
        if (this.row.deleted) {
            return;
        }
        if (this.column.editable && value) {
            this.editValue = this.value;
            this.gridAPI.set_cell_inEditMode(this.gridID, this);
            if (this.highlight && this.grid.lastSearchInfo.searchText) {
                this.highlight.observe();
            }
        }
        else {
            this.gridAPI.escape_editMode(this.gridID, this.cellID);
        }
        this.cdr.detectChanges();
        this.grid.cdr.markForCheck();
    }
    /**
     * Gets whether the cell is editable.
     * ```typescript
     * let isCellReadonly = this.cell.readonly;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get readonly() {
        return !this.column.editable;
    }
    /**
     * @hidden
     * @return {?}
     */
    get minHeight() {
        return this.grid ? this.grid.rowHeight : 32;
    }
    /**
     * Returns a string containing the grid `id` and the column `field` concatenated by "_".
     * ```typescript
     * let describedBy = this.cell.describedBy;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get describedby() {
        return `${this.row.gridID}_${this.column.field}`;
    }
    /**
     * Gets the style classes of the cell.
     * ```typescript
     * let cellStyleClasses = this.cell.styleClasses.
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get styleClasses() {
        return this.resolveStyleClasses();
    }
    /**
     * Gets the width of the cell.
     * ```typescript
     * let cellWidth = this.cell.width;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get width() {
        const /** @type {?} */ hasVerticalScroll = !this.grid.verticalScrollContainer.dc.instance.notVirtual;
        const /** @type {?} */ colWidth = this.column.width;
        const /** @type {?} */ isPercentageWidth = colWidth && typeof colWidth === 'string' && colWidth.indexOf('%') !== -1;
        if (colWidth && !isPercentageWidth) {
            let /** @type {?} */ cellWidth = this.isLastUnpinned && hasVerticalScroll ?
                parseInt(colWidth, 10) - 18 + '' : colWidth;
            if (typeof cellWidth !== 'string' || cellWidth.endsWith('px') === false) {
                cellWidth += 'px';
            }
            return cellWidth;
        }
        else {
            return colWidth;
        }
    }
    /**
     * Gets whether the cell is stored in a pinned column.
     * ```typescript
     * let isPinned = this.cell.isPinned;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get isPinned() {
        return this.column.pinned;
    }
    /**
     * Gets whether the cell is stored in the last column in the pinned area.
     * ```typescript
     * let isLastPinned = this.cell.isLastPinned;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get isLastPinned() {
        const /** @type {?} */ pinnedCols = this.grid.pinnedColumns;
        return pinnedCols[pinnedCols.length - 1] === this.column;
    }
    /**
     * Gets whether the cell is stored in the last column in the unpinned area.
     * ```typescript
     * let isLastUnpinned = this.cell.isLastUnpinned;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get isLastUnpinned() {
        const /** @type {?} */ unpinnedColumns = this.grid.unpinnedColumns;
        return unpinnedColumns[unpinnedColumns.length - 1] === this.column;
    }
    /**
     * Gets whether the cell is selected.
     * ```typescript
     * let isSelected = this.cell.selected;
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    get selected() {
        return this.isSelected = this.isCellSelected();
    }
    /**
     * Selects/deselects the cell.
     * ```typescript
     * this.cell.selected = true.
     * ```
     * \@memberof IgxGridCellComponent
     * @param {?} val
     * @return {?}
     */
    set selected(val) {
        this.isSelected = val;
    }
    /**
     * @return {?}
     */
    get dirty() {
        if (this.grid.rowEditable) {
            const /** @type {?} */ rowCurrentState = this.grid.transactions.getAggregatedValue(this.row.rowID, false);
            if (rowCurrentState) {
                return rowCurrentState && rowCurrentState[this.column.field];
            }
        }
        else {
            const /** @type {?} */ rowTransaction = this.grid.transactions.getState(this.row.rowID);
            return rowTransaction && rowTransaction.value && rowTransaction.value[this.column.field];
        }
        return false;
    }
    /**
     * @param {?=} fireFocus
     * @param {?=} event
     * @return {?}
     */
    _updateCellSelectionStatus(fireFocus = true, event) {
        if (this.selected) {
            return;
        }
        this._clearCellSelection();
        this._saveCellSelection();
        const /** @type {?} */ hasFilteredResults = this.grid.filteredData ? this.grid.filteredData.length > 0 : true;
        if (hasFilteredResults) {
            if (this.column.editable && this.previousCellEditMode && hasFilteredResults) {
                this.inEditMode = true;
            }
            if (!this.inEditMode && this.gridAPI.get_edit_row_state(this.gridID)) {
                // If there is a row being edited & this cell did not enter edit mode (!editable, row.deleted)
                this.grid.endEdit(true);
            }
            this.selected = true;
            if (fireFocus) {
                this.nativeElement.focus();
            }
            this.grid.onSelection.emit({ cell: this, event });
        }
    }
    /**
     * @return {?}
     */
    _clearCellSelection() {
        const /** @type {?} */ cell = this._getLastSelectedCell();
        if (cell) {
            cell.selected = false;
            cell.focused = false;
        }
        const /** @type {?} */ editCell = this.gridAPI.get_cell_inEditMode(this.gridID);
        if (editCell) {
            if (editCell.cell.column.field === this.grid.primaryKey) {
                if (editCell.cellID.rowIndex === this.cellID.rowIndex && editCell.cellID.columnID === this.cellID.columnID) {
                    this.previousCellEditMode = false;
                }
                else {
                    this.previousCellEditMode = true;
                    this.gridAPI.submit_value(this.gridID);
                }
            }
            else {
                this.previousCellEditMode = true;
                this.gridAPI.submit_value(this.gridID);
            }
        }
        else {
            this.previousCellEditMode = false;
        }
        this._saveCellSelection(this.selection.get_empty());
    }
    /**
     * @param {?=} newSelection
     * @return {?}
     */
    _saveCellSelection(newSelection) {
        const /** @type {?} */ sel = this.selection.get(this.cellSelectionID);
        if (sel && sel.size > 0) {
            this.selection.set(this.prevCellSelectionID, sel);
        }
        if (!newSelection) {
            newSelection = this.selection.add_item(this.cellSelectionID, this.cellID);
        }
        this.selection.set(this.cellSelectionID, newSelection);
    }
    /**
     * @return {?}
     */
    _getLastSelectedCell() {
        const /** @type {?} */ cellID = this.selection.first_item(this.cellSelectionID);
        if (cellID) {
            return this.gridAPI.get_cell_by_index(this.gridID, cellID.rowIndex, cellID.columnID);
        }
    }
    /**
     * Gets whether the cell is selected.
     * ```typescript
     * let isCellSelected = thid.cell.isCellSelected();
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    isCellSelected() {
        const /** @type {?} */ selectedCellID = this.selection.first_item(this.cellSelectionID);
        if (selectedCellID) {
            return this.cellID.rowID === selectedCellID.rowID &&
                this.cellID.columnID === selectedCellID.columnID;
        }
        return false;
    }
    /**
     * @hidden
     * @return {?}
     */
    ngOnInit() {
        this.cellSelectionID = `${this.gridID}-cell`;
        this.prevCellSelectionID = `${this.gridID}-prev-cell`;
    }
    /**
     * Sets new value to the cell.
     * ```typescript
     * this.cell.update('New Value');
     * ```
     * \@memberof IgxGridCellComponent
     * @param {?} val
     * @return {?}
     */
    update(val) {
        const /** @type {?} */ rowSelector = this.cellID.rowID;
        const /** @type {?} */ editableCell = this.gridAPI.get_cell_inEditMode(this.gridID);
        const /** @type {?} */ gridEditState = this.gridAPI.create_grid_edit_args(this.gridID, rowSelector, this.cellID.columnID, val);
        this.gridAPI.update_cell(this.gridID, rowSelector, this.cellID.columnID, val, gridEditState);
        if (editableCell && editableCell.cellID.rowID === this.cellID.rowID
            && editableCell.cellID.columnID === this.cellID.columnID) {
            if (gridEditState.args.cancel) {
                return;
            }
            this.gridAPI.escape_editMode(this.gridID, editableCell.cellID);
        }
        this.cdr.markForCheck();
        this.grid.refreshSearch();
    }
    /**
     * @hidden
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.highlight && this.grid.lastSearchInfo.searchText) {
            this.highlight.highlight(this.grid.lastSearchInfo.searchText, this.grid.lastSearchInfo.caseSensitive, this.grid.lastSearchInfo.exactMatch);
            this.highlight.activateIfNecessary();
        }
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onDoubleClick(event) {
        if (this.column.editable) {
            this.inEditMode = true;
        }
        this.grid.onDoubleClick.emit({
            cell: this,
            event
        });
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        this.grid.onCellClick.emit({
            cell: this,
            event
        });
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onContextMenu(event) {
        this.grid.onContextMenu.emit({
            cell: this,
            event
        });
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onFocus(event) {
        this.focused = true;
        this.row.focused = true;
        if (!this.selected) {
            this._updateCellSelectionStatus(false, event);
        }
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    onBlur(event) {
        this.focused = false;
        this.row.focused = false;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dispatchEvent(event) {
        const /** @type {?} */ key = event.key.toLowerCase();
        if (!this.isKeySupportedInCell(key)) {
            return;
        }
        const /** @type {?} */ shift = event.shiftKey;
        const /** @type {?} */ ctrl = event.ctrlKey;
        if (key === 'tab') {
            event.preventDefault();
            event.stopPropagation();
        }
        if (this.gridAPI.get_cell_inEditMode(this.gridID)) {
            event.stopPropagation();
        }
        if (this.inEditMode && isNavigationKey(key)) {
            return;
        }
        if (isNavigationKey(key)) {
            event.preventDefault();
            event.stopPropagation();
        }
        if (event.altKey) {
            if (this.row.nativeElement.tagName.toLowerCase() === 'igx-tree-grid-row' && this.isToggleKey(key)) {
                const /** @type {?} */ collapse = (/** @type {?} */ (this.row)).expanded && (key === 'left' || key === 'arrowleft');
                const /** @type {?} */ expand = !(/** @type {?} */ (this.row)).expanded && (key === 'right' || key === 'arrowright');
                if (collapse) {
                    (/** @type {?} */ (this.gridAPI)).trigger_row_expansion_toggle(this.gridID, this.row.treeRow, !this.row.expanded, event, this.visibleColumnIndex);
                }
                else if (expand) {
                    (/** @type {?} */ (this.gridAPI)).trigger_row_expansion_toggle(this.gridID, this.row.treeRow, !this.row.expanded, event, this.visibleColumnIndex);
                }
                return;
            }
        }
        const /** @type {?} */ args = { cell: this, groupRow: null, event: event, cancel: false };
        this.grid.onFocusChange.emit(args);
        if (args.cancel) {
            return;
        }
        switch (key) {
            case 'tab':
                if (shift) {
                    this.grid.navigation.performShiftTabKey(this.row.nativeElement, this.rowIndex, this.visibleColumnIndex);
                    break;
                }
                this.grid.navigation.performTab(this.row.nativeElement, this.rowIndex, this.visibleColumnIndex);
                break;
            case 'home':
            case 'arrowleft':
            case 'left':
                if (ctrl && key === 'home') {
                    this.grid.navigation.goToFirstCell();
                    return;
                }
                if (ctrl || key === 'home') {
                    this.grid.navigation.onKeydownHome(this.rowIndex);
                    break;
                }
                this.grid.navigation.onKeydownArrowLeft(this.nativeElement, this.rowIndex, this.visibleColumnIndex);
                break;
            case 'end':
            case 'arrowright':
            case 'right':
                if (ctrl && key === 'end') {
                    this.grid.navigation.goToLastCell();
                    return;
                }
                if (ctrl || key === 'end') {
                    this.grid.navigation.onKeydownEnd(this.rowIndex);
                    break;
                }
                this.grid.navigation.onKeydownArrowRight(this.nativeElement, this.rowIndex, this.visibleColumnIndex);
                break;
            case 'arrowup':
            case 'up':
                if (ctrl) {
                    this.grid.navigation.navigateTop(this.visibleColumnIndex);
                    break;
                }
                this.grid.navigation.navigateUp(this.row.nativeElement, this.rowIndex, this.visibleColumnIndex);
                break;
            case 'arrowdown':
            case 'down':
                if (ctrl) {
                    this.grid.navigation.navigateBottom(this.visibleColumnIndex);
                    break;
                }
                this.grid.navigation.navigateDown(this.row.nativeElement, this.rowIndex, this.visibleColumnIndex);
                break;
            case 'enter':
            case 'f2':
                this.onKeydownEnterEditMode(event);
                break;
            case 'escape':
            case 'esc':
                this.onKeydownExitEditMode(event);
                break;
            case ' ':
            case 'spacebar':
            case 'space':
                if (this.row.rowSelectable) {
                    this.row.checkboxElement.toggle();
                }
                break;
            default:
                return;
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeydownEnterEditMode(event) {
        if (this.column.editable) {
            if (this.inEditMode) {
                this.grid.endEdit(true);
                this.nativeElement.focus();
            }
            else {
                this.inEditMode = true;
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeydownExitEditMode(event) {
        if (this.column.editable) {
            const /** @type {?} */ editableCell = this;
            const /** @type {?} */ args = {
                cellID: editableCell.cellID,
                rowID: editableCell.cellID.rowID,
                oldValue: editableCell.value,
                newValue: editableCell.editValue,
                cancel: false
            };
            this.grid.onCellEditCancel.emit(args);
            if (args.cancel) {
                return;
            }
            this.grid.endEdit(false);
            this.inEditMode = false;
            this.nativeElement.focus();
        }
    }
    /**
     * If the provided string matches the text in the cell, the text gets highlighted.
     * ```typescript
     * this.cell.highlightText('Cell Value', true);
     * ```
     * \@memberof IgxGridCellComponent
     * @param {?} text
     * @param {?=} caseSensitive
     * @param {?=} exactMatch
     * @return {?}
     */
    highlightText(text, caseSensitive, exactMatch) {
        return this.highlight && this.column.searchable ? this.highlight.highlight(text, caseSensitive, exactMatch) : 0;
    }
    /**
     * Clears the highlight of the text in the cell.
     * ```typescript
     * this.cell.clearHighLight();
     * ```
     * \@memberof IgxGridCellComponent
     * @return {?}
     */
    clearHighlight() {
        if (this.highlight && this.column.searchable) {
            this.highlight.clearHighlight();
        }
    }
    /**
     * @param {?} key
     * @return {?}
     */
    isKeySupportedInCell(key) {
        return isNavigationKey(key) || key === 'tab' || key === 'enter' || key === 'f2' || key === 'escape' || key === 'esc';
    }
    /**
     * @hidden
     * @return {?}
     */
    resolveStyleClasses() {
        const /** @type {?} */ defaultClasses = ['igx-grid__td igx-grid__td--fw'];
        if (this.column.cellClasses) {
            Object.entries(this.column.cellClasses).forEach(([name, cb]) => {
                const /** @type {?} */ value = typeof cb === 'function' ? (/** @type {?} */ (cb))(this.row.rowData, this.column.field) : cb;
                if (value) {
                    defaultClasses.push(name);
                }
            }, this);
        }
        const /** @type {?} */ classList = {
            'igx_grid__cell--edit': this.inEditMode,
            'igx-grid__td--number': this.gridAPI.should_apply_number_style(this.column),
            'igx-grid__td--editing': this.inEditMode,
            'igx-grid__th--pinned': this.column.pinned,
            'igx-grid__th--pinned-last': this.isLastPinned,
            'igx-grid__td--selected': this.selected,
            'igx-grid__td--edited': this.dirty
        };
        Object.entries(classList).forEach(([klass, value]) => {
            if (value) {
                defaultClasses.push(klass);
            }
        });
        return defaultClasses.join(' ');
    }
    /**
     * @hidden
     * @param {?} range
     * @return {?}
     */
    calculateSizeToFit(range) {
        return Math.max(...Array.from(this.nativeElement.children)
            .map((child) => valToPxlsUsingRange(range, child)));
    }
    /**
     * @param {?} key
     * @return {?}
     */
    isToggleKey(key) {
        return ['left', 'right', 'arrowleft', 'arrowright'].indexOf(key.toLowerCase()) !== -1;
    }
}
IgxGridCellComponent.decorators = [
    { type: Component, args: [{
                changeDetection: ChangeDetectionStrategy.Default,
                preserveWhitespaces: false,
                selector: 'igx-grid-cell',
                template: `<ng-template #defaultCell igxTextHighlight [cssClass]="highlightClass" [activeCssClass]="activeHighlightClass" [groupName]="gridID"
        [value]="formatter ? formatter(value) : value" [row]="rowIndex" [column]="this.column.visibleIndex" [page]="this.grid.page" [containerClass]="'igx-grid__td-text'">
    <ng-container *ngIf="column.dataType === 'boolean' || column.dataType === 'string' || formatter; else default" >
        <div class="igx-grid__td-text">{{ formatter ? formatter(value) : value }}</div>
    </ng-container>
    <ng-template #default>
        <div class="igx-grid__td-text">{{ column.dataType === 'number' ? (value | igxdecimal) : (value | igxdate) }}</div>
    </ng-template>
</ng-template>
<ng-template #inlineEditor  let-cell="cell">
    <ng-container *ngIf="column.dataType === 'string'">
        <igx-input-group displayDensity="compact">
            <input igxInput [(ngModel)]="gridAPI.get_cell_inEditMode(gridID).cell.editValue" [igxFocus]="focused">
        </igx-input-group>
    </ng-container>
    <ng-container *ngIf="column.dataType === 'number'">
        <igx-input-group displayDensity="compact">
            <input igxInput [(ngModel)]="gridAPI.get_cell_inEditMode(gridID).cell.editValue" [igxFocus]="focused" type="number">
        </igx-input-group>
    </ng-container>
    <ng-container *ngIf="column.dataType === 'boolean'">
        <igx-checkbox [(ngModel)]="gridAPI.get_cell_inEditMode(gridID).cell.editValue" [checked]="gridAPI.get_cell_inEditMode(gridID).cell.editValue" [igxFocus]="focused" [disableRipple]="true"></igx-checkbox>
    </ng-container>
    <ng-container *ngIf="column.dataType === 'date'">
        <igx-datePicker [(ngModel)]="gridAPI.get_cell_inEditMode(gridID).cell.editValue" [igxFocus]="focused" [labelVisibility]="false"></igx-datePicker>
    </ng-container>
</ng-template>
<ng-container *ngTemplateOutlet="template; context: context">
</ng-container>
`
            },] },
];
/** @nocollapse */
IgxGridCellComponent.ctorParameters = () => [
    { type: GridBaseAPIService, },
    { type: IgxSelectionAPIService, },
    { type: ChangeDetectorRef, },
    { type: ElementRef, },
];
IgxGridCellComponent.propDecorators = {
    "column": [{ type: Input },],
    "row": [{ type: Input },],
    "cellTemplate": [{ type: Input },],
    "value": [{ type: Input },],
    "rowIndex": [{ type: HostBinding, args: ['attr.data-rowIndex',] },],
    "visibleColumnIndex": [{ type: HostBinding, args: ['attr.data-visibleIndex',] },],
    "tabindex": [{ type: HostBinding, args: ['attr.tabindex',] },],
    "role": [{ type: HostBinding, args: ['attr.role',] },],
    "readonly": [{ type: HostBinding, args: ['attr.aria-readonly',] },],
    "minHeight": [{ type: HostBinding, args: ['style.min-height.px',] },],
    "describedby": [{ type: HostBinding, args: ['attr.aria-describedby',] },],
    "styleClasses": [{ type: HostBinding, args: ['class',] },],
    "width": [{ type: HostBinding, args: ['style.min-width',] }, { type: HostBinding, args: ['style.flex-basis',] },],
    "selected": [{ type: HostBinding, args: ['attr.aria-selected',] },],
    "defaultCellTemplate": [{ type: ViewChild, args: ['defaultCell', { read: TemplateRef },] },],
    "inlineEditorTemplate": [{ type: ViewChild, args: ['inlineEditor', { read: TemplateRef },] },],
    "highlight": [{ type: ViewChild, args: [IgxTextHighlightDirective, { read: IgxTextHighlightDirective },] },],
    "onDoubleClick": [{ type: HostListener, args: ['dblclick', ['$event'],] },],
    "onClick": [{ type: HostListener, args: ['click', ['$event'],] },],
    "onContextMenu": [{ type: HostListener, args: ['contextmenu', ['$event'],] },],
    "onFocus": [{ type: HostListener, args: ['focus', ['$event'],] },],
    "onBlur": [{ type: HostListener, args: ['blur', ['$event'],] },],
    "dispatchEvent": [{ type: HostListener, args: ['keydown', ['$event'],] },],
};
function IgxGridCellComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxGridCellComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxGridCellComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxGridCellComponent.propDecorators;
    /**
     * Gets the column of the cell.
     * ```typescript
     *  let cellColumn = this.cell.column;
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.column;
    /**
     * Gets the row of the cell.
     * ```typescript
     * let cellRow = this.cell.row;
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.row;
    /**
     * Sets/gets the template of the cell.
     * ```html
     * <ng-template #cellTemplate igxCell let-value>
     *   <div style="font-style: oblique; color:blueviolet; background:red">
     *       <span>{{value}}</span>
     *   </div>
     * </ng-template>
     * ```
     * ```typescript
     * \@ViewChild('cellTemplate',{read: TemplateRef})
     * cellTemplate: TemplateRef<any>;
     * ```
     * ```typescript
     * this.cell.cellTemplate = this.cellTemplate;
     * ```
     * ```typescript
     * let template =  this.cell.cellTemplate;
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.cellTemplate;
    /**
     * Sets/gets the cell value.
     * ```typescript
     * this.cell.value = "Cell Value";
     * ```
     * ```typescript
     * let cellValue = this.cell.value;
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.value;
    /**
     * Sets/gets the highlight class of the cell.
     * Default value is `"igx-highlight"`.
     * ```typescript
     * let highlightClass = this.cell.highlightClass;
     * ```
     * ```typescript
     * this.cell.highlightClass = 'igx-cell-highlight';
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.highlightClass;
    /**
     * Sets/gets the active highlight class class of the cell.
     * Default value is `"igx-highlight__active"`.
     * ```typescript
     * let activeHighlightClass = this.cell.activeHighlightClass;
     * ```
     * ```typescript
     * this.cell.activeHighlightClass = 'igx-cell-highlight_active';
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.activeHighlightClass;
    /**
     * Sets/get the `tabindex` property of the cell.
     * Default value is `0`.
     * ```typescript
     * this.cell.tabindex = 1;
     * ```
     * ```typescript
     * let cellTabIndex = this.cell.tabindex;
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.tabindex;
    /**
     * Sets/get the `role` property of the cell.
     * Default value is `"gridcell"`.
     * ```typescript
     * this.cell.role = 'grid-cell';
     * ```
     * ```typescript
     * let cellRole = this.cell.role;
     * ```
     * \@memberof IgxGridCellComponent
     * @type {?}
     */
    IgxGridCellComponent.prototype.role;
    /** @type {?} */
    IgxGridCellComponent.prototype.defaultCellTemplate;
    /** @type {?} */
    IgxGridCellComponent.prototype.inlineEditorTemplate;
    /** @type {?} */
    IgxGridCellComponent.prototype.highlight;
    /**
     * @hidden
     * @type {?}
     */
    IgxGridCellComponent.prototype.editValue;
    /** @type {?} */
    IgxGridCellComponent.prototype.focused;
    /** @type {?} */
    IgxGridCellComponent.prototype.isSelected;
    /** @type {?} */
    IgxGridCellComponent.prototype.cellSelectionID;
    /** @type {?} */
    IgxGridCellComponent.prototype.prevCellSelectionID;
    /** @type {?} */
    IgxGridCellComponent.prototype.previousCellEditMode;
    /** @type {?} */
    IgxGridCellComponent.prototype.gridAPI;
    /** @type {?} */
    IgxGridCellComponent.prototype.selection;
    /** @type {?} */
    IgxGridCellComponent.prototype.cdr;
    /** @type {?} */
    IgxGridCellComponent.prototype.element;
}
