// igniteui-angular/esm2015/lib/grids/column-pinning.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { Directive, Component, NgModule, ChangeDetectorRef } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { IgxCheckboxModule } from '../checkbox/checkbox.component';
import { ColumnChooserBase } from './column-chooser-base';
import { ColumnChooserItemBase } from './column-chooser-item-base';
import { IgxInputGroupModule } from '../input-group/input-group.component';
export class IgxColumnPinningItemDirective extends ColumnChooserItemBase {
    constructor() {
        super('pinned');
    }
    /**
     * Returns whether a column could be pinned.
     * It's not possible to pin a column if there is not enough space for the unpinned area.
     * ```typescript
     * const columnItem: IgxColumnPinningItemDirective;
     * this.columnItem.pinnable;
     * ```
     * @return {?}
     */
    get pinnable() {
        if (this.column.grid.getUnpinnedWidth(true) - this.column.width < this.column.grid.unpinnedAreaMinWidth) {
            return false;
        }
        return true;
    }
}
IgxColumnPinningItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxColumnPinningItem]'
            },] },
];
/** @nocollapse */
IgxColumnPinningItemDirective.ctorParameters = () => [];
function IgxColumnPinningItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnPinningItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnPinningItemDirective.ctorParameters;
}
export class IgxColumnPinningComponent extends ColumnChooserBase {
    /**
     * @param {?} cdr
     */
    constructor(cdr) {
        super(cdr);
        this.cdr = cdr;
    }
    /**
     * @hidden
     * @param {?} container
     * @param {?} column
     * @return {?}
     */
    createColumnItem(container, column) {
        if (column.level !== 0) {
            return null;
        }
        const /** @type {?} */ item = new IgxColumnPinningItemDirective();
        item.container = container;
        item.column = column;
        return item;
    }
    /**
     * @hidden
     * @param {?} event
     * @param {?} columnItem
     * @return {?}
     */
    checkboxValueChange(event, columnItem) {
        if (event.checked && !columnItem.pinnable) {
            event.checkbox.checked = false;
            return false;
        }
        columnItem.value = !columnItem.value;
    }
}
IgxColumnPinningComponent.decorators = [
    { type: Component, args: [{
                preserveWhitespaces: false,
                selector: 'igx-column-pinning',
                template: `
<div class="igx-column-hiding__header">
    <h4 class="igx-column-hiding__header-title" *ngIf="title">{{ title }}</h4>

    <igx-input-group class="igx-column-hiding__header-input">
        <input igxInput
            type="text"
            [(ngModel)]="filterCriteria"
            [placeholder]="filterColumnsPrompt"
            autocomplete="off" />
    </igx-input-group>
</div>

<div class="igx-column-hiding__columns"
    [style.max-height]="columnsAreaMaxHeight">
    <igx-checkbox igxColumnPinningItem
        *ngFor="let columnItem of columnItems"
        class="igx-column-hiding__columns-item"
        (change)="checkboxValueChange($event, columnItem)"
        [checked]="columnItem.value">
        {{ columnItem.name }}
    </igx-checkbox>
</div>
`
            },] },
];
/** @nocollapse */
IgxColumnPinningComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, },
];
function IgxColumnPinningComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnPinningComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnPinningComponent.ctorParameters;
    /** @type {?} */
    IgxColumnPinningComponent.prototype.cdr;
}
/**
 * The IgxColumnPinningModule provides the {\@link IgxColumnPinningComponent}, {\@link IgxColumnPinningItemDirective} inside your application.
 */
export class IgxColumnPinningModule {
}
IgxColumnPinningModule.decorators = [
    { type: NgModule, args: [{
                declarations: [IgxColumnPinningComponent, IgxColumnPinningItemDirective],
                exports: [IgxColumnPinningComponent],
                imports: [
                    IgxCheckboxModule,
                    IgxInputGroupModule,
                    CommonModule,
                    FormsModule
                ]
            },] },
];
function IgxColumnPinningModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxColumnPinningModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxColumnPinningModule.ctorParameters;
}
