// igniteui-angular/esm2015/lib/animations/swing/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { animate, animation, style } from '@angular/animations';
import { EaseIn, EaseOut } from '../easings';
const /** @type {?} */ swingBase = [
    style({
        opacity: `{{startOpacity}}`,
        transform: `rotate{{direction}}({{startAngle}}deg)`,
        transformOrigin: `{{xPos}} {{yPos}}`
    }),
    animate(`{{duration}} {{delay}} {{easing}}`, style({
        opacity: `{{endOpacity}}`,
        transform: `rotate{{direction}}({{endAngle}}deg)`,
        transformOrigin: `{{xPos}} {{yPos}}`
    }))
];
const /** @type {?} */ swingParams = {
    delay: '0s',
    direction: 'X',
    duration: '.5s',
    easing: EaseOut.back,
    endAngle: 0,
    endOpacity: 1,
    startAngle: -100,
    startOpacity: 0,
    xPos: 'top',
    yPos: 'center'
};
const /** @type {?} */ swingOutParams = Object.assign({}, swingParams, { duration: '.55s', easing: EaseIn.back, endAngle: 70, endOpacity: 0, startAngle: 0, startOpacity: 1 });
const /** @type {?} */ swingInTopFwd = animation(swingBase, {
    params: Object.assign({}, swingParams)
});
const /** @type {?} */ swingInRightFwd = animation(swingBase, {
    params: Object.assign({}, swingParams, { direction: 'Y', xPos: 'center', yPos: 'right' })
});
const /** @type {?} */ swingInBottomFwd = animation(swingBase, {
    params: Object.assign({}, swingParams, { startAngle: 100, xPos: 'bottom' })
});
const /** @type {?} */ swingInLeftFwd = animation(swingBase, {
    params: Object.assign({}, swingParams, { direction: 'Y', startAngle: 100, xPos: 'center', yPos: 'left' })
});
const /** @type {?} */ swingInTopBck = animation(swingBase, {
    params: Object.assign({}, swingParams, { duration: '.6s', startAngle: 70 })
});
const /** @type {?} */ swingInRightBck = animation(swingBase, {
    params: Object.assign({}, swingParams, { direction: 'Y', duration: '.6s', startAngle: 70, xPos: 'center', yPos: 'right' })
});
const /** @type {?} */ swingInBottomBck = animation(swingBase, {
    params: Object.assign({}, swingParams, { duration: '.6s', startAngle: -70, xPos: 'bottom' })
});
const /** @type {?} */ swingInLeftBck = animation(swingBase, {
    params: Object.assign({}, swingParams, { direction: 'Y', duration: '.6s', startAngle: -70, xPos: 'center', yPos: 'left' })
});
const /** @type {?} */ swingOutTopFwd = animation(swingBase, {
    params: Object.assign({}, swingOutParams)
});
const /** @type {?} */ swingOutRightFwd = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { direction: 'Y', xPos: 'center', yPos: 'right' })
});
const /** @type {?} */ swingOutBottomFwd = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { endAngle: -70, xPos: 'bottom' })
});
const /** @type {?} */ swingOutLefttFwd = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { direction: 'Y', endAngle: -70, xPos: 'center', yPos: 'left' })
});
const /** @type {?} */ swingOutTopBck = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { duration: '.45s', endAngle: -100 })
});
const /** @type {?} */ swingOutRightBck = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { direction: 'Y', duration: '.45s', endAngle: -100, xPos: 'center', yPos: 'right' })
});
const /** @type {?} */ swingOutBottomBck = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { duration: '.45s', endAngle: 100, xPos: 'bottom' })
});
const /** @type {?} */ swingOutLeftBck = animation(swingBase, {
    params: Object.assign({}, swingOutParams, { direction: 'Y', duration: '.45s', endAngle: 100, xPos: 'center', yPos: 'left' })
});
export { swingInTopFwd, swingInRightFwd, swingInLeftFwd, swingInBottomFwd, swingInTopBck, swingInRightBck, swingInBottomBck, swingInLeftBck, swingOutTopFwd, swingOutRightFwd, swingOutBottomFwd, swingOutLefttFwd, swingOutTopBck, swingOutRightBck, swingOutBottomBck, swingOutLeftBck };
