// igniteui-angular/esm5/lib/directives/suffix/suffix.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, Directive } from '@angular/core';
/**
 * @hidden
 */
var IgxSuffixDirective = /** @class */ (function () {
    function IgxSuffixDirective() {
    }
    IgxSuffixDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'igx-suffix,[igxSuffix]'
                },] },
    ];
    return IgxSuffixDirective;
}());
export { IgxSuffixDirective };
function IgxSuffixDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxSuffixDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxSuffixDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxSuffixModule = /** @class */ (function () {
    function IgxSuffixModule() {
    }
    IgxSuffixModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IgxSuffixDirective],
                    exports: [IgxSuffixDirective]
                },] },
    ];
    return IgxSuffixModule;
}());
export { IgxSuffixModule };
function IgxSuffixModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxSuffixModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxSuffixModule.ctorParameters;
}
