// igniteui-angular/esm2015/lib/combo/combo.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive } from '@angular/core';
/**
 * @hidden
 */
export class IgxComboHeaderDirective {
}
IgxComboHeaderDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxComboHeader]'
            },] },
];
function IgxComboHeaderDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboHeaderDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboHeaderDirective.ctorParameters;
}
/**
 * @hidden
 */
export class IgxComboFooterDirective {
}
IgxComboFooterDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxComboFooter]'
            },] },
];
function IgxComboFooterDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboFooterDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboFooterDirective.ctorParameters;
}
/**
 * @hidden
 */
export class IgxComboItemDirective {
}
IgxComboItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxComboItem]'
            },] },
];
function IgxComboItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboItemDirective.ctorParameters;
}
/**
 * @hidden
 */
export class IgxComboEmptyDirective {
}
IgxComboEmptyDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxComboEmpty]'
            },] },
];
function IgxComboEmptyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboEmptyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboEmptyDirective.ctorParameters;
}
/**
 * @hidden
 */
export class IgxComboHeaderItemDirective {
}
IgxComboHeaderItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxComboHeaderItem]'
            },] },
];
function IgxComboHeaderItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboHeaderItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboHeaderItemDirective.ctorParameters;
}
/**
 * @hidden
 */
export class IgxComboAddItemDirective {
}
IgxComboAddItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxComboAddItem]'
            },] },
];
function IgxComboAddItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboAddItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboAddItemDirective.ctorParameters;
}
