// igniteui-angular/esm2015/lib/combo/combo-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, forwardRef, HostListener, HostBinding, Inject } from '@angular/core';
import { IgxSelectionAPIService } from '../core/selection';
import { IgxDropDownItemBase } from '../drop-down/drop-down-item.component';
import { IgxComboDropDownComponent } from './combo-dropdown.component';
/**
 * @hidden
 */
export class IgxComboItemComponent extends IgxDropDownItemBase {
    /**
     * @param {?} dropDown
     * @param {?} elementRef
     * @param {?} selection
     */
    constructor(dropDown, elementRef, selection) {
        super(dropDown, elementRef);
        this.dropDown = dropDown;
        this.elementRef = elementRef;
        this.selection = selection;
    }
    /**
     * @return {?}
     */
    get combo() {
        return this.dropDown.combo;
    }
    /**
     * Gets the height of a list item
     * @return {?}
     */
    get itemHeight() {
        return this.combo.itemHeight;
    }
    /**
     * @hidden
     * @return {?}
     */
    get itemID() {
        return this.combo.isRemote ? JSON.stringify(this.value) : this.value;
    }
    /**
     * @hidden
     * @return {?}
     */
    get isSelected() {
        return this.combo.isItemSelected(this.itemID);
    }
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    clicked(event) {
        if (this.disabled || this.isHeader) {
            const /** @type {?} */ focusedItem = this.dropDown.focusedItem;
            if (focusedItem) {
                focusedItem.element.nativeElement.focus({ preventScroll: true });
            }
            return;
        }
        this.dropDown.selectItem(this, event);
    }
}
IgxComboItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'igx-combo-item',
                template: `<ng-content></ng-content>`
            },] },
];
/** @nocollapse */
IgxComboItemComponent.ctorParameters = () => [
    { type: IgxComboDropDownComponent, decorators: [{ type: Inject, args: [forwardRef(() => IgxComboDropDownComponent),] },] },
    { type: ElementRef, },
    { type: IgxSelectionAPIService, },
];
IgxComboItemComponent.propDecorators = {
    "itemHeight": [{ type: HostBinding, args: ['style.height.px',] },],
    "clicked": [{ type: HostListener, args: ['click', ['$event'],] },],
};
function IgxComboItemComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboItemComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboItemComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxComboItemComponent.propDecorators;
    /** @type {?} */
    IgxComboItemComponent.prototype.dropDown;
    /** @type {?} */
    IgxComboItemComponent.prototype.elementRef;
    /** @type {?} */
    IgxComboItemComponent.prototype.selection;
}
