// igniteui-angular/esm5/lib/animations/rotate/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { animate, animation, style } from '@angular/animations';
import { EaseIn, EaseOut } from '../easings';
var /** @type {?} */ baseRecipe = [
    style({
        opacity: "{{startOpacity}}",
        transform: "rotate3d({{rotateX}},{{rotateY}},{{rotateZ}},{{startAngle}}deg)",
        transformOrigin: "{{xPos}} {{yPos}}"
    }),
    animate("{{duration}} {{delay}} {{easing}}", style({
        offset: 0,
        opacity: "{{endOpacity}}",
        transform: "rotate3d({{rotateX}},{{rotateY}},{{rotateZ}},{{endAngle}}deg)",
        transformOrigin: "{{xPos}} {{yPos}}"
    }))
];
var /** @type {?} */ baseInParams = {
    delay: '0s',
    duration: '600ms',
    easing: EaseOut.quad,
    endAngle: 0,
    endOpacity: 1,
    rotateX: 0,
    rotateY: 0,
    rotateZ: 1,
    startAngle: -360,
    startOpacity: 0,
    xPos: 'center',
    yPos: 'center'
};
var /** @type {?} */ baseOutParams = tslib_1.__assign({}, baseInParams, { easing: EaseIn.quad, endOpacity: 0, startOpacity: 1 });
var /** @type {?} */ rotateInCenter = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams)
});
var /** @type {?} */ rotateOutCenter = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams)
});
var /** @type {?} */ rotateInTop = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'top' })
});
var /** @type {?} */ rotateOutTop = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'top' })
});
var /** @type {?} */ rotateInRight = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'right' })
});
var /** @type {?} */ rotateOutRight = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'right' })
});
var /** @type {?} */ rotateInBottom = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'bottom' })
});
var /** @type {?} */ rotateOutBottom = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'bottom' })
});
var /** @type {?} */ rotateInLeft = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'left' })
});
var /** @type {?} */ rotateOutLeft = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'left' })
});
var /** @type {?} */ rotateInTr = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'right', yPos: 'top' })
});
var /** @type {?} */ rotateOutTr = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'right', yPos: 'top' })
});
var /** @type {?} */ rotateInBr = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'right', yPos: 'bottom' })
});
var /** @type {?} */ rotateOutBr = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'right', yPos: 'bottom' })
});
var /** @type {?} */ rotateInBl = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'left', yPos: 'bottom' })
});
var /** @type {?} */ rotateOutBl = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'left', yPos: 'bottom' })
});
var /** @type {?} */ rotateInTl = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { xPos: 'left', yPos: 'top' })
});
var /** @type {?} */ rotateOutTl = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { xPos: 'left', yPos: 'top' })
});
var /** @type {?} */ rotateInDiagonal1 = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { rotateX: 1, rotateY: 1, rotateZ: 0 })
});
var /** @type {?} */ rotateOutDiagonal1 = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { rotateX: 1, rotateY: 1, rotateZ: 0 })
});
var /** @type {?} */ rotateInDiagonal2 = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { rotateX: -1, rotateY: 1, rotateZ: 0 })
});
var /** @type {?} */ rotateOutDiagonal2 = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { rotateX: -1, rotateY: 1, rotateZ: 0 })
});
var /** @type {?} */ rotateInHor = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { rotateX: 0, rotateY: 1, rotateZ: 0 })
});
var /** @type {?} */ rotateOutHor = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { rotateX: 0, rotateY: 1, rotateZ: 0 })
});
var /** @type {?} */ rotateInVer = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseInParams, { rotateX: 1, rotateY: 0, rotateZ: 0 })
});
var /** @type {?} */ rotateOutVer = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseOutParams, { rotateX: 1, rotateY: 0, rotateZ: 0 })
});
export { rotateInCenter, rotateInTop, rotateInRight, rotateInLeft, rotateInBottom, rotateInTr, rotateInBr, rotateInBl, rotateInTl, rotateInDiagonal1, rotateInDiagonal2, rotateInHor, rotateInVer, rotateOutCenter, rotateOutTop, rotateOutRight, rotateOutLeft, rotateOutBottom, rotateOutTr, rotateOutBr, rotateOutBl, rotateOutTl, rotateOutDiagonal1, rotateOutDiagonal2, rotateOutHor, rotateOutVer };
