// igniteui-angular/esm2015/lib/grids/filtering/grid-filtering-row.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectorRef, Component, Input, TemplateRef, ViewChild, ViewChildren, QueryList, ElementRef, HostBinding, HostListener } from '@angular/core';
import { Subject } from 'rxjs';
import { DataType } from '../../data-operations/data-util';
import { IgxColumnComponent } from '../column.component';
import { IgxDropDownComponent } from '../../drop-down/drop-down.component';
import { FilteringLogic } from '../../data-operations/filtering-expression.interface';
import { HorizontalAlignment, VerticalAlignment } from '../../services/overlay/utilities';
import { ConnectedPositioningStrategy } from '../../services/overlay/position/connected-positioning-strategy';
import { IgxChipsAreaComponent } from '../../chips';
import { ExpressionUI } from './grid-filtering.service';
import { IgxFilteringService } from './grid-filtering.service';
import { AbsoluteScrollStrategy } from '../../services/overlay/scroll';
/**
 * @hidden
 */
export class IgxGridFilteringRowComponent {
    /**
     * @param {?} filteringService
     * @param {?} element
     * @param {?} cdr
     */
    constructor(filteringService, element, cdr) {
        this.filteringService = filteringService;
        this.element = element;
        this.cdr = cdr;
        this._positionSettings = {
            horizontalStartPoint: HorizontalAlignment.Left,
            verticalStartPoint: VerticalAlignment.Bottom
        };
        this._conditionsOverlaySettings = {
            closeOnOutsideClick: true,
            modal: false,
            scrollStrategy: new AbsoluteScrollStrategy(),
            positionStrategy: new ConnectedPositioningStrategy(this._positionSettings)
        };
        this._operatorsOverlaySettings = {
            closeOnOutsideClick: true,
            modal: false,
            scrollStrategy: new AbsoluteScrollStrategy(),
            positionStrategy: new ConnectedPositioningStrategy(this._positionSettings)
        };
        this.chipAreaScrollOffset = 0;
        this.conditionChanged = new Subject();
        this.unaryConditionChanged = new Subject();
        this._column = null;
        this.cssClass = 'igx-grid__filtering-row';
        this.unaryConditionChanged.subscribe(() => this.unaryConditionChangedCallback());
        this.conditionChanged.subscribe(() => this.conditionChangedCallback());
    }
    /**
     * @return {?}
     */
    get column() {
        return this._column;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set column(val) {
        if (val) {
            this._column = val;
            this.expressionsList = this.filteringService.getExpressions(this._column.field);
            this.resetExpression();
            this.chipAreaScrollOffset = 0;
            this.transform(this.chipAreaScrollOffset);
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this.expression ? this.expression.searchVal : null;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        if (!val && val !== 0) {
            this.expression.searchVal = null;
        }
        else {
            this.expression.searchVal = this.transformValue(val);
            if (this.expressionsList.find(item => item.expression === this.expression) === undefined) {
                this.addExpression(true);
            }
        }
        this.filter();
    }
    /**
     * @return {?}
     */
    get locale() {
        return window.navigator.language;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this._conditionsOverlaySettings.outlet = this.column.grid.outletDirective;
        this._operatorsOverlaySettings.outlet = this.column.grid.outletDirective;
        if (this.column.dataType === DataType.Date) {
            // TODO: revise usage of cdr.detectChanges() here
            this.cdr.detectChanges();
        }
        this.input.nativeElement.focus();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.conditionChanged.unsubscribe();
        this.unaryConditionChanged.unsubscribe();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onTabKeydown(event) {
        event.stopPropagation();
        if (document.activeElement === this.closeButton.nativeElement && !event.shiftKey) {
            event.preventDefault();
        }
    }
    /**
     * @return {?}
     */
    get disabled() {
        return !(this.column.filteringExpressionsTree && this.column.filteringExpressionsTree.filteringOperands.length > 0);
    }
    /**
     * @return {?}
     */
    get template() {
        if (this.column.dataType === DataType.Date) {
            return this.defaultDateUI;
        }
        return this.defaultFilterUI;
    }
    /**
     * @return {?}
     */
    get type() {
        switch (this.column.dataType) {
            case DataType.String:
            case DataType.Boolean:
                return 'text';
            case DataType.Number:
                return 'number';
        }
    }
    /**
     * @return {?}
     */
    get conditions() {
        return this.column.filters.instance().conditionList();
    }
    /**
     * @return {?}
     */
    get isUnaryCondition() {
        if (this.expression.condition) {
            return this.expression.condition.isUnary;
        }
        else {
            return true;
        }
    }
    /**
     * @return {?}
     */
    get placeholder() {
        if (this.expression.condition && this.expression.condition.isUnary) {
            return this.filteringService.getChipLabel(this.expression);
        }
        else {
            return 'Add filter value';
        }
    }
    /**
     * Event handler for keydown on the input group's prefix.
     * @param {?} event
     * @return {?}
     */
    onPrefixKeyDown(event) {
        if ((event.key === "Enter" /* ENTER */ || event.key === " " /* SPACE */ || event.key === "Spacebar" /* SPACE_IE */) &&
            this.dropDownConditions.collapsed) {
            this._conditionsOverlaySettings.positionStrategy.settings.target = this.inputGroupPrefix.nativeElement;
            this.dropDownConditions.toggle(this._conditionsOverlaySettings);
            event.stopImmediatePropagation();
        }
        else if (event.key === "Tab" /* TAB */ && event.shiftKey) {
            event.preventDefault();
            event.stopPropagation();
        }
    }
    /**
     * Event handler for keydown on the input.
     * @param {?} event
     * @return {?}
     */
    onInputKeyDown(event) {
        if (event.key === "Enter" /* ENTER */) {
            this.chipsArea.chipsList.filter(chip => chip.selected = false);
            let /** @type {?} */ indexToDeselect = -1;
            for (let /** @type {?} */ index = 0; index < this.expressionsList.length; index++) {
                const /** @type {?} */ expression = this.expressionsList[index].expression;
                if (expression.searchVal === null && !expression.condition.isUnary) {
                    indexToDeselect = index;
                }
            }
            if (indexToDeselect !== -1) {
                this.removeExpression(indexToDeselect, this.expression);
            }
            this.resetExpression();
            this.scrollChipsWhenAddingExpression();
        }
        else if (event.key === "ArrowDown" /* DOWN_ARROW */) {
            this.input.nativeElement.blur();
            this.inputGroupPrefix.nativeElement.focus();
            this.toggleConditionsDropDown(this.inputGroupPrefix.nativeElement);
        }
        event.stopPropagation();
    }
    /**
     * Event handler for datepicker's close.
     * @return {?}
     */
    datePickerClose() {
        this.input.nativeElement.focus();
    }
    /**
     * Returns the filtering operation condition for a given value.
     * @param {?} value
     * @return {?}
     */
    getCondition(value) {
        return this.column.filters.instance().condition(value);
    }
    /**
     * Returns the icon name of the current condition.
     * @return {?}
     */
    getIconName() {
        if (this.column.dataType === DataType.Boolean && this.expression.searchVal === null) {
            return this.getCondition(this.conditions[0]).iconName;
        }
        else {
            return this.expression.condition.iconName;
        }
    }
    /**
     * Returns whether a given condition is selected in dropdown.
     * @param {?} conditionName
     * @return {?}
     */
    isConditionSelected(conditionName) {
        if (this.expression.condition) {
            return this.expression.condition.name === conditionName;
        }
        else {
            return false;
        }
    }
    /**
     * Clears the current filtering.
     * @return {?}
     */
    clearFiltering() {
        this.filteringService.clearFilter(this.column.field);
        this.resetExpression();
        this.cdr.detectChanges();
        this.chipAreaScrollOffset = 0;
        this.transform(this.chipAreaScrollOffset);
    }
    /**
     * Clears the value of the input.
     * @return {?}
     */
    clearInput() {
        this.value = null;
    }
    /**
     * Event handler for keydown on clear button.
     * @param {?} eventArgs
     * @return {?}
     */
    onClearKeyDown(eventArgs) {
        if (eventArgs.key === "Enter" /* ENTER */) {
            eventArgs.preventDefault();
            this.clearInput();
        }
    }
    /**
     * Closes the filtering edit row.
     * @return {?}
     */
    close() {
        if (this.expressionsList.length === 1 &&
            this.expressionsList[0].expression.searchVal === null &&
            this.expressionsList[0].expression.condition.isUnary === false) {
            this.filteringService.clearFilter(this.column.field);
        }
        else {
            this.expressionsList.forEach((item) => {
                if (item.expression.searchVal === null && !item.expression.condition.isUnary) {
                    this.filteringService.removeExpression(this.column.field, this.expressionsList.indexOf(item));
                }
            });
        }
        this.filteringService.isFilterRowVisible = false;
        this.filteringService.filteredColumn = null;
        this.filteringService.selectedExpression = null;
        this.cdr.detectChanges();
        this.chipAreaScrollOffset = 0;
        this.transform(this.chipAreaScrollOffset);
    }
    /**
     * Opens the conditions dropdown.
     * @param {?} target
     * @return {?}
     */
    toggleConditionsDropDown(target) {
        this._conditionsOverlaySettings.positionStrategy.settings.target = target;
        this.dropDownConditions.toggle(this._conditionsOverlaySettings);
    }
    /**
     * Opens the logic operators dropdown.
     * @param {?} eventArgs
     * @param {?} index
     * @return {?}
     */
    toggleOperatorsDropDown(eventArgs, index) {
        this._operatorsOverlaySettings.positionStrategy.settings.target = eventArgs.target.parentElement;
        this.dropDownOperators.toArray()[index].toggle(this._operatorsOverlaySettings);
    }
    /**
     * Event handler for change event in conditions dropdown.
     * @param {?} eventArgs
     * @return {?}
     */
    onConditionsChanged(eventArgs) {
        const /** @type {?} */ value = (/** @type {?} */ (eventArgs.newSelection)).value;
        this.expression.condition = this.getCondition(value);
        if (this.expression.condition.isUnary) {
            this.unaryConditionChanged.next(value);
        }
        else {
            this.conditionChanged.next(value);
        }
        if (this.input) {
            this.input.nativeElement.focus();
        }
    }
    /**
     *  Event handler for chip selected event.
     * @param {?} eventArgs
     * @param {?} expression
     * @return {?}
     */
    onChipSelected(eventArgs, expression) {
        if (eventArgs.selected) {
            if (this.chipsArea.chipsList) {
                this.chipsArea.chipsList.forEach((chip) => {
                    if (chip !== eventArgs.owner) {
                        chip.selected = false;
                    }
                });
            }
            this.expression = expression;
            if (this.input) {
                this.input.nativeElement.focus();
            }
        }
        else if (this.expression === expression) {
            this.resetExpression();
        }
    }
    /**
     * Event handler for chip keydown event.
     * @param {?} eventArgs
     * @param {?} chip
     * @return {?}
     */
    onChipKeyDown(eventArgs, chip) {
        if (eventArgs.key === "Enter" /* ENTER */) {
            eventArgs.preventDefault();
            chip.selected = !chip.selected;
        }
    }
    /**
     * Scrolls the first chip into view if the tab key is pressed on the left arrow.
     * @param {?} event
     * @return {?}
     */
    onLeftArrowKeyDown(event) {
        if (event.key === "Tab" /* TAB */) {
            this.chipAreaScrollOffset = 0;
            this.transform(this.chipAreaScrollOffset);
        }
    }
    /**
     * Event handler for chip removed event.
     * @param {?} eventArgs
     * @param {?} item
     * @return {?}
     */
    onChipRemoved(eventArgs, item) {
        const /** @type {?} */ indexToRemove = this.expressionsList.indexOf(item);
        this.removeExpression(indexToRemove, item.expression);
        this.scrollChipsOnRemove();
    }
    /**
     * Event handler for logic operator changed event.
     * @param {?} eventArgs
     * @param {?} expression
     * @return {?}
     */
    onLogicOperatorChanged(eventArgs, expression) {
        if (eventArgs.oldSelection) {
            expression.afterOperator = (/** @type {?} */ (eventArgs.newSelection)).value;
            this.expressionsList[this.expressionsList.indexOf(expression) + 1].beforeOperator = expression.afterOperator;
            this.filter();
        }
    }
    /**
     * Scrolls the chips into the chip area when left or right arrows are pressed.
     * @param {?} arrowPosition
     * @return {?}
     */
    scrollChipsOnArrowPress(arrowPosition) {
        let /** @type {?} */ count = 0;
        const /** @type {?} */ chipAraeChildren = this.chipsArea.element.nativeElement.children;
        const /** @type {?} */ containerRect = this.container.nativeElement.getBoundingClientRect();
        if (arrowPosition === 'right') {
            for (let /** @type {?} */ index = 0; index < chipAraeChildren.length; index++) {
                if (Math.ceil(chipAraeChildren[index].getBoundingClientRect().right) < Math.ceil(containerRect.right)) {
                    count++;
                }
            }
            if (count < chipAraeChildren.length) {
                this.chipAreaScrollOffset -= Math.ceil(chipAraeChildren[count].getBoundingClientRect().right) -
                    Math.ceil(containerRect.right) + 1;
                this.transform(this.chipAreaScrollOffset);
            }
        }
        if (arrowPosition === 'left') {
            for (let /** @type {?} */ index = 0; index < chipAraeChildren.length; index++) {
                if (Math.ceil(chipAraeChildren[index].getBoundingClientRect().left) < Math.ceil(containerRect.left)) {
                    count++;
                }
            }
            if (count > 0) {
                this.chipAreaScrollOffset += Math.ceil(containerRect.left) -
                    Math.ceil(chipAraeChildren[count - 1].getBoundingClientRect().left) + 1;
                this.transform(this.chipAreaScrollOffset);
            }
        }
    }
    /**
     * @return {?}
     */
    showHideArrowButtons() {
        requestAnimationFrame(() => {
            const /** @type {?} */ containerWidth = this.container.nativeElement.getBoundingClientRect().width;
            this.chipsAreaWidth = this.chipsArea.element.nativeElement.getBoundingClientRect().width;
            this.showArrows = this.chipsAreaWidth >= containerWidth;
            // TODO: revise the cdr.detectChanges() usage here
            this.cdr.detectChanges();
        });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    transformValue(value) {
        if (this.column.dataType === DataType.Number) {
            value = parseFloat(value);
        }
        else if (this.column.dataType === DataType.Boolean) {
            value = Boolean(value);
        }
        return value;
    }
    /**
     * @param {?} isSelected
     * @return {?}
     */
    addExpression(isSelected) {
        const /** @type {?} */ exprUI = new ExpressionUI();
        exprUI.expression = this.expression;
        exprUI.beforeOperator = this.expressionsList.length > 0 ? FilteringLogic.And : null;
        exprUI.isSelected = isSelected;
        this.expressionsList.push(exprUI);
        const /** @type {?} */ length = this.expressionsList.length;
        if (this.expressionsList[length - 2]) {
            this.expressionsList[length - 2].afterOperator = this.expressionsList[length - 1].beforeOperator;
        }
        this.showHideArrowButtons();
    }
    /**
     * @param {?} indexToRemove
     * @param {?} expression
     * @return {?}
     */
    removeExpression(indexToRemove, expression) {
        if (indexToRemove === 0 && this.expressionsList.length === 1) {
            this.clearFiltering();
            return;
        }
        this.filteringService.removeExpression(this.column.field, indexToRemove);
        this.filter();
        if (this.expression === expression) {
            this.resetExpression();
        }
        this.showHideArrowButtons();
    }
    /**
     * @return {?}
     */
    resetExpression() {
        this.expression = {
            fieldName: this.column.field,
            condition: null,
            searchVal: null,
            ignoreCase: this.column.filteringIgnoreCase
        };
        if (this.column.dataType !== DataType.Boolean) {
            this.expression.condition = this.getCondition(this.conditions[0]);
        }
        if (this.column.dataType === DataType.Date && this.input) {
            this.input.nativeElement.value = null;
        }
        this.showHideArrowButtons();
    }
    /**
     * @return {?}
     */
    scrollChipsWhenAddingExpression() {
        const /** @type {?} */ chipAraeChildren = this.chipsArea.element.nativeElement.children;
        if (!chipAraeChildren || chipAraeChildren.length === 0) {
            return;
        }
        const /** @type {?} */ containerRectRight = Math.ceil(this.container.nativeElement.getBoundingClientRect().right);
        const /** @type {?} */ lastChipRectRight = Math.ceil(chipAraeChildren[chipAraeChildren.length - 1].getBoundingClientRect().right);
        if (lastChipRectRight >= containerRectRight) {
            this.chipAreaScrollOffset -= lastChipRectRight - containerRectRight;
            this.transform(this.chipAreaScrollOffset);
        }
    }
    /**
     * @param {?} offset
     * @return {?}
     */
    transform(offset) {
        requestAnimationFrame(() => {
            this.chipsArea.element.nativeElement.style.transform = `translate(${offset}px)`;
        });
    }
    /**
     * @return {?}
     */
    scrollChipsOnRemove() {
        let /** @type {?} */ count = 0;
        const /** @type {?} */ chipAraeChildren = this.chipsArea.element.nativeElement.children;
        const /** @type {?} */ containerRect = this.container.nativeElement.getBoundingClientRect();
        for (let /** @type {?} */ index = 0; index < chipAraeChildren.length; index++) {
            if (Math.ceil(chipAraeChildren[index].getBoundingClientRect().left) < Math.ceil(containerRect.left)) {
                count++;
            }
        }
        if (count <= 2) {
            this.chipAreaScrollOffset = 0;
        }
        else {
            const /** @type {?} */ dif = chipAraeChildren[count].id === 'chip' ? count - 2 : count - 1;
            this.chipAreaScrollOffset += Math.ceil(containerRect.left) - Math.ceil(chipAraeChildren[dif].getBoundingClientRect().left) + 1;
        }
        this.transform(this.chipAreaScrollOffset);
    }
    /**
     * @return {?}
     */
    conditionChangedCallback() {
        if (!!this.expression.searchVal || this.expression.searchVal === 0) {
            this.filter();
        }
        else if (this.value) {
            this.value = null;
        }
    }
    /**
     * @return {?}
     */
    unaryConditionChangedCallback() {
        if (this.value) {
            this.value = null;
        }
        if (this.expressionsList.find(item => item.expression === this.expression) === undefined) {
            this.addExpression(true);
        }
        this.filter();
    }
    /**
     * @return {?}
     */
    filter() {
        this.rootExpressionsTree = this.filteringService.createSimpleFilteringTree(this.column.field);
        this.filteringService.filter(this.column.field, this.rootExpressionsTree);
    }
}
IgxGridFilteringRowComponent.decorators = [
    { type: Component, args: [{
                preserveWhitespaces: false,
                selector: 'igx-grid-filtering-row',
                template: `<!-- Have to apply styles inline because of the overlay outlet ... -->
<igx-drop-down #inputGroupConditions [height]="'200px'" (onSelection)="onConditionsChanged($event)">
    <igx-drop-down-item
        *ngFor="let condition of conditions"
        [value]="condition"
        [isSelected]="isConditionSelected(condition)">
        <igx-icon fontSet="filtering-icons" [name]="getCondition(condition).iconName"></igx-icon>
        <span style="margin-left: 16px">{{ condition | filterCondition | titlecase }}</span>
    </igx-drop-down-item>
</igx-drop-down>

<ng-template #defaultFilterUI>
    <igx-input-group type="box" [displayDensity]="'compact'" [supressInputAutofocus]="true">
        <igx-prefix #inputGroupPrefix
                    (click)="toggleConditionsDropDown(inputGroupPrefix)"
                    (keydown)="onPrefixKeyDown($event)"
                    tabindex="0"
                    [igxDropDownItemNavigation]="inputGroupConditions">
            <igx-icon fontSet="filtering-icons" [name]="getIconName()"></igx-icon>
        </igx-prefix>
        <input
            #input
            igxInput
            tabindex="0"
            [placeholder]="placeholder"
            autocomplete="off"
            [(ngModel)]="value"
            [type]="type"
            [readonly]="isUnaryCondition"
            (keydown)="onInputKeyDown($event)"/>
            <igx-suffix *ngIf="input.value.length > 0" (keydown)="onClearKeyDown($event)" (click)="clearInput()" tabindex="0">
                <igx-icon>clear</igx-icon>
            </igx-suffix>
    </igx-input-group>
</ng-template>

<ng-template #defaultDateUI>
    <igx-datePicker tabindex="0" [(ngModel)]="value" [locale]="locale" (onClose)="datePickerClose()">
        <ng-template igxDatePickerTemplate let-openDialog="openDialog" let-displayData="displayData">
            <igx-input-group type="box" [displayDensity]="'compact'" [supressInputAutofocus]="true">
                <igx-prefix #inputGroupPrefix
                            tabindex="0"
                            (click)="toggleConditionsDropDown(inputGroupPrefix)"
                            (keydown)="onPrefixKeyDown($event)"
                            [igxDropDownItemNavigation]="inputGroupConditions">
                    <igx-icon fontSet="filtering-icons" [name]="expression.condition.iconName"></igx-icon>
                </igx-prefix>
                <input #input
                       igxInput
                       tabindex="0"
                (click)="openDialog()"
                [placeholder]="placeholder"
                autocomplete="off"
                       [value]="displayData | date"
                       [readonly]="true"
                       (keydown)="onInputKeyDown($event)"/>
                <igx-suffix *ngIf="input.value.length > 0" (keydown)="onClearKeyDown($event)" (click)="clearInput()" tabindex="0">
                    <igx-icon>clear</igx-icon>
                </igx-suffix>
            </igx-input-group>
            </ng-template>
    </igx-datePicker>
</ng-template>

<ng-container *ngTemplateOutlet="template; context: { $implicit: this }"></ng-container>

<button igxButton="icon" class="igx-grid__filtering-row-scroll-start" *ngIf="showArrows" (keydown)="onLeftArrowKeyDown($event)" (click)="scrollChipsOnArrowPress('left')">
    <igx-icon>navigate_before</igx-icon>
</button>

<div #container class="igx-grid__filtering-row-main">
    <div>
         <igx-chips-area #chipsArea>
            <ng-container *ngFor="let item of expressionsList; index as i; let last = last;" tabindex="0">
                <igx-chip #chip id='chip'
                    (onSelection)="onChipSelected($event, item.expression)"
                    (keydown)="onChipKeyDown($event, chip)"
                    (onRemove)="onChipRemoved($event, item)"
                    [selectable]="true"
                    [selected]="item.isSelected"
                    [displayDensity]="'cosy'"
                    [removable]="true">
                    <igx-icon
                        igxPrefix
                        fontSet="filtering-icons"
                        [name]="item.expression.condition.iconName">
                    </igx-icon>
                    <span>{{filteringService.getChipLabel(item.expression)}}</span>
                </igx-chip>

                <span id='operand' *ngIf="!last">
                    <button igxButton (click)="toggleOperatorsDropDown($event, i)" [igxDropDownItemNavigation]="operators">
                        <igx-icon>expand_more</igx-icon>
                        <span>{{filteringService.getOperatorAsString(item.afterOperator)}}</span>
                    </button>
                    <igx-drop-down #operators (onSelection)="onLogicOperatorChanged($event, item)">
                            <igx-drop-down-item [value]="0" [isSelected]="item.afterOperator === 0">And</igx-drop-down-item>
                            <igx-drop-down-item [value]="1" [isSelected]="item.afterOperator === 1">Or</igx-drop-down-item>
                    </igx-drop-down>
                </span>
            </ng-container>
        </igx-chips-area>
    </div>
</div>

<button igxButton="icon" class="igx-grid__filtering-row-scroll-end" *ngIf="showArrows" (click)="scrollChipsOnArrowPress('right')">
    <igx-icon>navigate_next</igx-icon>
</button>

<div #buttonsContainer class="igx-grid__filtering-row-editing-buttons">
    <button igxButton igxRipple (click)="clearFiltering()" [disabled]="disabled">Reset</button>
    <button #closeButton igxButton igxRipple (click)="close()">Close</button>
</div>
`
            },] },
];
/** @nocollapse */
IgxGridFilteringRowComponent.ctorParameters = () => [
    { type: IgxFilteringService, },
    { type: ElementRef, },
    { type: ChangeDetectorRef, },
];
IgxGridFilteringRowComponent.propDecorators = {
    "column": [{ type: Input },],
    "value": [{ type: Input },],
    "defaultFilterUI": [{ type: ViewChild, args: ['defaultFilterUI', { read: TemplateRef },] },],
    "defaultDateUI": [{ type: ViewChild, args: ['defaultDateUI', { read: TemplateRef },] },],
    "input": [{ type: ViewChild, args: ['input', { read: ElementRef },] },],
    "dropDownConditions": [{ type: ViewChild, args: ['inputGroupConditions', { read: IgxDropDownComponent },] },],
    "chipsArea": [{ type: ViewChild, args: ['chipsArea', { read: IgxChipsAreaComponent },] },],
    "dropDownOperators": [{ type: ViewChildren, args: ['operators', { read: IgxDropDownComponent },] },],
    "inputGroupPrefix": [{ type: ViewChild, args: ['inputGroupPrefix', { read: ElementRef },] },],
    "container": [{ type: ViewChild, args: ['container',] },],
    "operand": [{ type: ViewChild, args: ['operand',] },],
    "closeButton": [{ type: ViewChild, args: ['closeButton',] },],
    "cssClass": [{ type: HostBinding, args: ['class.igx-grid__filtering-row',] },],
    "onTabKeydown": [{ type: HostListener, args: ['keydown.tab', ['$event'],] },],
};
function IgxGridFilteringRowComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxGridFilteringRowComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxGridFilteringRowComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxGridFilteringRowComponent.propDecorators;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype._positionSettings;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype._conditionsOverlaySettings;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype._operatorsOverlaySettings;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.rootExpressionsTree;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.chipsAreaWidth;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.chipAreaScrollOffset;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.conditionChanged;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.unaryConditionChanged;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype._column;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.showArrows;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.expression;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.expressionsList;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.defaultFilterUI;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.defaultDateUI;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.input;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.dropDownConditions;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.chipsArea;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.dropDownOperators;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.inputGroupPrefix;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.container;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.operand;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.closeButton;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.cssClass;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.filteringService;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.element;
    /** @type {?} */
    IgxGridFilteringRowComponent.prototype.cdr;
}
