// igniteui-angular/esm5/lib/combo/combo.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive } from '@angular/core';
/**
 * @hidden
 */
var IgxComboHeaderDirective = /** @class */ (function () {
    function IgxComboHeaderDirective() {
    }
    IgxComboHeaderDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxComboHeader]'
                },] },
    ];
    return IgxComboHeaderDirective;
}());
export { IgxComboHeaderDirective };
function IgxComboHeaderDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboHeaderDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboHeaderDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxComboFooterDirective = /** @class */ (function () {
    function IgxComboFooterDirective() {
    }
    IgxComboFooterDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxComboFooter]'
                },] },
    ];
    return IgxComboFooterDirective;
}());
export { IgxComboFooterDirective };
function IgxComboFooterDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboFooterDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboFooterDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxComboItemDirective = /** @class */ (function () {
    function IgxComboItemDirective() {
    }
    IgxComboItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxComboItem]'
                },] },
    ];
    return IgxComboItemDirective;
}());
export { IgxComboItemDirective };
function IgxComboItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboItemDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxComboEmptyDirective = /** @class */ (function () {
    function IgxComboEmptyDirective() {
    }
    IgxComboEmptyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxComboEmpty]'
                },] },
    ];
    return IgxComboEmptyDirective;
}());
export { IgxComboEmptyDirective };
function IgxComboEmptyDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboEmptyDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboEmptyDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxComboHeaderItemDirective = /** @class */ (function () {
    function IgxComboHeaderItemDirective() {
    }
    IgxComboHeaderItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxComboHeaderItem]'
                },] },
    ];
    return IgxComboHeaderItemDirective;
}());
export { IgxComboHeaderItemDirective };
function IgxComboHeaderItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboHeaderItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboHeaderItemDirective.ctorParameters;
}
/**
 * @hidden
 */
var IgxComboAddItemDirective = /** @class */ (function () {
    function IgxComboAddItemDirective() {
    }
    IgxComboAddItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxComboAddItem]'
                },] },
    ];
    return IgxComboAddItemDirective;
}());
export { IgxComboAddItemDirective };
function IgxComboAddItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxComboAddItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxComboAddItemDirective.ctorParameters;
}
