// igniteui-angular/esm2015/lib/animations/slide/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { animate, animation, style } from '@angular/animations';
import { EaseIn, EaseOut } from '../easings';
const /** @type {?} */ base = [
    style({
        opacity: `{{startOpacity}}`,
        transform: `{{fromPosition}}`
    }),
    animate(`{{duration}} {{delay}} {{easing}}`, style({
        opacity: `{{endOpacity}}`,
        transform: `{{toPosition}}`
    }))
];
const /** @type {?} */ baseInParams = {
    delay: '0s',
    duration: '350ms',
    easing: EaseOut.quad,
    endOpacity: 1,
    fromPosition: 'translateY(-500px)',
    startOpacity: 0,
    toPosition: 'translateY(0)'
};
const /** @type {?} */ baseOutParams = {
    delay: '0s',
    duration: '350ms',
    easing: EaseIn.quad,
    endOpacity: 0,
    fromPosition: 'translateY(0)',
    startOpacity: 1,
    toPosition: 'translateY(-500px)'
};
const /** @type {?} */ slideInTop = animation(base, { params: baseInParams });
const /** @type {?} */ slideInLeft = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseOut.quad,
        endOpacity: 1,
        fromPosition: 'translateX(-500px)',
        startOpacity: 0,
        toPosition: 'translateY(0)'
    }
});
const /** @type {?} */ slideInRight = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseOut.quad,
        endOpacity: 1,
        fromPosition: 'translateX(500px)',
        startOpacity: 0,
        toPosition: 'translateY(0)'
    }
});
const /** @type {?} */ slideInBottom = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseOut.quad,
        endOpacity: 1,
        fromPosition: 'translateY(500px)',
        startOpacity: 0,
        toPosition: 'translateY(0)'
    }
});
const /** @type {?} */ slideInTr = animation(base, {
    params: Object.assign({}, baseInParams, { fromPosition: 'translateY(-500px) translateX(500px)', toPosition: 'translateY(0) translateX(0)' })
});
const /** @type {?} */ slideInTl = animation(base, {
    params: Object.assign({}, baseInParams, { fromPosition: 'translateY(-500px) translateX(-500px)', toPosition: 'translateY(0) translateX(0)' })
});
const /** @type {?} */ slideInBr = animation(base, {
    params: Object.assign({}, baseInParams, { fromPosition: 'translateY(500px) translateX(500px)', toPosition: 'translateY(0) translateX(0)' })
});
const /** @type {?} */ slideInBl = animation(base, {
    params: Object.assign({}, baseInParams, { fromPosition: 'translateY(500px) translateX(-500px)', toPosition: 'translateY(0) translateX(0)' })
});
const /** @type {?} */ slideOutTop = animation(base, { params: baseOutParams });
const /** @type {?} */ slideOutRight = animation(base, {
    params: Object.assign({}, baseOutParams, { toPosition: 'translateX(500px)' })
});
const /** @type {?} */ slideOutBottom = animation(base, {
    params: {
        delay: '0s',
        duration: '350ms',
        easing: EaseIn.quad,
        endOpacity: 0,
        fromPosition: 'translateY(0)',
        startOpacity: 1,
        toPosition: 'translateY(500px)'
    }
});
const /** @type {?} */ slideOutLeft = animation(base, {
    params: Object.assign({}, baseOutParams, { toPosition: 'translateX(-500px)' })
});
const /** @type {?} */ slideOutTr = animation(base, {
    params: Object.assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(-500px) translateX(500px)' })
});
const /** @type {?} */ slideOutBr = animation(base, {
    params: Object.assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(500px) translateX(500px)' })
});
const /** @type {?} */ slideOutBl = animation(base, {
    params: Object.assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(500px) translateX(-500px)' })
});
const /** @type {?} */ slideOutTl = animation(base, {
    params: Object.assign({}, baseOutParams, { fromPosition: 'translateY(0) translateX(0)', toPosition: 'translateY(-500px) translateX(-500px)' })
});
export { slideInTop, slideInRight, slideInBottom, slideInLeft, slideInTr, slideInBr, slideInBl, slideInTl, slideOutTop, slideOutBottom, slideOutRight, slideOutLeft, slideOutTr, slideOutBr, slideOutBl, slideOutTl };
