// igniteui-angular/esm5/lib/core/navigation/directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, HostListener, Input, NgModule } from '@angular/core';
import { IgxNavigationService } from './nav.service';
/**
 * Directive that can toggle targets through provided NavigationService.
 *
 * Usage:
 * ```
 * <button igxNavToggle="ID"> Toggle </button>
 * ```
 * Where the `ID` matches the ID of compatible `IToggleView` component.
 */
var IgxNavigationToggleDirective = /** @class */ (function () {
    function IgxNavigationToggleDirective(nav) {
        this.state = nav;
    }
    /**
     * @return {?}
     */
    IgxNavigationToggleDirective.prototype.toggleNavigationDrawer = /**
     * @return {?}
     */
    function () {
        this.state.toggle(this.target, true);
    };
    IgxNavigationToggleDirective.decorators = [
        { type: Directive, args: [{ selector: '[igxNavToggle]' },] },
    ];
    /** @nocollapse */
    IgxNavigationToggleDirective.ctorParameters = function () { return [
        { type: IgxNavigationService, },
    ]; };
    IgxNavigationToggleDirective.propDecorators = {
        "target": [{ type: Input, args: ['igxNavToggle',] },],
        "toggleNavigationDrawer": [{ type: HostListener, args: ['click',] },],
    };
    return IgxNavigationToggleDirective;
}());
export { IgxNavigationToggleDirective };
function IgxNavigationToggleDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavigationToggleDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavigationToggleDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxNavigationToggleDirective.propDecorators;
    /** @type {?} */
    IgxNavigationToggleDirective.prototype.state;
    /** @type {?} */
    IgxNavigationToggleDirective.prototype.target;
}
/**
 * Directive that can close targets through provided NavigationService.
 *
 * Usage:
 * ```
 * <button igxNavClose="ID"> Close </button>
 * ```
 * Where the `ID` matches the ID of compatible `IToggleView` component.
 */
var IgxNavigationCloseDirective = /** @class */ (function () {
    function IgxNavigationCloseDirective(nav) {
        this.state = nav;
    }
    /**
     * @return {?}
     */
    IgxNavigationCloseDirective.prototype.closeNavigationDrawer = /**
     * @return {?}
     */
    function () {
        this.state.close(this.target, true);
    };
    IgxNavigationCloseDirective.decorators = [
        { type: Directive, args: [{ selector: '[igxNavClose]' },] },
    ];
    /** @nocollapse */
    IgxNavigationCloseDirective.ctorParameters = function () { return [
        { type: IgxNavigationService, },
    ]; };
    IgxNavigationCloseDirective.propDecorators = {
        "target": [{ type: Input, args: ['igxNavClose',] },],
        "closeNavigationDrawer": [{ type: HostListener, args: ['click',] },],
    };
    return IgxNavigationCloseDirective;
}());
export { IgxNavigationCloseDirective };
function IgxNavigationCloseDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavigationCloseDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavigationCloseDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxNavigationCloseDirective.propDecorators;
    /** @type {?} */
    IgxNavigationCloseDirective.prototype.state;
    /** @type {?} */
    IgxNavigationCloseDirective.prototype.target;
}
var IgxNavigationModule = /** @class */ (function () {
    function IgxNavigationModule() {
    }
    IgxNavigationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IgxNavigationCloseDirective, IgxNavigationToggleDirective],
                    exports: [IgxNavigationCloseDirective, IgxNavigationToggleDirective],
                    providers: [IgxNavigationService]
                },] },
    ];
    return IgxNavigationModule;
}());
export { IgxNavigationModule };
function IgxNavigationModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavigationModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavigationModule.ctorParameters;
}
