// igniteui-cli/templates/angular/igx-ts/projects/empty/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Util_1 = require("../../../../../lib/Util");
class EmptyJsBlocksProject {
    constructor() {
        this.id = "angular";
        this.name = "empty";
        this.description = "Default empty angular project structure for Ignite UI for Angular";
        this.dependencies = [];
        this.framework = "angular";
        this.projectType = "igx-ts";
        this.hasExtraConfiguration = false;
        this.CUSTOM_THEME = `/* See: https://www.infragistics.com/products/ignite-ui-angular/angular/components/themes.html */
@import "~igniteui-angular/lib/core/styles/themes/index";

$primary: #731963 !default;
$secondary: #ce5712 !default;

$app-palette: igx-palette($primary, $secondary);

@include igx-core();
@include igx-theme($app-palette);
`;
        this.DEFAULT_THEME = `,
              "node_modules/igniteui-angular/styles/igniteui-angular.css"`;
    }
    installModules() {
        throw new Error("Method not implemented.");
    }
    upgradeIgniteUIPackage(projectPath, packagePath) {
        // noop
    }
    getExtraConfiguration() {
        return [];
    }
    setExtraConfiguration(extraConfigKeys) { }
    generateFiles(outputPath, name, theme, ...options) {
        const config = {
            "$(CustomTheme)": "",
            "$(DefaultTheme)": "",
            "$(cliVersion)": Util_1.Util.version(),
            "$(dash-name)": Util_1.Util.lowerDashed(name),
            "$(name)": name,
            "$(theme)": theme,
            "__path__": name
        };
        switch (theme) {
            case "Custom":
                config["$(CustomTheme)"] = this.CUSTOM_THEME;
                break;
            case "Default":
            default:
                config["$(DefaultTheme)"] = this.DEFAULT_THEME;
                break;
        }
        const pathsConfig = {};
        return Util_1.Util.processTemplates(path.join(__dirname, "./files"), path.join(outputPath, name), config, pathsConfig);
    }
}
module.exports = new EmptyJsBlocksProject();
