// igniteui-angular/esm5/lib/directives/radio/radio-group.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, NgModule, Input, QueryList, Output, EventEmitter, ContentChildren } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { IgxRadioComponent, RadioLabelPosition } from '../../radio/radio.component';
import { IgxRippleModule } from '../ripple/ripple.directive';
import { takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
var /** @type {?} */ noop = function () { };
var ɵ0 = noop;
var /** @type {?} */ nextId = 0;
/**
 * **Ignite UI for Angular Radio Group** -
 * [Documentation](https://www.infragistics.com/products/ignite-ui-angular/angular/components/radio_group.html)
 *
 * The Ignite UI Radio Group allows the user to select a single option from an available set of options that are listed side by side.
 *
 * Example:
 * ```html
 * <igx-radio-group name="radioGroup">
 *   <igx-radio *ngFor="let item of ['Foo', 'Bar', 'Baz']" value="{{item}}">
 *      {{item}}
 *   </igx-radio>
 * </igx-radio-group>
 * ```
 */
var IgxRadioGroupDirective = /** @class */ (function () {
    function IgxRadioGroupDirective() {
        /**
         * An event that is emitted after the radio group `value` is changed.
         * Provides references to the selected `IgxRadioComponent` and the `value` property as event arguments.
         * \@memberof IgxRadioGroupDirective
         */
        this.change = new EventEmitter();
        /**
         * @hidden
         */
        this._onChangeCallback = noop;
        /**
         * @hidden
         */
        this._name = "igx-radio-group-" + nextId++;
        /**
         * @hidden
         */
        this._value = null;
        /**
         * @hidden
         */
        this._selected = null;
        /**
         * @hidden
         */
        this._isInitialized = false;
        /**
         * @hidden
         */
        this._labelPosition = 'after';
        /**
         * @hidden
         */
        this._disabled = false;
        /**
         * @hidden
         */
        this._required = false;
        /**
         * @hidden
         */
        this.destroy$ = new Subject();
    }
    Object.defineProperty(IgxRadioGroupDirective.prototype, "value", {
        get: /**
         * Sets/gets the `value` attribute.
         * ```html
         * <igx-radio-group [value] = "'radioButtonValue'"></igx-radio-group>
         * ```
         * ```typescript
         * let value =  this.radioGroup.value;
         * ```
         * \@memberof IgxRadioGroupDirective
         * @return {?}
         */
        function () { return this._value; },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            if (this._value !== newValue) {
                this._value = newValue;
                this._selectRadioButton();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxRadioGroupDirective.prototype, "name", {
        get: /**
         * Sets/gets the `name` attribute of the radio group component. All child radio buttons inherits this name.
         * ```html
         * <igx-radio-group name = "Radio1"></igx-radio-group>
         *  ```
         * ```typescript
         * let name =  this.radioGroup.name;
         * ```
         * \@memberof IgxRadioGroupDirective
         * @return {?}
         */
        function () { return this._name; },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            if (this._name !== newValue) {
                this._name = newValue;
                this._setRadioButtonNames();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxRadioGroupDirective.prototype, "required", {
        get: /**
         * Sets/gets whether the radio group is required.
         * If not set, `required` will have value `false`.
         * ```html
         * <igx-radio-group [required] = "true"></igx-radio-group>
         * ```
         * ```typescript
         * let isRequired =  this.radioGroup.required;
         * ```
         * \@memberof IgxRadioGroupDirective
         * @return {?}
         */
        function () { return this._required; },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            if (this._required !== newValue) {
                this._required = newValue;
                this._setRadioButtonsRequired();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxRadioGroupDirective.prototype, "disabled", {
        get: /**
         * An \@Input property that allows you to disable the radio group. By default it's false.
         * ```html
         * <igx-radio-group [disabled]="true"></igx-radio-group>
         * ```
         * \@memberof IgxRadioGroupDirective
         * @return {?}
         */
        function () { return this._disabled; },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            if (this._disabled !== newValue) {
                this._disabled = newValue;
                this._disableRadioButtons();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxRadioGroupDirective.prototype, "labelPosition", {
        get: /**
         * Sets/gets the position of the `label` in the child radio buttons.
         * If not set, `labelPosition` will have value `"after"`.
         * ```html
         * <igx-radio-group labelPosition = "before"></igx-radio-group>
         * ```
         * ```typescript
         * let labelPosition =  this.radioGroup.labelPosition;
         * ```
         * \@memberof IgxRadioGroupDirective
         * @return {?}
         */
        function () { return this._labelPosition; },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            if (this._labelPosition !== newValue) {
                this._labelPosition = newValue === RadioLabelPosition.BEFORE ? RadioLabelPosition.BEFORE : RadioLabelPosition.AFTER;
                this._setRadioButtonLabelPosition();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxRadioGroupDirective.prototype, "selected", {
        get: /**
         * Sets/gets the selected child radio button.
         * ```typescript
         * let selectedButton = this.radioGroup.selected;
         * this.radioGroup.selected = selectedButton;
         * ```
         * \@memberof IgxRadioGroupDirective
         * @return {?}
         */
        function () { return this._selected; },
        set: /**
         * @param {?} selected
         * @return {?}
         */
        function (selected) {
            if (this._selected !== selected) {
                this._selected = selected;
                this.value = selected ? selected.value : null;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxRadioGroupDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // The initial value can possibly be set by NgModel and it is possible that
        // the OnInit of the NgModel occurs after the OnInit of this class.
        this._isInitialized = true;
        setTimeout(function () { _this._initRadioButtons(); });
    };
    /**
     * Checks whether the provided value is consistent to the current radio button.
     * If it is, the checked attribute will have value `true` and selected property will contain the selected `IgxRadioComponent`.
     * ```typescript
     * this.radioGroup.writeValue('radioButtonValue');
     * ```
     * @param {?} value
     * @return {?}
     */
    IgxRadioGroupDirective.prototype.writeValue = /**
     * Checks whether the provided value is consistent to the current radio button.
     * If it is, the checked attribute will have value `true` and selected property will contain the selected `IgxRadioComponent`.
     * ```typescript
     * this.radioGroup.writeValue('radioButtonValue');
     * ```
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    IgxRadioGroupDirective.prototype.registerOnChange = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) { this._onChangeCallback = fn; };
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    IgxRadioGroupDirective.prototype.registerOnTouched = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                button.registerOnTouched(fn);
            });
        }
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype.ngOnDestroy = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.destroy$.next(true);
        this.destroy$.complete();
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._initRadioButtons = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                button.name = _this._name;
                button.labelPosition = _this._labelPosition;
                button.disabled = _this._disabled;
                button.required = _this._required;
                if (_this._value && button.value === _this._value) {
                    button.checked = true;
                    _this._selected = button;
                }
                button.change.pipe(takeUntil(_this.destroy$)).subscribe(function (ev) { return _this._selectedRadioButtonChanged(ev); });
            });
        }
    };
    /**
     * @hidden
     * @param {?} args
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._selectedRadioButtonChanged = /**
     * @hidden
     * @param {?} args
     * @return {?}
     */
    function (args) {
        if (this._selected !== args.radio) {
            if (this._selected) {
                this._selected.checked = false;
            }
            this._selected = args.radio;
        }
        this._value = args.value;
        if (this._isInitialized) {
            this.change.emit(args);
            this._onChangeCallback(this.value);
        }
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._setRadioButtonNames = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                button.name = _this._name;
            });
        }
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._selectRadioButton = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                if (!_this._value) {
                    // no value - uncheck all radio buttons
                    if (button.checked) {
                        button.checked = false;
                    }
                }
                else {
                    if (_this._value === button.value) {
                        // selected button
                        if (_this._selected !== button) {
                            _this._selected = button;
                        }
                        if (!button.checked) {
                            button.select();
                        }
                    }
                    else {
                        // non-selected button
                        if (button.checked) {
                            button.checked = false;
                        }
                    }
                }
            });
        }
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._setRadioButtonLabelPosition = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                button.labelPosition = _this._labelPosition;
            });
        }
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._disableRadioButtons = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                button.disabled = _this._disabled;
            });
        }
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxRadioGroupDirective.prototype._setRadioButtonsRequired = /**
     * @hidden
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.radioButtons) {
            this.radioButtons.forEach(function (button) {
                button.required = _this._required;
            });
        }
    };
    IgxRadioGroupDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'igx-radio-group, [igxRadioGroup]',
                    providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: IgxRadioGroupDirective, multi: true }]
                },] },
    ];
    /** @nocollapse */
    IgxRadioGroupDirective.propDecorators = {
        "radioButtons": [{ type: ContentChildren, args: [IgxRadioComponent,] },],
        "value": [{ type: Input },],
        "name": [{ type: Input },],
        "required": [{ type: Input },],
        "disabled": [{ type: Input },],
        "labelPosition": [{ type: Input },],
        "selected": [{ type: Input },],
        "change": [{ type: Output },],
    };
    return IgxRadioGroupDirective;
}());
export { IgxRadioGroupDirective };
function IgxRadioGroupDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxRadioGroupDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxRadioGroupDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxRadioGroupDirective.propDecorators;
    /**
     * Returns reference to the child radio buttons.
     * ```typescript
     * let radioButtons =  this.radioGroup.radioButtons;
     * ```
     * \@memberof IgxRadioGroupDirective
     * @type {?}
     */
    IgxRadioGroupDirective.prototype.radioButtons;
    /**
     * An event that is emitted after the radio group `value` is changed.
     * Provides references to the selected `IgxRadioComponent` and the `value` property as event arguments.
     * \@memberof IgxRadioGroupDirective
     * @type {?}
     */
    IgxRadioGroupDirective.prototype.change;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._onChangeCallback;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._name;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._value;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._selected;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._isInitialized;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._labelPosition;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._disabled;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype._required;
    /**
     * @hidden
     * @type {?}
     */
    IgxRadioGroupDirective.prototype.destroy$;
}
/**
 * The IgxRadioModule provides the {\@link IgxRadioGroupDirective} and {\@link IgxRadioComponent} inside your application.
 */
var IgxRadioModule = /** @class */ (function () {
    function IgxRadioModule() {
    }
    IgxRadioModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IgxRadioGroupDirective, IgxRadioComponent],
                    exports: [IgxRadioGroupDirective, IgxRadioComponent],
                    imports: [IgxRippleModule]
                },] },
    ];
    return IgxRadioModule;
}());
export { IgxRadioModule };
function IgxRadioModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxRadioModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxRadioModule.ctorParameters;
}
export { ɵ0 };
