// igniteui-angular/esm5/lib/services/excel/worksheet-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ExportUtilities } from '../exporter-common/export-utilities';
import { WorksheetDataDictionary } from './worksheet-data-dictionary';
/**
 * @hidden
 */
var /**
 * @hidden
 */
WorksheetData = /** @class */ (function () {
    function WorksheetData(_data, options, indexOfLastPinnedColumn, sort) {
        this._data = _data;
        this.options = options;
        this.indexOfLastPinnedColumn = indexOfLastPinnedColumn;
        this.sort = sort;
        this.initializeData();
    }
    Object.defineProperty(WorksheetData.prototype, "data", {
        get: /**
         * @return {?}
         */
        function () {
            return this._data;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WorksheetData.prototype, "columnCount", {
        get: /**
         * @return {?}
         */
        function () {
            return this._columnCount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WorksheetData.prototype, "rowCount", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rowCount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WorksheetData.prototype, "isEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.rowCount || !this._columnCount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WorksheetData.prototype, "keys", {
        get: /**
         * @return {?}
         */
        function () {
            return this._keys;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WorksheetData.prototype, "isSpecialData", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isSpecialData;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WorksheetData.prototype, "dataDictionary", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dataDictionary;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    WorksheetData.prototype.initializeData = /**
     * @return {?}
     */
    function () {
        if (!this._data || this._data.length === 0) {
            return;
        }
        this._keys = ExportUtilities.getKeysFromData(this._data);
        if (this._keys.length === 0) {
            return;
        }
        this._isSpecialData = ExportUtilities.isSpecialData(this._data);
        this._columnCount = this._keys.length;
        this._rowCount = this._data.length + 1;
        this._dataDictionary = new WorksheetDataDictionary(this._columnCount, this.options.columnWidth);
    };
    return WorksheetData;
}());
/**
 * @hidden
 */
export { WorksheetData };
function WorksheetData_tsickle_Closure_declarations() {
    /** @type {?} */
    WorksheetData.prototype._columnCount;
    /** @type {?} */
    WorksheetData.prototype._rowCount;
    /** @type {?} */
    WorksheetData.prototype._dataDictionary;
    /** @type {?} */
    WorksheetData.prototype._keys;
    /** @type {?} */
    WorksheetData.prototype._isSpecialData;
    /** @type {?} */
    WorksheetData.prototype._data;
    /** @type {?} */
    WorksheetData.prototype.options;
    /** @type {?} */
    WorksheetData.prototype.indexOfLastPinnedColumn;
    /** @type {?} */
    WorksheetData.prototype.sort;
}
