// igniteui-angular/esm5/lib/services/excel/worksheet-data-dictionary.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ExportUtilities } from '../exporter-common/export-utilities';
/**
 * @hidden
 */
var WorksheetDataDictionary = /** @class */ (function () {
    function WorksheetDataDictionary(columnCount, columnWidth) {
        this.hasNonStringValues = false;
        this._dictionary = {};
        this._widthsDictionary = {};
        this._counter = 0;
        this.dirtyKeyCollections();
        this._calculateColumnWidth = !columnWidth;
        this._columnWidths = new Array(columnCount);
        this._columnTypeInfo = new Array(columnCount);
        if (!this._calculateColumnWidth) {
            this._columnWidths.fill(columnWidth);
        }
        this.stringsCount = 0;
    }
    Object.defineProperty(WorksheetDataDictionary.prototype, "columnWidths", {
        get: /**
         * @return {?}
         */
        function () {
            return this._columnWidths;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @param {?} column
     * @param {?} isHeader
     * @return {?}
     */
    WorksheetDataDictionary.prototype.saveValue = /**
     * @param {?} value
     * @param {?} column
     * @param {?} isHeader
     * @return {?}
     */
    function (value, column, isHeader) {
        if (this._columnTypeInfo[column] === undefined && isHeader === false) {
            this._columnTypeInfo[column] = typeof value === 'string' ||
                typeof value === 'boolean' ||
                value instanceof Date;
        }
        var /** @type {?} */ sanitizedValue = '';
        var /** @type {?} */ isSavedAsString = this._columnTypeInfo[column] || isHeader;
        if (isSavedAsString) {
            sanitizedValue = this.sanitizeValue(value);
            if (this._dictionary[sanitizedValue] === undefined) {
                this._dictionary[sanitizedValue] = this._counter++;
                this.dirtyKeyCollections();
            }
            this.stringsCount++;
        }
        else {
            this.hasNonStringValues = true;
        }
        if (this._calculateColumnWidth) {
            var /** @type {?} */ width = this.getTextWidth(value);
            var /** @type {?} */ maxWidth = Math.max(this._columnWidths[column] || 0, width);
            this._columnWidths[column] = maxWidth;
        }
        return isSavedAsString ? this.getSanitizedValue(sanitizedValue) : -1;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    WorksheetDataDictionary.prototype.getValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.getSanitizedValue(this.sanitizeValue(value));
    };
    /**
     * @param {?} sanitizedValue
     * @return {?}
     */
    WorksheetDataDictionary.prototype.getSanitizedValue = /**
     * @param {?} sanitizedValue
     * @return {?}
     */
    function (sanitizedValue) {
        return this._dictionary[sanitizedValue];
    };
    /**
     * @return {?}
     */
    WorksheetDataDictionary.prototype.getKeys = /**
     * @return {?}
     */
    function () {
        if (!this._keysAreValid) {
            this._keys = Object.keys(this._dictionary);
            this._keysAreValid = true;
        }
        return this._keys;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    WorksheetDataDictionary.prototype.getTextWidth = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this._widthsDictionary[value] === undefined) {
            var /** @type {?} */ context = this.getContext();
            var /** @type {?} */ metrics = context.measureText(value);
            this._widthsDictionary[value] = metrics.width + WorksheetDataDictionary.TEXT_PADDING;
        }
        return this._widthsDictionary[value];
    };
    /**
     * @return {?}
     */
    WorksheetDataDictionary.prototype.getContext = /**
     * @return {?}
     */
    function () {
        if (!this._context) {
            var /** @type {?} */ canvas = document.createElement('canvas');
            this._context = canvas.getContext('2d');
            this._context.font = WorksheetDataDictionary.DEFAULT_FONT;
        }
        return this._context;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    WorksheetDataDictionary.prototype.sanitizeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (ExportUtilities.hasValue(value) === false) {
            return '';
        }
        else {
            var /** @type {?} */ stringValue = String(value);
            return stringValue.replace(/&/g, '&amp;')
                .replace(/</g, '&lt;')
                .replace(/>/g, '&gt;')
                .replace(/"/g, '&quot;')
                .replace(/'/g, '&apos;');
        }
    };
    /**
     * @return {?}
     */
    WorksheetDataDictionary.prototype.dirtyKeyCollections = /**
     * @return {?}
     */
    function () {
        this._keysAreValid = false;
    };
    WorksheetDataDictionary.DEFAULT_FONT = '11pt Calibri';
    WorksheetDataDictionary.TEXT_PADDING = 5;
    return WorksheetDataDictionary;
}());
export { WorksheetDataDictionary };
function WorksheetDataDictionary_tsickle_Closure_declarations() {
    /** @type {?} */
    WorksheetDataDictionary.DEFAULT_FONT;
    /** @type {?} */
    WorksheetDataDictionary.TEXT_PADDING;
    /** @type {?} */
    WorksheetDataDictionary.prototype._dictionary;
    /** @type {?} */
    WorksheetDataDictionary.prototype._widthsDictionary;
    /** @type {?} */
    WorksheetDataDictionary.prototype._sortedKeysByValue;
    /** @type {?} */
    WorksheetDataDictionary.prototype._sortedKeysByValueAreValid;
    /** @type {?} */
    WorksheetDataDictionary.prototype._keys;
    /** @type {?} */
    WorksheetDataDictionary.prototype._keysAreValid;
    /** @type {?} */
    WorksheetDataDictionary.prototype._counter;
    /** @type {?} */
    WorksheetDataDictionary.prototype._calculateColumnWidth;
    /** @type {?} */
    WorksheetDataDictionary.prototype._columnWidths;
    /** @type {?} */
    WorksheetDataDictionary.prototype._context;
    /** @type {?} */
    WorksheetDataDictionary.prototype._columnTypeInfo;
    /** @type {?} */
    WorksheetDataDictionary.prototype.hasNonStringValues;
    /** @type {?} */
    WorksheetDataDictionary.prototype.stringsCount;
}
