// igniteui-angular/esm2015/lib/directives/suffix/suffix.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, Directive } from '@angular/core';
/**
 * @hidden
 */
export class IgxSuffixDirective {
}
IgxSuffixDirective.decorators = [
    { type: Directive, args: [{
                selector: 'igx-suffix,[igxSuffix]'
            },] },
];
function IgxSuffixDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxSuffixDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxSuffixDirective.ctorParameters;
}
/**
 * @hidden
 */
export class IgxSuffixModule {
}
IgxSuffixModule.decorators = [
    { type: NgModule, args: [{
                declarations: [IgxSuffixDirective],
                exports: [IgxSuffixDirective]
            },] },
];
function IgxSuffixModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxSuffixModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxSuffixModule.ctorParameters;
}
