// igniteui-angular/esm5/lib/grids/tree-grid/tree-grid.pipes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Pipe } from '@angular/core';
import { cloneArray, cloneHierarchicalArray } from '../../core/utils';
import { DataUtil } from '../../data-operations/data-util';
import { GridBaseAPIService } from '../api.service';
/**
 * @hidden
 */
var IgxTreeGridHierarchizingPipe = /** @class */ (function () {
    function IgxTreeGridHierarchizingPipe(gridAPI) {
        this.gridAPI = /** @type {?} */ (gridAPI);
    }
    /**
     * @param {?} collection
     * @param {?} primaryKey
     * @param {?} foreignKey
     * @param {?} childDataKey
     * @param {?} id
     * @param {?} pipeTrigger
     * @return {?}
     */
    IgxTreeGridHierarchizingPipe.prototype.transform = /**
     * @param {?} collection
     * @param {?} primaryKey
     * @param {?} foreignKey
     * @param {?} childDataKey
     * @param {?} id
     * @param {?} pipeTrigger
     * @return {?}
     */
    function (collection, primaryKey, foreignKey, childDataKey, id, pipeTrigger) {
        var /** @type {?} */ grid = this.gridAPI.get(id);
        var /** @type {?} */ hierarchicalRecords = [];
        var /** @type {?} */ treeGridRecordsMap = new Map();
        if (primaryKey && foreignKey) {
            hierarchicalRecords = this.hierarchizeFlatData(id, collection, primaryKey, foreignKey, treeGridRecordsMap);
            grid.flatData = grid.data;
        }
        else if (childDataKey) {
            var /** @type {?} */ flatData = [];
            hierarchicalRecords = this.hierarchizeRecursive(id, collection, primaryKey, childDataKey, undefined, flatData, 0, treeGridRecordsMap);
            grid.flatData = flatData;
        }
        grid.records = treeGridRecordsMap;
        grid.rootRecords = hierarchicalRecords;
        return hierarchicalRecords;
    };
    /**
     * @param {?} primaryKey
     * @param {?} rowData
     * @return {?}
     */
    IgxTreeGridHierarchizingPipe.prototype.getRowID = /**
     * @param {?} primaryKey
     * @param {?} rowData
     * @return {?}
     */
    function (primaryKey, rowData) {
        return primaryKey ? rowData[primaryKey] : rowData;
    };
    /**
     * @param {?} id
     * @param {?} collection
     * @param {?} primaryKey
     * @param {?} foreignKey
     * @param {?} map
     * @return {?}
     */
    IgxTreeGridHierarchizingPipe.prototype.hierarchizeFlatData = /**
     * @param {?} id
     * @param {?} collection
     * @param {?} primaryKey
     * @param {?} foreignKey
     * @param {?} map
     * @return {?}
     */
    function (id, collection, primaryKey, foreignKey, map) {
        var _this = this;
        var /** @type {?} */ result = [];
        var /** @type {?} */ missingParentRecords = [];
        collection.forEach(function (row) {
            var /** @type {?} */ record = {
                rowID: _this.getRowID(primaryKey, row),
                data: row,
                children: [],
                path: []
            };
            var /** @type {?} */ parent = map.get(row[foreignKey]);
            if (parent) {
                record.parent = parent;
                if (parent) {
                    (_a = record.path).push.apply(_a, tslib_1.__spread(parent.path));
                    record.path.push(parent.rowID);
                }
                parent.children.push(record);
            }
            else {
                missingParentRecords.push(record);
            }
            map.set(row[primaryKey], record);
            var _a;
        });
        missingParentRecords.forEach(function (record) {
            var /** @type {?} */ parent = map.get(record.data[foreignKey]);
            if (parent) {
                record.parent = parent;
                parent.children.push(record);
            }
            else {
                result.push(record);
            }
        });
        this.setIndentationLevels(id, result, 0);
        return result;
    };
    /**
     * @param {?} id
     * @param {?} collection
     * @param {?} indentationLevel
     * @return {?}
     */
    IgxTreeGridHierarchizingPipe.prototype.setIndentationLevels = /**
     * @param {?} id
     * @param {?} collection
     * @param {?} indentationLevel
     * @return {?}
     */
    function (id, collection, indentationLevel) {
        for (var /** @type {?} */ i = 0; i < collection.length; i++) {
            var /** @type {?} */ record = collection[i];
            record.level = indentationLevel;
            record.expanded = this.gridAPI.get_row_expansion_state(id, record.rowID, record.level);
            if (record.children && record.children.length > 0) {
                this.setIndentationLevels(id, record.children, indentationLevel + 1);
            }
        }
    };
    /**
     * @param {?} id
     * @param {?} collection
     * @param {?} primaryKey
     * @param {?} childDataKey
     * @param {?} parent
     * @param {?} flatData
     * @param {?} indentationLevel
     * @param {?} map
     * @return {?}
     */
    IgxTreeGridHierarchizingPipe.prototype.hierarchizeRecursive = /**
     * @param {?} id
     * @param {?} collection
     * @param {?} primaryKey
     * @param {?} childDataKey
     * @param {?} parent
     * @param {?} flatData
     * @param {?} indentationLevel
     * @param {?} map
     * @return {?}
     */
    function (id, collection, primaryKey, childDataKey, parent, flatData, indentationLevel, map) {
        var /** @type {?} */ result = [];
        for (var /** @type {?} */ i = 0; i < collection.length; i++) {
            var /** @type {?} */ item = collection[i];
            var /** @type {?} */ record = {
                rowID: this.getRowID(primaryKey, item),
                data: item,
                parent: parent,
                level: indentationLevel,
                path: []
            };
            if (parent) {
                (_a = record.path).push.apply(_a, tslib_1.__spread(parent.path));
                record.path.push(parent.rowID);
            }
            record.expanded = this.gridAPI.get_row_expansion_state(id, record.rowID, record.level);
            flatData.push(item);
            map.set(record.rowID, record);
            record.children = item[childDataKey] ?
                this.hierarchizeRecursive(id, item[childDataKey], primaryKey, childDataKey, record, flatData, indentationLevel + 1, map) :
                undefined;
            result.push(record);
        }
        return result;
        var _a;
    };
    IgxTreeGridHierarchizingPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'treeGridHierarchizing',
                    pure: true
                },] },
    ];
    /** @nocollapse */
    IgxTreeGridHierarchizingPipe.ctorParameters = function () { return [
        { type: GridBaseAPIService, },
    ]; };
    return IgxTreeGridHierarchizingPipe;
}());
export { IgxTreeGridHierarchizingPipe };
function IgxTreeGridHierarchizingPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTreeGridHierarchizingPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTreeGridHierarchizingPipe.ctorParameters;
    /** @type {?} */
    IgxTreeGridHierarchizingPipe.prototype.gridAPI;
}
/**
 * @hidden
 */
var IgxTreeGridFlatteningPipe = /** @class */ (function () {
    function IgxTreeGridFlatteningPipe(gridAPI) {
        this.gridAPI = /** @type {?} */ (gridAPI);
    }
    /**
     * @param {?} collection
     * @param {?} id
     * @param {?} expandedLevels
     * @param {?} expandedStates
     * @param {?} pipeTrigger
     * @return {?}
     */
    IgxTreeGridFlatteningPipe.prototype.transform = /**
     * @param {?} collection
     * @param {?} id
     * @param {?} expandedLevels
     * @param {?} expandedStates
     * @param {?} pipeTrigger
     * @return {?}
     */
    function (collection, id, expandedLevels, expandedStates, pipeTrigger) {
        var /** @type {?} */ grid = this.gridAPI.get(id);
        var /** @type {?} */ data = [];
        grid.processedRootRecords = collection;
        grid.processedRecords = new Map();
        this.getFlatDataRecursive(collection, data, expandedLevels, expandedStates, id, true);
        return data;
    };
    /**
     * @param {?} collection
     * @param {?=} data
     * @param {?=} expandedLevels
     * @param {?=} expandedStates
     * @param {?=} gridID
     * @param {?=} parentExpanded
     * @return {?}
     */
    IgxTreeGridFlatteningPipe.prototype.getFlatDataRecursive = /**
     * @param {?} collection
     * @param {?=} data
     * @param {?=} expandedLevels
     * @param {?=} expandedStates
     * @param {?=} gridID
     * @param {?=} parentExpanded
     * @return {?}
     */
    function (collection, data, expandedLevels, expandedStates, gridID, parentExpanded) {
        if (data === void 0) { data = []; }
        if (!collection || !collection.length) {
            return;
        }
        for (var /** @type {?} */ i = 0; i < collection.length; i++) {
            var /** @type {?} */ hierarchicalRecord = collection[i];
            if (parentExpanded) {
                data.push(hierarchicalRecord);
            }
            var /** @type {?} */ grid = this.gridAPI.get(gridID);
            hierarchicalRecord.expanded = this.gridAPI.get_row_expansion_state(gridID, hierarchicalRecord.rowID, hierarchicalRecord.level);
            this.updateNonProcessedRecordExpansion(grid, hierarchicalRecord);
            grid.processedRecords.set(hierarchicalRecord.rowID, hierarchicalRecord);
            this.getFlatDataRecursive(hierarchicalRecord.children, data, expandedLevels, expandedStates, gridID, parentExpanded && hierarchicalRecord.expanded);
        }
    };
    /**
     * @param {?} grid
     * @param {?} record
     * @return {?}
     */
    IgxTreeGridFlatteningPipe.prototype.updateNonProcessedRecordExpansion = /**
     * @param {?} grid
     * @param {?} record
     * @return {?}
     */
    function (grid, record) {
        var /** @type {?} */ rec = grid.records.get(record.rowID);
        rec.expanded = record.expanded;
    };
    IgxTreeGridFlatteningPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'treeGridFlattening',
                    pure: true
                },] },
    ];
    /** @nocollapse */
    IgxTreeGridFlatteningPipe.ctorParameters = function () { return [
        { type: GridBaseAPIService, },
    ]; };
    return IgxTreeGridFlatteningPipe;
}());
export { IgxTreeGridFlatteningPipe };
function IgxTreeGridFlatteningPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTreeGridFlatteningPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTreeGridFlatteningPipe.ctorParameters;
    /** @type {?} */
    IgxTreeGridFlatteningPipe.prototype.gridAPI;
}
/**
 * @hidden
 */
var IgxTreeGridSortingPipe = /** @class */ (function () {
    function IgxTreeGridSortingPipe(gridAPI) {
        this.gridAPI = /** @type {?} */ (gridAPI);
    }
    /**
     * @param {?} hierarchicalData
     * @param {?} expressions
     * @param {?} id
     * @param {?} pipeTrigger
     * @return {?}
     */
    IgxTreeGridSortingPipe.prototype.transform = /**
     * @param {?} hierarchicalData
     * @param {?} expressions
     * @param {?} id
     * @param {?} pipeTrigger
     * @return {?}
     */
    function (hierarchicalData, expressions, id, pipeTrigger) {
        var /** @type {?} */ state = { expressions: [] };
        var /** @type {?} */ grid = this.gridAPI.get(id);
        state.expressions = grid.sortingExpressions;
        var /** @type {?} */ result;
        if (!state.expressions.length) {
            result = hierarchicalData;
        }
        else {
            result = DataUtil.hierarchicalSort(hierarchicalData, state, undefined);
        }
        return result;
    };
    IgxTreeGridSortingPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'treeGridSorting',
                    pure: true
                },] },
    ];
    /** @nocollapse */
    IgxTreeGridSortingPipe.ctorParameters = function () { return [
        { type: GridBaseAPIService, },
    ]; };
    return IgxTreeGridSortingPipe;
}());
export { IgxTreeGridSortingPipe };
function IgxTreeGridSortingPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTreeGridSortingPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTreeGridSortingPipe.ctorParameters;
    /** @type {?} */
    IgxTreeGridSortingPipe.prototype.gridAPI;
}
/**
 * @hidden
 */
var IgxTreeGridPagingPipe = /** @class */ (function () {
    function IgxTreeGridPagingPipe(gridAPI) {
        this.gridAPI = /** @type {?} */ (gridAPI);
    }
    /**
     * @param {?} collection
     * @param {?=} page
     * @param {?=} perPage
     * @param {?=} id
     * @param {?=} pipeTrigger
     * @return {?}
     */
    IgxTreeGridPagingPipe.prototype.transform = /**
     * @param {?} collection
     * @param {?=} page
     * @param {?=} perPage
     * @param {?=} id
     * @param {?=} pipeTrigger
     * @return {?}
     */
    function (collection, page, perPage, id, pipeTrigger) {
        if (page === void 0) { page = 0; }
        if (perPage === void 0) { perPage = 15; }
        if (!this.gridAPI.get(id).paging) {
            return collection;
        }
        var /** @type {?} */ state = {
            index: page,
            recordsPerPage: perPage
        };
        var /** @type {?} */ result = DataUtil.page(cloneArray(collection), state);
        this.gridAPI.get(id).pagingState = state;
        return result;
    };
    IgxTreeGridPagingPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'treeGridPaging',
                    pure: true
                },] },
    ];
    /** @nocollapse */
    IgxTreeGridPagingPipe.ctorParameters = function () { return [
        { type: GridBaseAPIService, },
    ]; };
    return IgxTreeGridPagingPipe;
}());
export { IgxTreeGridPagingPipe };
function IgxTreeGridPagingPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTreeGridPagingPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTreeGridPagingPipe.ctorParameters;
    /** @type {?} */
    IgxTreeGridPagingPipe.prototype.gridAPI;
}
/**
 * @hidden
 */
var IgxTreeGridTransactionPipe = /** @class */ (function () {
    function IgxTreeGridTransactionPipe(gridAPI) {
        this.gridAPI = /** @type {?} */ (gridAPI);
    }
    /**
     * @param {?} collection
     * @param {?} id
     * @param {?} pipeTrigger
     * @return {?}
     */
    IgxTreeGridTransactionPipe.prototype.transform = /**
     * @param {?} collection
     * @param {?} id
     * @param {?} pipeTrigger
     * @return {?}
     */
    function (collection, id, pipeTrigger) {
        var /** @type {?} */ grid = this.gridAPI.get(id);
        if (collection && grid.transactions.enabled) {
            var /** @type {?} */ primaryKey = grid.primaryKey;
            if (!primaryKey) {
                return collection;
            }
            var /** @type {?} */ foreignKey = grid.foreignKey;
            var /** @type {?} */ childDataKey = grid.childDataKey;
            if (foreignKey) {
                return DataUtil.mergeTransactions(cloneArray(collection), grid.transactions.getAggregatedChanges(true), grid.primaryKey);
            }
            else if (childDataKey) {
                return DataUtil.mergeHierarchicalTransactions(cloneHierarchicalArray(collection, childDataKey), grid.transactions.getAggregatedChanges(true), childDataKey, grid.primaryKey);
            }
        }
        return collection;
    };
    IgxTreeGridTransactionPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'treeGridTransaction',
                    pure: true
                },] },
    ];
    /** @nocollapse */
    IgxTreeGridTransactionPipe.ctorParameters = function () { return [
        { type: GridBaseAPIService, },
    ]; };
    return IgxTreeGridTransactionPipe;
}());
export { IgxTreeGridTransactionPipe };
function IgxTreeGridTransactionPipe_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTreeGridTransactionPipe.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTreeGridTransactionPipe.ctorParameters;
    /** @type {?} */
    IgxTreeGridTransactionPipe.prototype.gridAPI;
}
