// igniteui-angular/esm2015/lib/core/displayDensity.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { InjectionToken, Input } from '@angular/core';
/** @enum {string} */
const DisplayDensity = {
    comfortable: 'comfortable',
    cosy: 'cosy',
    compact: 'compact',
};
export { DisplayDensity };
/**
 * Describes the object used to configure the DisplayDensity in Angular DI.
 * @record
 */
export function IDisplayDensity() { }
function IDisplayDensity_tsickle_Closure_declarations() {
    /** @type {?} */
    IDisplayDensity.prototype.displayDensity;
}
/**
 * Defines the DisplayDensity DI token.
 */
export const /** @type {?} */ DisplayDensityToken = new InjectionToken('DisplayDensity');
/**
 * Base class containing all logic required for implementing DisplayDensity.
 */
export class DisplayDensityBase {
    /**
     * @param {?} displayDensityOptions
     */
    constructor(displayDensityOptions) {
        this.displayDensityOptions = displayDensityOptions;
    }
    /**
     * Returns the theme of the component.
     * The default theme is `comfortable`.
     * Available options are `comfortable`, `cosy`, `compact`.
     * ```typescript
     * let componentTheme = this.component.displayDensity;
     * ```
     * @return {?}
     */
    get displayDensity() {
        return this._displayDensity;
    }
    /**
     * Sets the theme of the component.
     * @param {?} val
     * @return {?}
     */
    set displayDensity(val) {
        switch (val) {
            case 'compact':
                this._displayDensity = "compact" /* compact */;
                break;
            case 'cosy':
                this._displayDensity = "cosy" /* cosy */;
                break;
            case 'comfortable':
            default:
                this._displayDensity = "comfortable" /* comfortable */;
        }
    }
    /**
     * @hidden
     * @return {?}
     */
    isCosy() {
        return this._displayDensity === "cosy" /* cosy */ ||
            (!this._displayDensity && this.displayDensityOptions && this.displayDensityOptions.displayDensity === "cosy" /* cosy */);
    }
    /**
     * @hidden
     * @return {?}
     */
    isComfortable() {
        return this._displayDensity === "comfortable" /* comfortable */ ||
            (!this._displayDensity && (!this.displayDensityOptions ||
                this.displayDensityOptions.displayDensity === "comfortable" /* comfortable */));
    }
    /**
     * @hidden
     * @return {?}
     */
    isCompact() {
        return this._displayDensity === "compact" /* compact */ ||
            (!this._displayDensity && this.displayDensityOptions && this.displayDensityOptions.displayDensity === "compact" /* compact */);
    }
}
DisplayDensityBase.propDecorators = {
    "displayDensity": [{ type: Input },],
};
function DisplayDensityBase_tsickle_Closure_declarations() {
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    DisplayDensityBase.propDecorators;
    /** @type {?} */
    DisplayDensityBase.prototype._displayDensity;
    /** @type {?} */
    DisplayDensityBase.prototype.displayDensityOptions;
}
