// igniteui-angular/esm5/lib/animations/misc/pulsate.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { animate, animation, keyframes, style } from '@angular/animations';
var /** @type {?} */ heartbeatBase = [
    style({
        animationTimingFunction: "ease-out",
        transform: "scale(1)",
        transformOrigin: "center center"
    }),
    animate("{{duration}} {{delay}} {{easing}}", keyframes([
        style({
            animationTimingFunction: "ease-in",
            offset: 0.1,
            transform: "scale(0.91)"
        }),
        style({
            animationTimingFunction: "ease-out",
            offset: 0.17,
            transform: "scale(0.98)"
        }),
        style({
            animationTimingFunction: "ease-in",
            offset: 0.33,
            transform: "scale(0.87)"
        }),
        style({
            animationTimingFunction: "ease-out",
            offset: 0.45,
            transform: "scale(1)"
        })
    ]))
];
var /** @type {?} */ heartbeatParams = {
    delay: '0s',
    duration: '1.5s',
    easing: 'ease-in-out'
};
var /** @type {?} */ pulsateBase = [
    animate("{{duration}} {{delay}} {{easing}}", keyframes([
        style({
            offset: 0,
            transform: "scale({{fromScale}})"
        }),
        style({
            offset: 0.5,
            transform: "scale({{toScale}})"
        }),
        style({
            offset: 1,
            transform: "scale({{fromScale}})"
        })
    ]))
];
var /** @type {?} */ pulsateParams = {
    delay: '0s',
    duration: '.5s',
    easing: 'ease-in-out',
    fromScale: 1,
    toScale: 1.1
};
var /** @type {?} */ blinkBase = [
    animate("{{duration}} {{delay}} {{easing}}", keyframes([
        style({
            offset: 0,
            opacity: .8,
            transform: "scale({{fromScale}})"
        }),
        style({
            offset: 0.8,
            opacity: 0,
            transform: "scale({{midScale}})"
        }),
        style({
            offset: 1,
            opacity: 0,
            transform: "scale({{toScale}})"
        })
    ]))
];
var /** @type {?} */ blinkParams = {
    delay: '0s',
    duration: '.8s',
    easing: 'ease-in-out',
    fromScale: .2,
    midScale: 1.2,
    toScale: 2.2
};
var /** @type {?} */ pulsateFwd = animation(pulsateBase, {
    params: tslib_1.__assign({}, pulsateParams)
});
var /** @type {?} */ pulsateBck = animation(pulsateBase, {
    params: tslib_1.__assign({}, pulsateParams, { toScale: .9 })
});
var /** @type {?} */ heartbeat = animation(heartbeatBase, {
    params: tslib_1.__assign({}, heartbeatParams)
});
var /** @type {?} */ blink = animation(blinkBase, {
    params: tslib_1.__assign({}, blinkParams)
});
export { heartbeat, pulsateFwd, pulsateBck, blink };
