// igniteui-angular/esm2015/lib/tabs/tabs.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, forwardRef, Host, HostBinding, Inject, TemplateRef } from '@angular/core';
import { IgxTabsComponent } from './tabs.component';
/** @enum {string} */
const ButtonStyle = {
    VISIBLE: 'visible',
    HIDDEN: 'hidden',
    NOT_DISPLAYED: 'not_displayed',
};
export class IgxRightButtonStyleDirective {
    /**
     * @param {?} tabs
     */
    constructor(tabs) {
        this.tabs = tabs;
    }
    /**
     * @return {?}
     */
    getRightButtonStyle() {
        const /** @type {?} */ viewPortWidth = this.tabs.viewPort.nativeElement.offsetWidth;
        const /** @type {?} */ itemsContainerWidth = this.tabs.itemsContainer.nativeElement.offsetWidth;
        const /** @type {?} */ headerContainerWidth = this.tabs.headerContainer.nativeElement.offsetWidth;
        const /** @type {?} */ offset = this.tabs.offset;
        const /** @type {?} */ total = offset + viewPortWidth;
        if (itemsContainerWidth <= headerContainerWidth && offset === 0) {
            return ButtonStyle.NOT_DISPLAYED;
        }
        if (itemsContainerWidth > total) {
            return ButtonStyle.VISIBLE;
        }
        else {
            return ButtonStyle.HIDDEN;
        }
    }
    /**
     * @return {?}
     */
    get visibleCSS() {
        return (this.getRightButtonStyle() === ButtonStyle.VISIBLE) ? true : false;
    }
    /**
     * @return {?}
     */
    get hiddenCSS() {
        return (this.getRightButtonStyle() === ButtonStyle.HIDDEN) ? true : false;
    }
    /**
     * @return {?}
     */
    get notDisplayedCSS() {
        return (this.getRightButtonStyle() === ButtonStyle.NOT_DISPLAYED) ? true : false;
    }
}
IgxRightButtonStyleDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxRightButtonStyle]'
            },] },
];
/** @nocollapse */
IgxRightButtonStyleDirective.ctorParameters = () => [
    { type: IgxTabsComponent, decorators: [{ type: Host }, { type: Inject, args: [forwardRef(() => IgxTabsComponent),] },] },
];
IgxRightButtonStyleDirective.propDecorators = {
    "visibleCSS": [{ type: HostBinding, args: ['class.igx-tabs__header-button',] },],
    "hiddenCSS": [{ type: HostBinding, args: ['class.igx-tabs__header-button--hidden',] },],
    "notDisplayedCSS": [{ type: HostBinding, args: ['class.igx-tabs__header-button--none',] },],
};
function IgxRightButtonStyleDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxRightButtonStyleDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxRightButtonStyleDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxRightButtonStyleDirective.propDecorators;
    /** @type {?} */
    IgxRightButtonStyleDirective.prototype.tabs;
}
export class IgxLeftButtonStyleDirective {
    /**
     * @param {?} tabs
     */
    constructor(tabs) {
        this.tabs = tabs;
    }
    /**
     * @return {?}
     */
    getLeftButtonStyle() {
        const /** @type {?} */ itemsContainerWidth = this.tabs.itemsContainer.nativeElement.offsetWidth;
        const /** @type {?} */ headerContainerWidth = this.tabs.headerContainer.nativeElement.offsetWidth;
        const /** @type {?} */ offset = this.tabs.offset;
        if (offset === 0) {
            if (itemsContainerWidth <= headerContainerWidth) {
                return ButtonStyle.NOT_DISPLAYED;
            }
            return ButtonStyle.HIDDEN;
        }
        else {
            return ButtonStyle.VISIBLE;
        }
    }
    /**
     * @return {?}
     */
    get visibleCSS() {
        return (this.getLeftButtonStyle() === ButtonStyle.VISIBLE) ? true : false;
    }
    /**
     * @return {?}
     */
    get hiddenCSS() {
        return (this.getLeftButtonStyle() === ButtonStyle.HIDDEN) ? true : false;
    }
    /**
     * @return {?}
     */
    get notDisplayedCSS() {
        return (this.getLeftButtonStyle() === ButtonStyle.NOT_DISPLAYED) ? true : false;
    }
}
IgxLeftButtonStyleDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxLeftButtonStyle]'
            },] },
];
/** @nocollapse */
IgxLeftButtonStyleDirective.ctorParameters = () => [
    { type: IgxTabsComponent, decorators: [{ type: Host }, { type: Inject, args: [forwardRef(() => IgxTabsComponent),] },] },
];
IgxLeftButtonStyleDirective.propDecorators = {
    "visibleCSS": [{ type: HostBinding, args: ['class.igx-tabs__header-button',] },],
    "hiddenCSS": [{ type: HostBinding, args: ['class.igx-tabs__header-button--hidden',] },],
    "notDisplayedCSS": [{ type: HostBinding, args: ['class.igx-tabs__header-button--none',] },],
};
function IgxLeftButtonStyleDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxLeftButtonStyleDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxLeftButtonStyleDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxLeftButtonStyleDirective.propDecorators;
    /** @type {?} */
    IgxLeftButtonStyleDirective.prototype.tabs;
}
export class IgxTabItemTemplateDirective {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
IgxTabItemTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxTab]'
            },] },
];
/** @nocollapse */
IgxTabItemTemplateDirective.ctorParameters = () => [
    { type: TemplateRef, },
];
function IgxTabItemTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxTabItemTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxTabItemTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxTabItemTemplateDirective.prototype.template;
}
