// igniteui-angular/esm2015/lib/expansion-panel/expansion-panel-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, HostBinding, Inject, forwardRef, ElementRef, Input, ChangeDetectorRef } from '@angular/core';
import { IgxExpansionPanelComponent } from './expansion-panel.component';
export class IgxExpansionPanelBodyComponent {
    /**
     * @param {?} panel
     * @param {?} element
     * @param {?} cdr
     */
    constructor(panel, element, cdr) {
        this.panel = panel;
        this.element = element;
        this.cdr = cdr;
        this._labelledBy = '';
        this._label = '';
        this.cssClass = `igx-expansion-panel__body`;
        this._title = '';
        this.role = 'region';
    }
    /**
     * @return {?}
     */
    get label() {
        return this._label || this.panel.id + '-region';
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set label(val) {
        this._label = val;
    }
    /**
     * @return {?}
     */
    get labelledBy() {
        return this._labelledBy;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set labelledBy(val) {
        this._labelledBy = val;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.labelledBy = this.panel.headerId;
        this.label = this.panel.id + '-region';
    }
}
IgxExpansionPanelBodyComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-expansion-panel-body',
                template: `<ng-content></ng-content>`
            },] },
];
/** @nocollapse */
IgxExpansionPanelBodyComponent.ctorParameters = () => [
    { type: IgxExpansionPanelComponent, decorators: [{ type: Inject, args: [forwardRef(() => IgxExpansionPanelComponent),] },] },
    { type: ElementRef, },
    { type: ChangeDetectorRef, },
];
IgxExpansionPanelBodyComponent.propDecorators = {
    "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel__body',] },],
    "label": [{ type: Input }, { type: HostBinding, args: ['attr.aria-label',] },],
    "labelledBy": [{ type: Input }, { type: HostBinding, args: ['attr.aria-labelledby',] },],
    "role": [{ type: Input }, { type: HostBinding, args: ['attr.role',] },],
};
function IgxExpansionPanelBodyComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelBodyComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelBodyComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelBodyComponent.propDecorators;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype._labelledBy;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype._label;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.cssClass;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype._title;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.role;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.panel;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.element;
    /** @type {?} */
    IgxExpansionPanelBodyComponent.prototype.cdr;
}
