// igniteui-angular/esm2015/lib/expansion-panel/expansion-panel.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, HostBinding } from '@angular/core';
/**
 * @hidden
 */
export class IgxExpansionPanelTitleDirective {
    constructor() {
        this.cssClass = `igx-expansion-panel__header-title`;
    }
}
IgxExpansionPanelTitleDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-expansion-panel-title'
            },] },
];
/** @nocollapse */
IgxExpansionPanelTitleDirective.propDecorators = {
    "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel__header-title',] },],
};
function IgxExpansionPanelTitleDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelTitleDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelTitleDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelTitleDirective.propDecorators;
    /** @type {?} */
    IgxExpansionPanelTitleDirective.prototype.cssClass;
}
/**
 * @hidden
 */
export class IgxExpansionPanelDescriptionDirective {
    constructor() {
        this.cssClass = `igx-expansion-panel__header-description`;
    }
}
IgxExpansionPanelDescriptionDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-expansion-panel-description'
            },] },
];
/** @nocollapse */
IgxExpansionPanelDescriptionDirective.propDecorators = {
    "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel__header-description',] },],
};
function IgxExpansionPanelDescriptionDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelDescriptionDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelDescriptionDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelDescriptionDirective.propDecorators;
    /** @type {?} */
    IgxExpansionPanelDescriptionDirective.prototype.cssClass;
}
/**
 * @hidden
 */
export class IgxExpansionPanelIconDirective {
}
IgxExpansionPanelIconDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'igx-expansion-panel-icon'
            },] },
];
function IgxExpansionPanelIconDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelIconDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelIconDirective.ctorParameters;
}
