// igniteui-angular/esm2015/lib/expansion-panel/expansion-panel-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ChangeDetectorRef, ElementRef, HostBinding, HostListener, Input, Host, EventEmitter, Output, ContentChild } from '@angular/core';
import { IgxExpansionPanelComponent } from './expansion-panel.component';
import { IgxExpansionPanelIconDirective } from './expansion-panel.directives';
/**
 * @record
 */
export function IExpansionPanelEventArgs() { }
function IExpansionPanelEventArgs_tsickle_Closure_declarations() {
    /** @type {?} */
    IExpansionPanelEventArgs.prototype.event;
    /** @type {?} */
    IExpansionPanelEventArgs.prototype.panel;
}
/** @enum {string} */
const ICON_POSITION = {
    LEFT: 'left',
    NONE: 'none',
    RIGHT: 'right',
};
export { ICON_POSITION };
export class IgxExpansionPanelHeaderComponent {
    /**
     * @param {?} panel
     * @param {?} cdr
     * @param {?} elementRef
     */
    constructor(panel, cdr, elementRef) {
        this.panel = panel;
        this.cdr = cdr;
        this.elementRef = elementRef;
        this._iconTemplate = false;
        this.id = '';
        this.lv = '3';
        this.role = 'heading';
        this.iconPosition = ICON_POSITION.LEFT;
        this.onInteraction = new EventEmitter();
        this.cssClass = 'igx-expansion-panel__header';
        this.disabled = false;
        this.id = `${this.panel.id}-header`;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set iconTemplate(val) {
        this._iconTemplate = /** @type {?} */ (val);
    }
    /**
     * @return {?}
     */
    get iconTemplate() {
        return this._iconTemplate;
    }
    /**
     * @return {?}
     */
    get controls() {
        return this.panel.id;
    }
    /**
     * @return {?}
     */
    get isExpanded() {
        return !this.panel.collapsed;
    }
    /**
     * @param {?=} evt
     * @return {?}
     */
    onAction(evt) {
        if (this.disabled) {
            evt.stopPropagation();
            return;
        }
        this.onInteraction.emit({ event: evt, panel: this.panel });
        this.panel.toggle(evt);
        evt.preventDefault();
    }
    /**
     * @return {?}
     */
    get iconPositionClass() {
        switch (this.iconPosition) {
            case (ICON_POSITION.LEFT):
                return `igx-expansion-panel__header-icon--start`;
            case (ICON_POSITION.RIGHT):
                return `igx-expansion-panel__header-icon--end`;
            case (ICON_POSITION.NONE):
                return `igx-expansion-panel__header-icon--none`;
            default:
                return '';
        }
    }
}
IgxExpansionPanelHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'igx-expansion-panel-header',
                template: `<div class="igx-expansion-panel__header-inner" tabindex="0" role="button" [attr.id]="id"
[attr.aria-disabled]="disabled" [attr.aria-expanded]="isExpanded" [attr.aria-controls]="controls">
    <div class="igx-expansion-panel__title-wrapper">
        <ng-content select="igx-expansion-panel-title"></ng-content>
        <ng-content select="igx-expansion-panel-description"></ng-content>
    </div>
    <ng-content></ng-content>
    <div [class]="iconPositionClass">
        <ng-content *ngIf="iconTemplate" select="igx-expansion-panel-icon"></ng-content>
        <igx-icon *ngIf="!iconTemplate" fontSet="material">
            {{panel.collapsed? 'expand_more':'expand_less'}}
        </igx-icon>
    </div>
</div>
`
            },] },
];
/** @nocollapse */
IgxExpansionPanelHeaderComponent.ctorParameters = () => [
    { type: IgxExpansionPanelComponent, decorators: [{ type: Host },] },
    { type: ChangeDetectorRef, },
    { type: ElementRef, },
];
IgxExpansionPanelHeaderComponent.propDecorators = {
    "iconTemplate": [{ type: ContentChild, args: [IgxExpansionPanelIconDirective,] },],
    "lv": [{ type: HostBinding, args: ['attr.aria-level',] }, { type: Input },],
    "role": [{ type: HostBinding, args: ['attr.role',] }, { type: Input },],
    "iconPosition": [{ type: Input },],
    "onInteraction": [{ type: Output },],
    "cssClass": [{ type: HostBinding, args: ['class.igx-expansion-panel__header',] },],
    "isExpanded": [{ type: HostBinding, args: ['class.igx-expansion-panel__header--expanded',] },],
    "disabled": [{ type: Input }, { type: HostBinding, args: ['class.igx-expansion-panel--disabled',] },],
    "onAction": [{ type: HostListener, args: ['keydown.Enter', ['$event'],] }, { type: HostListener, args: ['keydown.Space', ['$event'],] }, { type: HostListener, args: ['keydown.Spacebar', ['$event'],] }, { type: HostListener, args: ['click', ['$event'],] },],
};
function IgxExpansionPanelHeaderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxExpansionPanelHeaderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxExpansionPanelHeaderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxExpansionPanelHeaderComponent.propDecorators;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype._iconTemplate;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.id;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.lv;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.role;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.iconPosition;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.onInteraction;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.cssClass;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.disabled;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.panel;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.cdr;
    /** @type {?} */
    IgxExpansionPanelHeaderComponent.prototype.elementRef;
}
