// igniteui-angular/esm2015/lib/data-operations/filtering-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { FilteringLogic } from './filtering-expression.interface';
import { FilteringExpressionsTree } from './filtering-expressions-tree';
/**
 * @record
 */
export function IFilteringStrategy() { }
function IFilteringStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    IFilteringStrategy.prototype.filter;
}
/**
 * @abstract
 */
export class BaseFilteringStrategy {
    /**
     * @param {?} rec
     * @param {?} expr
     * @return {?}
     */
    findMatchByExpression(rec, expr) {
        const /** @type {?} */ cond = expr.condition;
        const /** @type {?} */ val = this.getFieldValue(rec, expr.fieldName);
        return cond.logic(val, expr.searchVal, expr.ignoreCase);
    }
    /**
     * @param {?} rec
     * @param {?} expressions
     * @return {?}
     */
    matchRecord(rec, expressions) {
        if (expressions) {
            if (expressions instanceof FilteringExpressionsTree) {
                const /** @type {?} */ expressionsTree = /** @type {?} */ (expressions);
                const /** @type {?} */ operator = /** @type {?} */ (expressionsTree.operator);
                let /** @type {?} */ match, /** @type {?} */ matchOperand, /** @type {?} */ operand;
                if (expressionsTree.filteringOperands) {
                    for (let /** @type {?} */ i = 0; i < expressionsTree.filteringOperands.length; i++) {
                        operand = expressionsTree.filteringOperands[i];
                        matchOperand = this.matchRecord(rec, operand);
                        if (match === undefined) {
                            match = matchOperand;
                        }
                        else if (operator === FilteringLogic.And) {
                            match = match && matchOperand;
                        }
                        else if (operator === FilteringLogic.Or) {
                            match = match || matchOperand;
                        }
                    }
                }
                return match === undefined ? true : match;
            }
            else {
                const /** @type {?} */ expression = /** @type {?} */ (expressions);
                return this.findMatchByExpression(rec, expression);
            }
        }
        return true;
    }
}
function BaseFilteringStrategy_tsickle_Closure_declarations() {
    /**
     * @abstract
     * @param {?} data
     * @param {?} expressionsTree
     * @return {?}
     */
    BaseFilteringStrategy.prototype.filter = function (data, expressionsTree) { };
    /**
     * @abstract
     * @param {?} rec
     * @param {?} fieldName
     * @return {?}
     */
    BaseFilteringStrategy.prototype.getFieldValue = function (rec, fieldName) { };
}
export class FilteringStrategy extends BaseFilteringStrategy {
    /**
     * @template T
     * @param {?} data
     * @param {?} expressionsTree
     * @return {?}
     */
    filter(data, expressionsTree) {
        let /** @type {?} */ i;
        let /** @type {?} */ rec;
        const /** @type {?} */ len = data.length;
        const /** @type {?} */ res = [];
        if (!expressionsTree || !expressionsTree.filteringOperands || expressionsTree.filteringOperands.length === 0 || !len) {
            return data;
        }
        for (i = 0; i < len; i++) {
            rec = data[i];
            if (this.matchRecord(rec, expressionsTree)) {
                res.push(rec);
            }
        }
        return res;
    }
    /**
     * @param {?} rec
     * @param {?} fieldName
     * @return {?}
     */
    getFieldValue(rec, fieldName) {
        return rec[fieldName];
    }
}
