// igniteui-angular/esm2015/lib/animations/flip/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { animate, animation, keyframes, style } from '@angular/animations';
import { EaseOut } from '../easings';
const /** @type {?} */ baseRecipe = [
    style({
        backfaceVisibility: 'hidden',
        transformStyle: 'preserve-3d'
    }),
    animate(`{{duration}} {{delay}} {{easing}}`, keyframes([
        style({
            offset: 0,
            transform: `translateZ({{startDistance}})
                rotate3d({{rotateX}}, {{rotateY}}, {{rotateZ}}, {{startAngle}}deg)`
        }),
        style({
            offset: 1,
            transform: `translateZ({{endDistance}})
                rotate3d({{rotateX}}, {{rotateY}}, {{rotateZ}}, {{endAngle}}deg)`
        })
    ]))
];
const /** @type {?} */ baseParams = {
    delay: '0s',
    duration: '600ms',
    easing: EaseOut.quad,
    endAngle: 180,
    endDistance: '0px',
    rotateX: 1,
    rotateY: 0,
    rotateZ: 0,
    startAngle: 0,
    startDistance: '0px'
};
const /** @type {?} */ flipTop = animation(baseRecipe, {
    params: Object.assign({}, baseParams)
});
const /** @type {?} */ flipBottom = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { endAngle: -180 })
});
const /** @type {?} */ flipLeft = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { rotateX: 0, rotateY: 1 })
});
const /** @type {?} */ flipRight = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { endAngle: -180, rotateX: 0, rotateY: 1 })
});
const /** @type {?} */ flipHorFwd = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { endDistance: '170px' })
});
const /** @type {?} */ flipHorBck = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { endDistance: '-170px' })
});
const /** @type {?} */ flipVerFwd = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { endDistance: '170px', rotateX: 0, rotateY: 1 })
});
const /** @type {?} */ flipVerBck = animation(baseRecipe, {
    params: Object.assign({}, baseParams, { endDistance: '-170px', rotateX: 0, rotateY: 1 })
});
export { flipTop, flipRight, flipBottom, flipLeft, flipHorFwd, flipHorBck, flipVerFwd, flipVerBck };
