// igniteui-angular/esm5/lib/navigation-drawer/navigation-drawer.directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, HostBinding, Input, TemplateRef } from '@angular/core';
var IgxNavDrawerItemDirective = /** @class */ (function () {
    function IgxNavDrawerItemDirective() {
        /**
         * @hidden
         */
        this.active = false;
        /**
         * @hidden
         */
        this.isHeader = false;
        /**
         * @hidden
         */
        this.activeClass = 'igx-nav-drawer__item--active';
    }
    Object.defineProperty(IgxNavDrawerItemDirective.prototype, "defaultCSS", {
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return !this.active && !this.isHeader;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxNavDrawerItemDirective.prototype, "currentCSS", {
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this.active && !this.isHeader;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxNavDrawerItemDirective.prototype, "headerCSS", {
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this.isHeader;
        },
        enumerable: true,
        configurable: true
    });
    IgxNavDrawerItemDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxDrawerItem]',
                    exportAs: 'igxDrawerItem'
                },] },
    ];
    /** @nocollapse */
    IgxNavDrawerItemDirective.propDecorators = {
        "active": [{ type: Input, args: ['active',] },],
        "isHeader": [{ type: Input, args: ['isHeader',] },],
        "defaultCSS": [{ type: HostBinding, args: ['class.igx-nav-drawer__item',] },],
        "currentCSS": [{ type: HostBinding, args: ['class.igx-nav-drawer__item--active',] },],
        "headerCSS": [{ type: HostBinding, args: ['class.igx-nav-drawer__item--header',] },],
    };
    return IgxNavDrawerItemDirective;
}());
export { IgxNavDrawerItemDirective };
function IgxNavDrawerItemDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavDrawerItemDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavDrawerItemDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxNavDrawerItemDirective.propDecorators;
    /**
     * @hidden
     * @type {?}
     */
    IgxNavDrawerItemDirective.prototype.active;
    /**
     * @hidden
     * @type {?}
     */
    IgxNavDrawerItemDirective.prototype.isHeader;
    /**
     * @hidden
     * @type {?}
     */
    IgxNavDrawerItemDirective.prototype.activeClass;
}
var IgxNavDrawerTemplateDirective = /** @class */ (function () {
    function IgxNavDrawerTemplateDirective(template) {
        this.template = template;
    }
    IgxNavDrawerTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxDrawer]'
                },] },
    ];
    /** @nocollapse */
    IgxNavDrawerTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxNavDrawerTemplateDirective;
}());
export { IgxNavDrawerTemplateDirective };
function IgxNavDrawerTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavDrawerTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavDrawerTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxNavDrawerTemplateDirective.prototype.template;
}
var IgxNavDrawerMiniTemplateDirective = /** @class */ (function () {
    function IgxNavDrawerMiniTemplateDirective(template) {
        this.template = template;
    }
    IgxNavDrawerMiniTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxDrawerMini]'
                },] },
    ];
    /** @nocollapse */
    IgxNavDrawerMiniTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxNavDrawerMiniTemplateDirective;
}());
export { IgxNavDrawerMiniTemplateDirective };
function IgxNavDrawerMiniTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxNavDrawerMiniTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxNavDrawerMiniTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxNavDrawerMiniTemplateDirective.prototype.template;
}
