// igniteui-angular/esm5/lib/animations/flip/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { animate, animation, keyframes, style } from '@angular/animations';
import { EaseOut } from '../easings';
var /** @type {?} */ baseRecipe = [
    style({
        backfaceVisibility: 'hidden',
        transformStyle: 'preserve-3d'
    }),
    animate("{{duration}} {{delay}} {{easing}}", keyframes([
        style({
            offset: 0,
            transform: "translateZ({{startDistance}})\n                rotate3d({{rotateX}}, {{rotateY}}, {{rotateZ}}, {{startAngle}}deg)"
        }),
        style({
            offset: 1,
            transform: "translateZ({{endDistance}})\n                rotate3d({{rotateX}}, {{rotateY}}, {{rotateZ}}, {{endAngle}}deg)"
        })
    ]))
];
var /** @type {?} */ baseParams = {
    delay: '0s',
    duration: '600ms',
    easing: EaseOut.quad,
    endAngle: 180,
    endDistance: '0px',
    rotateX: 1,
    rotateY: 0,
    rotateZ: 0,
    startAngle: 0,
    startDistance: '0px'
};
var /** @type {?} */ flipTop = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams)
});
var /** @type {?} */ flipBottom = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { endAngle: -180 })
});
var /** @type {?} */ flipLeft = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { rotateX: 0, rotateY: 1 })
});
var /** @type {?} */ flipRight = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { endAngle: -180, rotateX: 0, rotateY: 1 })
});
var /** @type {?} */ flipHorFwd = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { endDistance: '170px' })
});
var /** @type {?} */ flipHorBck = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { endDistance: '-170px' })
});
var /** @type {?} */ flipVerFwd = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { endDistance: '170px', rotateX: 0, rotateY: 1 })
});
var /** @type {?} */ flipVerBck = animation(baseRecipe, {
    params: tslib_1.__assign({}, baseParams, { endDistance: '-170px', rotateX: 0, rotateY: 1 })
});
export { flipTop, flipRight, flipBottom, flipLeft, flipHorFwd, flipHorBck, flipVerFwd, flipVerBck };
