// igniteui-angular/esm5/lib/grids/filtering/grid-filtering.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { GridBaseAPIService } from '../api.service';
import { IgxIconService } from '../../icon/icon.service';
import { FilteringExpressionsTree } from '../../data-operations/filtering-expressions-tree';
import icons from './svgIcons';
import { FilteringLogic } from '../../data-operations/filtering-expression.interface';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { IgxGridFilterConditionPipe } from '../grid-common.pipes';
import { TitleCasePipe, DatePipe } from '@angular/common';
var /** @type {?} */ FILTERING_ICONS_FONT_SET = 'filtering-icons';
/**
 * @hidden
 */
var /**
 * @hidden
 */
ExpressionUI = /** @class */ (function () {
    function ExpressionUI() {
        this.isSelected = false;
    }
    return ExpressionUI;
}());
/**
 * @hidden
 */
export { ExpressionUI };
function ExpressionUI_tsickle_Closure_declarations() {
    /** @type {?} */
    ExpressionUI.prototype.expression;
    /** @type {?} */
    ExpressionUI.prototype.beforeOperator;
    /** @type {?} */
    ExpressionUI.prototype.afterOperator;
    /** @type {?} */
    ExpressionUI.prototype.isSelected;
}
/**
 * @hidden
 */
var IgxFilteringService = /** @class */ (function () {
    function IgxFilteringService(gridAPI, iconService) {
        this.gridAPI = gridAPI;
        this.iconService = iconService;
        this.columnsWithComplexFilter = new Set();
        this.areEventsSubscribed = false;
        this.destroy$ = new Subject();
        this.isFiltering = false;
        this.columnToExpressionsMap = new Map();
        this.filterPipe = new IgxGridFilterConditionPipe();
        this.titlecasePipe = new TitleCasePipe();
        this.datePipe = new DatePipe(window.navigator.language);
        this.isFilterRowVisible = false;
        this.filteredColumn = null;
        this.selectedExpression = null;
        this.columnToChipToFocus = new Map();
        this.columnToMoreIconHidden = new Map();
        this.columnStartIndex = -1;
    }
    /**
     * @return {?}
     */
    IgxFilteringService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy$.next(true);
        this.destroy$.complete();
    };
    Object.defineProperty(IgxFilteringService.prototype, "grid", {
        get: /**
         * @return {?}
         */
        function () {
            return this.gridAPI.get(this.gridId);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Subscribe to grid's events.
     * @return {?}
     */
    IgxFilteringService.prototype.subscribeToEvents = /**
     * Subscribe to grid's events.
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.areEventsSubscribed) {
            this.areEventsSubscribed = true;
            this.grid.onColumnResized.pipe(takeUntil(this.destroy$)).subscribe(function (eventArgs) {
                _this.updateFilteringCell(eventArgs.column.field);
            });
            this.grid.parentVirtDir.onChunkLoad.pipe(takeUntil(this.destroy$)).subscribe(function (eventArgs) {
                if (eventArgs.startIndex !== _this.columnStartIndex) {
                    _this.columnStartIndex = eventArgs.startIndex;
                    _this.grid.filterCellList.forEach(function (filterCell) {
                        filterCell.updateFilterCellArea();
                        if (filterCell.getChipToFocus()) {
                            _this.columnToChipToFocus.set(filterCell.column.field, false);
                            filterCell.focusChip();
                        }
                    });
                }
            });
            this.grid.onColumnMovingEnd.pipe(takeUntil(this.destroy$)).subscribe(function (event) {
                _this.grid.filterCellList.forEach(function (filterCell) {
                    filterCell.updateFilterCellArea();
                });
            });
        }
    };
    /**
     * Execute filtering on the grid.
     * @param {?} field
     * @param {?} expressionsTree
     * @return {?}
     */
    IgxFilteringService.prototype.filter = /**
     * Execute filtering on the grid.
     * @param {?} field
     * @param {?} expressionsTree
     * @return {?}
     */
    function (field, expressionsTree) {
        var _this = this;
        this.isFiltering = true;
        this.grid.filter(field, null, expressionsTree);
        // Wait for the change detection to update filtered data through the pipes and then emit the event.
        requestAnimationFrame(function () { return _this.grid.onFilteringDone.emit(expressionsTree); });
        this.isFiltering = false;
    };
    /**
     * Clear the filter of a given column.
     * @param {?} field
     * @return {?}
     */
    IgxFilteringService.prototype.clearFilter = /**
     * Clear the filter of a given column.
     * @param {?} field
     * @return {?}
     */
    function (field) {
        var _this = this;
        this.isFiltering = true;
        this.grid.clearFilter(field);
        var /** @type {?} */ expr = this.grid.filteringExpressionsTree.find(field);
        // Wait for the change detection to update filtered data through the pipes and then emit the event.
        requestAnimationFrame(function () { return _this.grid.onFilteringDone.emit(/** @type {?} */ (expr)); });
        var /** @type {?} */ expressions = this.getExpressions(field);
        expressions.length = 0;
        this.isFiltering = false;
    };
    /**
     * Register filtering SVG icons in the icon service.
     * @return {?}
     */
    IgxFilteringService.prototype.registerSVGIcons = /**
     * Register filtering SVG icons in the icon service.
     * @return {?}
     */
    function () {
        try {
            for (var icons_1 = tslib_1.__values(icons), icons_1_1 = icons_1.next(); !icons_1_1.done; icons_1_1 = icons_1.next()) {
                var icon = icons_1_1.value;
                if (!this.iconService.isSvgIconCached(icon.name, FILTERING_ICONS_FONT_SET)) {
                    this.iconService.addSvgIconFromText(icon.name, icon.value, FILTERING_ICONS_FONT_SET);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (icons_1_1 && !icons_1_1.done && (_a = icons_1.return)) _a.call(icons_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var e_1, _a;
    };
    /**
     * Returns the ExpressionUI array for a given column.
     * @param {?} columnId
     * @return {?}
     */
    IgxFilteringService.prototype.getExpressions = /**
     * Returns the ExpressionUI array for a given column.
     * @param {?} columnId
     * @return {?}
     */
    function (columnId) {
        if (!this.columnToExpressionsMap.has(columnId)) {
            var /** @type {?} */ column = this.grid.columns.find(function (col) { return col.field === columnId; });
            var /** @type {?} */ expressionUIs = new Array();
            this.generateExpressionsList(column.filteringExpressionsTree, this.grid.filteringExpressionsTree.operator, expressionUIs);
            this.columnToExpressionsMap.set(columnId, expressionUIs);
            return expressionUIs;
        }
        return this.columnToExpressionsMap.get(columnId);
    };
    /**
     * Recreates all ExpressionUIs for all columns. Executed after filtering to refresh the cache.
     * @return {?}
     */
    IgxFilteringService.prototype.refreshExpressions = /**
     * Recreates all ExpressionUIs for all columns. Executed after filtering to refresh the cache.
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.isFiltering) {
            this.columnsWithComplexFilter.clear();
            this.columnToExpressionsMap.forEach(function (value, key) {
                var /** @type {?} */ column = _this.grid.columns.find(function (col) { return col.field === key; });
                value.length = 0;
                _this.generateExpressionsList(column.filteringExpressionsTree, _this.grid.filteringExpressionsTree.operator, value);
                var /** @type {?} */ isComplex = _this.isFilteringTreeComplex(column.filteringExpressionsTree);
                if (isComplex) {
                    _this.columnsWithComplexFilter.add(key);
                }
                _this.updateFilteringCell(key);
            });
        }
    };
    /**
     * Remove an ExpressionUI for a given column.
     * @param {?} columnId
     * @param {?} indexToRemove
     * @return {?}
     */
    IgxFilteringService.prototype.removeExpression = /**
     * Remove an ExpressionUI for a given column.
     * @param {?} columnId
     * @param {?} indexToRemove
     * @return {?}
     */
    function (columnId, indexToRemove) {
        var /** @type {?} */ expressionsList = this.getExpressions(columnId);
        if (indexToRemove === 0 && expressionsList.length > 1) {
            expressionsList[1].beforeOperator = null;
        }
        else if (indexToRemove === expressionsList.length - 1) {
            expressionsList[indexToRemove - 1].afterOperator = null;
        }
        else {
            expressionsList[indexToRemove - 1].afterOperator = expressionsList[indexToRemove + 1].beforeOperator;
            expressionsList[0].beforeOperator = null;
            expressionsList[expressionsList.length - 1].afterOperator = null;
        }
        expressionsList.splice(indexToRemove, 1);
    };
    /**
     * Generate filtering tree for a given column from existing ExpressionUIs.
     * @param {?} columnId
     * @return {?}
     */
    IgxFilteringService.prototype.createSimpleFilteringTree = /**
     * Generate filtering tree for a given column from existing ExpressionUIs.
     * @param {?} columnId
     * @return {?}
     */
    function (columnId) {
        var /** @type {?} */ expressionsList = this.getExpressions(columnId);
        var /** @type {?} */ expressionsTree = new FilteringExpressionsTree(FilteringLogic.Or, columnId);
        var /** @type {?} */ currAndBranch;
        var /** @type {?} */ currExpressionUI;
        for (var /** @type {?} */ i = 0; i < expressionsList.length; i++) {
            currExpressionUI = expressionsList[i];
            if ((currExpressionUI.beforeOperator === undefined || currExpressionUI.beforeOperator === null ||
                currExpressionUI.beforeOperator === FilteringLogic.Or) &&
                currExpressionUI.afterOperator === FilteringLogic.And) {
                currAndBranch = new FilteringExpressionsTree(FilteringLogic.And, columnId);
                expressionsTree.filteringOperands.push(currAndBranch);
                currAndBranch.filteringOperands.push(currExpressionUI.expression);
            }
            else if (currExpressionUI.beforeOperator === FilteringLogic.And) {
                currAndBranch.filteringOperands.push(currExpressionUI.expression);
            }
            else {
                expressionsTree.filteringOperands.push(currExpressionUI.expression);
                currAndBranch = null;
            }
        }
        return expressionsTree;
    };
    /**
     * Returns whether a complex filter is applied to a given column.
     * @param {?} columnId
     * @return {?}
     */
    IgxFilteringService.prototype.isFilterComplex = /**
     * Returns whether a complex filter is applied to a given column.
     * @param {?} columnId
     * @return {?}
     */
    function (columnId) {
        if (this.columnsWithComplexFilter.has(columnId)) {
            return true;
        }
        var /** @type {?} */ column = this.grid.columns.find(function (col) { return col.field === columnId; });
        var /** @type {?} */ isComplex = this.isFilteringTreeComplex(column.filteringExpressionsTree);
        if (isComplex) {
            this.columnsWithComplexFilter.add(columnId);
        }
        return isComplex;
    };
    /**
     * Returns the string representation of the FilteringLogic operator.
     * @param {?} operator
     * @return {?}
     */
    IgxFilteringService.prototype.getOperatorAsString = /**
     * Returns the string representation of the FilteringLogic operator.
     * @param {?} operator
     * @return {?}
     */
    function (operator) {
        return FilteringLogic[operator];
    };
    /**
     * Genererate the label of a chip from a given filtering expression.
     * @param {?} expression
     * @return {?}
     */
    IgxFilteringService.prototype.getChipLabel = /**
     * Genererate the label of a chip from a given filtering expression.
     * @param {?} expression
     * @return {?}
     */
    function (expression) {
        if (expression.condition.isUnary) {
            return this.titlecasePipe.transform(this.filterPipe.transform(expression.condition.name));
        }
        else if (expression.searchVal instanceof Date) {
            return this.datePipe.transform(expression.searchVal);
        }
        else {
            return expression.searchVal;
        }
    };
    /**
     * @param {?} columnId
     * @return {?}
     */
    IgxFilteringService.prototype.updateFilteringCell = /**
     * @param {?} columnId
     * @return {?}
     */
    function (columnId) {
        var /** @type {?} */ filterCell = this.grid.filterCellList.find(function (cell) { return cell.column.field === columnId; });
        if (filterCell) {
            filterCell.updateFilterCellArea();
        }
    };
    /**
     * @param {?} expressions
     * @return {?}
     */
    IgxFilteringService.prototype.isFilteringTreeComplex = /**
     * @param {?} expressions
     * @return {?}
     */
    function (expressions) {
        if (!expressions) {
            return false;
        }
        if (expressions instanceof FilteringExpressionsTree) {
            var /** @type {?} */ expressionsTree = /** @type {?} */ (expressions);
            if (expressionsTree.operator === FilteringLogic.Or) {
                var /** @type {?} */ andOperatorsCount = this.getChildAndOperatorsCount(expressionsTree);
                // having more that 'And' and operator in the sub-tree means that the filter could not be represented without parentheses.
                return andOperatorsCount > 1;
            }
            var /** @type {?} */ isComplex = false;
            for (var /** @type {?} */ i = 0; i < expressionsTree.filteringOperands.length; i++) {
                isComplex = isComplex || this.isFilteringTreeComplex(expressionsTree.filteringOperands[i]);
            }
            return isComplex;
        }
        return false;
    };
    /**
     * @param {?} expressions
     * @return {?}
     */
    IgxFilteringService.prototype.getChildAndOperatorsCount = /**
     * @param {?} expressions
     * @return {?}
     */
    function (expressions) {
        var /** @type {?} */ count = 0;
        var /** @type {?} */ operand;
        for (var /** @type {?} */ i = 0; i < expressions.filteringOperands.length; i++) {
            operand = expressions[i];
            if (operand instanceof FilteringExpressionsTree) {
                if (operand.operator === FilteringLogic.And) {
                    count++;
                }
                count = count + this.getChildAndOperatorsCount(operand);
            }
        }
        return count;
    };
    /**
     * @param {?} expressions
     * @param {?} operator
     * @param {?} expressionsUIs
     * @return {?}
     */
    IgxFilteringService.prototype.generateExpressionsList = /**
     * @param {?} expressions
     * @param {?} operator
     * @param {?} expressionsUIs
     * @return {?}
     */
    function (expressions, operator, expressionsUIs) {
        if (!expressions) {
            return;
        }
        if (expressions instanceof FilteringExpressionsTree) {
            var /** @type {?} */ expressionsTree = /** @type {?} */ (expressions);
            for (var /** @type {?} */ i = 0; i < expressionsTree.filteringOperands.length; i++) {
                this.generateExpressionsList(expressionsTree.filteringOperands[i], expressionsTree.operator, expressionsUIs);
            }
        }
        else {
            var /** @type {?} */ exprUI = new ExpressionUI();
            exprUI.expression = /** @type {?} */ (expressions);
            if (expressionsUIs.length !== 0) {
                exprUI.beforeOperator = operator;
            }
            var /** @type {?} */ prevExprUI = expressionsUIs[expressionsUIs.length - 1];
            if (prevExprUI) {
                prevExprUI.afterOperator = operator;
            }
            expressionsUIs.push(exprUI);
        }
    };
    IgxFilteringService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    IgxFilteringService.ctorParameters = function () { return [
        { type: GridBaseAPIService, },
        { type: IgxIconService, },
    ]; };
    return IgxFilteringService;
}());
export { IgxFilteringService };
function IgxFilteringService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxFilteringService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxFilteringService.ctorParameters;
    /** @type {?} */
    IgxFilteringService.prototype.columnsWithComplexFilter;
    /** @type {?} */
    IgxFilteringService.prototype.areEventsSubscribed;
    /** @type {?} */
    IgxFilteringService.prototype.destroy$;
    /** @type {?} */
    IgxFilteringService.prototype.isFiltering;
    /** @type {?} */
    IgxFilteringService.prototype.columnToExpressionsMap;
    /** @type {?} */
    IgxFilteringService.prototype.filterPipe;
    /** @type {?} */
    IgxFilteringService.prototype.titlecasePipe;
    /** @type {?} */
    IgxFilteringService.prototype.datePipe;
    /** @type {?} */
    IgxFilteringService.prototype.gridId;
    /** @type {?} */
    IgxFilteringService.prototype.isFilterRowVisible;
    /** @type {?} */
    IgxFilteringService.prototype.filteredColumn;
    /** @type {?} */
    IgxFilteringService.prototype.selectedExpression;
    /** @type {?} */
    IgxFilteringService.prototype.columnToChipToFocus;
    /** @type {?} */
    IgxFilteringService.prototype.columnToMoreIconHidden;
    /** @type {?} */
    IgxFilteringService.prototype.columnStartIndex;
    /** @type {?} */
    IgxFilteringService.prototype.gridAPI;
    /** @type {?} */
    IgxFilteringService.prototype.iconService;
}
