// igniteui-angular/esm2015/lib/services/excel/excel-files.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ExcelStrings } from './excel-strings';
/**
 * @hidden
 */
export class RootRelsFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('.rels', ExcelStrings.getRels());
    }
}
/**
 * @hidden
 */
export class AppFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('app.xml', ExcelStrings.getApp());
    }
}
/**
 * @hidden
 */
export class CoreFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('core.xml', ExcelStrings.getCore());
    }
}
/**
 * @hidden
 */
export class WorkbookRelsFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        const /** @type {?} */ hasSharedStrings = worksheetData.isEmpty === false;
        folder.file('workbook.xml.rels', ExcelStrings.getWorkbookRels(hasSharedStrings));
    }
}
/**
 * @hidden
 */
export class ThemeFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('theme1.xml', ExcelStrings.getTheme());
    }
}
/**
 * @hidden
 */
export class WorksheetFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        const /** @type {?} */ sheetData = [];
        const /** @type {?} */ cols = [];
        let /** @type {?} */ dimension;
        const /** @type {?} */ dictionary = worksheetData.dataDictionary;
        let /** @type {?} */ freezePane = '';
        if (worksheetData.isEmpty) {
            sheetData.push('<sheetData/>');
            dimension = 'A1';
        }
        else {
            sheetData.push('<sheetData>');
            const /** @type {?} */ height = worksheetData.options.rowHeight;
            const /** @type {?} */ rowHeight = height ? ' ht="' + height + '" customHeight="1"' : '';
            sheetData.push(`<row r="1"${rowHeight}>`);
            for (let /** @type {?} */ i = 0; i < worksheetData.columnCount; i++) {
                const /** @type {?} */ column = ExcelStrings.getExcelColumn(i) + 1;
                const /** @type {?} */ value = dictionary.saveValue(worksheetData.keys[i], i, true);
                sheetData.push(`<c r="${column}" t="s"><v>${value}</v></c>`);
            }
            sheetData.push('</row>');
            for (let /** @type {?} */ i = 1; i < worksheetData.rowCount; i++) {
                sheetData.push(`<row r="${(i + 1)}"${rowHeight}>`);
                for (let /** @type {?} */ j = 0; j < worksheetData.columnCount; j++) {
                    const /** @type {?} */ cellData = WorksheetFile.getCellData(worksheetData, i, j);
                    sheetData.push(cellData);
                }
                sheetData.push('</row>');
            }
            sheetData.push('</sheetData>');
            dimension = 'A1:' + ExcelStrings.getExcelColumn(worksheetData.columnCount - 1) + worksheetData.rowCount;
            cols.push('<cols>');
            for (let /** @type {?} */ i = 0; i < worksheetData.columnCount; i++) {
                const /** @type {?} */ width = dictionary.columnWidths[i];
                // Use the width provided in the options if it exists
                const /** @type {?} */ widthInTwips = worksheetData.options.columnWidth ?
                    worksheetData.options.columnWidth :
                    Math.max(((width / 96) * 14.4), WorksheetFile.MIN_WIDTH);
                cols.push(`<col min="${(i + 1)}" max="${(i + 1)}" width="${widthInTwips}" customWidth="1"/>`);
            }
            cols.push('</cols>');
            if (worksheetData.indexOfLastPinnedColumn !== -1 &&
                !worksheetData.options.ignorePinning &&
                !worksheetData.options.ignoreColumnsOrder) {
                const /** @type {?} */ frozenColumnCount = worksheetData.indexOfLastPinnedColumn + 1;
                const /** @type {?} */ firstCell = ExcelStrings.getExcelColumn(frozenColumnCount) + '1';
                freezePane = `<pane xSplit="${frozenColumnCount}" topLeftCell="${firstCell}" activePane="topRight" state="frozen"/>`;
            }
        }
        const /** @type {?} */ hasTable = !worksheetData.isEmpty && worksheetData.options.exportAsTable;
        folder.file('sheet1.xml', ExcelStrings.getSheetXML(dimension, freezePane, cols.join(''), sheetData.join(''), hasTable));
    }
    /**
     * @param {?} worksheetData
     * @param {?} row
     * @param {?} column
     * @return {?}
     */
    static getCellData(worksheetData, row, column) {
        const /** @type {?} */ dictionary = worksheetData.dataDictionary;
        const /** @type {?} */ columnName = ExcelStrings.getExcelColumn(column) + (row + 1);
        const /** @type {?} */ columnHeader = worksheetData.keys[column];
        const /** @type {?} */ cellValue = worksheetData.isSpecialData ?
            worksheetData.data[row - 1] :
            worksheetData.data[row - 1][columnHeader];
        if (cellValue === undefined || cellValue === null) {
            return `<c r="${columnName}" s="1"/>`;
        }
        else {
            const /** @type {?} */ savedValue = dictionary.saveValue(cellValue, column, false);
            const /** @type {?} */ isSavedAsString = savedValue !== -1;
            const /** @type {?} */ value = isSavedAsString ? savedValue : cellValue;
            const /** @type {?} */ type = isSavedAsString ? ` t="s"` : '';
            const /** @type {?} */ format = isSavedAsString ? '' : ` s="1"`;
            return `<c r="${columnName}"${type}${format}><v>${value}</v></c>`;
        }
    }
}
WorksheetFile.MIN_WIDTH = 8.34;
function WorksheetFile_tsickle_Closure_declarations() {
    /** @type {?} */
    WorksheetFile.MIN_WIDTH;
}
/**
 * @hidden
 */
export class StyleFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('styles.xml', ExcelStrings.getStyles(worksheetData.dataDictionary && worksheetData.dataDictionary.hasNonStringValues));
    }
}
/**
 * @hidden
 */
export class WorkbookFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('workbook.xml', ExcelStrings.getWorkbook());
    }
}
/**
 * @hidden
 */
export class ContentTypesFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('[Content_Types].xml', ExcelStrings.getContentTypesXML(!worksheetData.isEmpty, worksheetData.options.exportAsTable));
    }
}
/**
 * @hidden
 */
export class SharedStringsFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        const /** @type {?} */ dict = worksheetData.dataDictionary;
        const /** @type {?} */ sortedValues = dict.getKeys();
        const /** @type {?} */ sharedStrings = new Array(sortedValues.length);
        for (const /** @type {?} */ value of sortedValues) {
            sharedStrings[dict.getSanitizedValue(value)] = '<si><t>' + value + '</t></si>';
        }
        folder.file('sharedStrings.xml', ExcelStrings.getSharedStringXML(dict.stringsCount, sortedValues.length, sharedStrings.join('')));
    }
}
/**
 * @hidden
 */
export class TablesFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        const /** @type {?} */ columnCount = worksheetData.columnCount;
        const /** @type {?} */ lastColumn = ExcelStrings.getExcelColumn(columnCount - 1) + worksheetData.rowCount;
        const /** @type {?} */ dimension = 'A1:' + lastColumn;
        const /** @type {?} */ values = worksheetData.keys;
        let /** @type {?} */ sortString = '';
        let /** @type {?} */ tableColumns = '<tableColumns count="' + columnCount + '">';
        for (let /** @type {?} */ i = 0; i < columnCount; i++) {
            const /** @type {?} */ value = values[i];
            tableColumns += '<tableColumn id="' + (i + 1) + '" name="' + value + '"/>';
        }
        tableColumns += '</tableColumns>';
        if (worksheetData.sort) {
            const /** @type {?} */ sortingExpression = worksheetData.sort;
            const /** @type {?} */ sc = ExcelStrings.getExcelColumn(values.indexOf(sortingExpression.fieldName));
            const /** @type {?} */ dir = sortingExpression.dir - 1;
            sortString = `<sortState ref="A2:${lastColumn}"><sortCondition descending="${dir}" ref="${sc}1:${sc}15"/></sortState>`;
        }
        folder.file('table1.xml', ExcelStrings.getTablesXML(dimension, tableColumns, sortString));
    }
}
/**
 * @hidden
 */
export class WorksheetRelsFile {
    /**
     * @param {?} folder
     * @param {?} worksheetData
     * @return {?}
     */
    writeElement(folder, worksheetData) {
        folder.file('sheet1.xml.rels', ExcelStrings.getWorksheetRels());
    }
}
