// igniteui-angular/esm2015/lib/services/excel/worksheet-data-dictionary.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ExportUtilities } from '../exporter-common/export-utilities';
/**
 * @hidden
 */
export class WorksheetDataDictionary {
    /**
     * @param {?} columnCount
     * @param {?} columnWidth
     */
    constructor(columnCount, columnWidth) {
        this.hasNonStringValues = false;
        this._dictionary = {};
        this._widthsDictionary = {};
        this._counter = 0;
        this.dirtyKeyCollections();
        this._calculateColumnWidth = !columnWidth;
        this._columnWidths = new Array(columnCount);
        this._columnTypeInfo = new Array(columnCount);
        if (!this._calculateColumnWidth) {
            this._columnWidths.fill(columnWidth);
        }
        this.stringsCount = 0;
    }
    /**
     * @return {?}
     */
    get columnWidths() {
        return this._columnWidths;
    }
    /**
     * @param {?} value
     * @param {?} column
     * @param {?} isHeader
     * @return {?}
     */
    saveValue(value, column, isHeader) {
        if (this._columnTypeInfo[column] === undefined && isHeader === false) {
            this._columnTypeInfo[column] = typeof value === 'string' ||
                typeof value === 'boolean' ||
                value instanceof Date;
        }
        let /** @type {?} */ sanitizedValue = '';
        const /** @type {?} */ isSavedAsString = this._columnTypeInfo[column] || isHeader;
        if (isSavedAsString) {
            sanitizedValue = this.sanitizeValue(value);
            if (this._dictionary[sanitizedValue] === undefined) {
                this._dictionary[sanitizedValue] = this._counter++;
                this.dirtyKeyCollections();
            }
            this.stringsCount++;
        }
        else {
            this.hasNonStringValues = true;
        }
        if (this._calculateColumnWidth) {
            const /** @type {?} */ width = this.getTextWidth(value);
            const /** @type {?} */ maxWidth = Math.max(this._columnWidths[column] || 0, width);
            this._columnWidths[column] = maxWidth;
        }
        return isSavedAsString ? this.getSanitizedValue(sanitizedValue) : -1;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    getValue(value) {
        return this.getSanitizedValue(this.sanitizeValue(value));
    }
    /**
     * @param {?} sanitizedValue
     * @return {?}
     */
    getSanitizedValue(sanitizedValue) {
        return this._dictionary[sanitizedValue];
    }
    /**
     * @return {?}
     */
    getKeys() {
        if (!this._keysAreValid) {
            this._keys = Object.keys(this._dictionary);
            this._keysAreValid = true;
        }
        return this._keys;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    getTextWidth(value) {
        if (this._widthsDictionary[value] === undefined) {
            const /** @type {?} */ context = this.getContext();
            const /** @type {?} */ metrics = context.measureText(value);
            this._widthsDictionary[value] = metrics.width + WorksheetDataDictionary.TEXT_PADDING;
        }
        return this._widthsDictionary[value];
    }
    /**
     * @return {?}
     */
    getContext() {
        if (!this._context) {
            const /** @type {?} */ canvas = document.createElement('canvas');
            this._context = canvas.getContext('2d');
            this._context.font = WorksheetDataDictionary.DEFAULT_FONT;
        }
        return this._context;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    sanitizeValue(value) {
        if (ExportUtilities.hasValue(value) === false) {
            return '';
        }
        else {
            const /** @type {?} */ stringValue = String(value);
            return stringValue.replace(/&/g, '&amp;')
                .replace(/</g, '&lt;')
                .replace(/>/g, '&gt;')
                .replace(/"/g, '&quot;')
                .replace(/'/g, '&apos;');
        }
    }
    /**
     * @return {?}
     */
    dirtyKeyCollections() {
        this._keysAreValid = false;
    }
}
WorksheetDataDictionary.DEFAULT_FONT = '11pt Calibri';
WorksheetDataDictionary.TEXT_PADDING = 5;
function WorksheetDataDictionary_tsickle_Closure_declarations() {
    /** @type {?} */
    WorksheetDataDictionary.DEFAULT_FONT;
    /** @type {?} */
    WorksheetDataDictionary.TEXT_PADDING;
    /** @type {?} */
    WorksheetDataDictionary.prototype._dictionary;
    /** @type {?} */
    WorksheetDataDictionary.prototype._widthsDictionary;
    /** @type {?} */
    WorksheetDataDictionary.prototype._sortedKeysByValue;
    /** @type {?} */
    WorksheetDataDictionary.prototype._sortedKeysByValueAreValid;
    /** @type {?} */
    WorksheetDataDictionary.prototype._keys;
    /** @type {?} */
    WorksheetDataDictionary.prototype._keysAreValid;
    /** @type {?} */
    WorksheetDataDictionary.prototype._counter;
    /** @type {?} */
    WorksheetDataDictionary.prototype._calculateColumnWidth;
    /** @type {?} */
    WorksheetDataDictionary.prototype._columnWidths;
    /** @type {?} */
    WorksheetDataDictionary.prototype._context;
    /** @type {?} */
    WorksheetDataDictionary.prototype._columnTypeInfo;
    /** @type {?} */
    WorksheetDataDictionary.prototype.hasNonStringValues;
    /** @type {?} */
    WorksheetDataDictionary.prototype.stringsCount;
}
