// igniteui-angular/esm2015/lib/services/excel/worksheet-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ExportUtilities } from '../exporter-common/export-utilities';
import { WorksheetDataDictionary } from './worksheet-data-dictionary';
/**
 * @hidden
 */
export class WorksheetData {
    /**
     * @param {?} _data
     * @param {?} options
     * @param {?} indexOfLastPinnedColumn
     * @param {?} sort
     */
    constructor(_data, options, indexOfLastPinnedColumn, sort) {
        this._data = _data;
        this.options = options;
        this.indexOfLastPinnedColumn = indexOfLastPinnedColumn;
        this.sort = sort;
        this.initializeData();
    }
    /**
     * @return {?}
     */
    get data() {
        return this._data;
    }
    /**
     * @return {?}
     */
    get columnCount() {
        return this._columnCount;
    }
    /**
     * @return {?}
     */
    get rowCount() {
        return this._rowCount;
    }
    /**
     * @return {?}
     */
    get isEmpty() {
        return !this.rowCount || !this._columnCount;
    }
    /**
     * @return {?}
     */
    get keys() {
        return this._keys;
    }
    /**
     * @return {?}
     */
    get isSpecialData() {
        return this._isSpecialData;
    }
    /**
     * @return {?}
     */
    get dataDictionary() {
        return this._dataDictionary;
    }
    /**
     * @return {?}
     */
    initializeData() {
        if (!this._data || this._data.length === 0) {
            return;
        }
        this._keys = ExportUtilities.getKeysFromData(this._data);
        if (this._keys.length === 0) {
            return;
        }
        this._isSpecialData = ExportUtilities.isSpecialData(this._data);
        this._columnCount = this._keys.length;
        this._rowCount = this._data.length + 1;
        this._dataDictionary = new WorksheetDataDictionary(this._columnCount, this.options.columnWidth);
    }
}
function WorksheetData_tsickle_Closure_declarations() {
    /** @type {?} */
    WorksheetData.prototype._columnCount;
    /** @type {?} */
    WorksheetData.prototype._rowCount;
    /** @type {?} */
    WorksheetData.prototype._dataDictionary;
    /** @type {?} */
    WorksheetData.prototype._keys;
    /** @type {?} */
    WorksheetData.prototype._isSpecialData;
    /** @type {?} */
    WorksheetData.prototype._data;
    /** @type {?} */
    WorksheetData.prototype.options;
    /** @type {?} */
    WorksheetData.prototype.indexOfLastPinnedColumn;
    /** @type {?} */
    WorksheetData.prototype.sort;
}
