// igniteui-angular/esm2015/lib/services/exporter-common/export-utilities.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @hidden
 */
export class ExportUtilities {
    /**
     * @param {?} data
     * @return {?}
     */
    static getKeysFromData(data) {
        const /** @type {?} */ length = data.length;
        if (length === 0) {
            return [];
        }
        const /** @type {?} */ dataEntry = data[0];
        const /** @type {?} */ dataEntryMiddle = data[Math.floor(length / 2)];
        const /** @type {?} */ dataEntryLast = data[length - 1];
        const /** @type {?} */ keys1 = Object.keys(dataEntry);
        const /** @type {?} */ keys2 = Object.keys(dataEntryMiddle);
        const /** @type {?} */ keys3 = Object.keys(dataEntryLast);
        const /** @type {?} */ keys = new Set(keys1.concat(keys2).concat(keys3));
        return !ExportUtilities.isSpecialData(data) ? Array.from(keys) : ['Column 1'];
    }
    /**
     * @param {?} blob
     * @param {?} fileName
     * @return {?}
     */
    static saveBlobToFile(blob, fileName) {
        const /** @type {?} */ a = document.createElement('a');
        if (window.navigator && window.navigator.msSaveOrOpenBlob) {
            window.navigator.msSaveOrOpenBlob(blob, fileName);
        }
        else {
            const /** @type {?} */ url = window.URL.createObjectURL(blob);
            a.download = fileName;
            a.href = url;
            document.body.appendChild(a);
            a.click();
            document.body.removeChild(a);
            window.URL.revokeObjectURL(url);
        }
    }
    /**
     * @param {?} s
     * @return {?}
     */
    static stringToArrayBuffer(s) {
        const /** @type {?} */ buf = new ArrayBuffer(s.length);
        const /** @type {?} */ view = new Uint8Array(buf);
        for (let /** @type {?} */ i = 0; i !== s.length; ++i) {
            /* tslint:disable no-bitwise */
            view[i] = s.charCodeAt(i) & 0xFF;
            /* tslint:enable no-bitwise */
        }
        return buf;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    static isSpecialData(data) {
        const /** @type {?} */ dataEntry = data[0];
        return (typeof dataEntry === 'string' ||
            typeof dataEntry === 'number' ||
            dataEntry instanceof Date);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    static hasValue(value) {
        return value !== undefined && value !== null;
    }
}
