// igniteui-angular/esm5/lib/services/overlay/scroll/block-scroll-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var BlockScrollStrategy = /** @class */ (function () {
    function BlockScrollStrategy(scrollContainer) {
        var _this = this;
        this._initialized = false;
        this.onScroll = function (ev) {
            ev.preventDefault();
            if (!_this._sourceElement || _this._sourceElement !== ev.srcElement) {
                _this._sourceElement = ev.srcElement;
                _this._initialScrollTop = _this._sourceElement.scrollTop;
                _this._initialScrollLeft = _this._sourceElement.scrollLeft;
            }
            _this._sourceElement.scrollTop = _this._initialScrollTop;
            _this._sourceElement.scrollLeft = _this._initialScrollLeft;
        };
    }
    /**
     * @param {?} document
     * @param {?} overlayService
     * @param {?} id
     * @return {?}
     */
    BlockScrollStrategy.prototype.initialize = /**
     * @param {?} document
     * @param {?} overlayService
     * @param {?} id
     * @return {?}
     */
    function (document, overlayService, id) {
        if (this._initialized) {
            return;
        }
        this._document = document;
        this._initialized = true;
    };
    /**
     * @return {?}
     */
    BlockScrollStrategy.prototype.attach = /**
     * @return {?}
     */
    function () {
        this._document.addEventListener('scroll', this.onScroll, true);
        this._document.addEventListener('wheel', this.onWheel, true);
    };
    /**
     * @return {?}
     */
    BlockScrollStrategy.prototype.detach = /**
     * @return {?}
     */
    function () {
        this._document.removeEventListener('scroll', this.onScroll, true);
        this._document.removeEventListener('wheel', this.onWheel, true);
        this._sourceElement = null;
        this._initialScrollTop = 0;
        this._initialScrollLeft = 0;
        this._initialized = false;
    };
    /**
     * @param {?} ev
     * @return {?}
     */
    BlockScrollStrategy.prototype.onWheel = /**
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        ev.stopImmediatePropagation();
        ev.preventDefault();
    };
    return BlockScrollStrategy;
}());
export { BlockScrollStrategy };
function BlockScrollStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    BlockScrollStrategy.prototype._initialized;
    /** @type {?} */
    BlockScrollStrategy.prototype._document;
    /** @type {?} */
    BlockScrollStrategy.prototype._initialScrollTop;
    /** @type {?} */
    BlockScrollStrategy.prototype._initialScrollLeft;
    /** @type {?} */
    BlockScrollStrategy.prototype._sourceElement;
    /** @type {?} */
    BlockScrollStrategy.prototype.onScroll;
}
