// igniteui-angular/esm5/lib/list/list.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, TemplateRef } from '@angular/core';
/**
 * @record
 */
export function IListChild() { }
function IListChild_tsickle_Closure_declarations() {
    /** @type {?} */
    IListChild.prototype.index;
}
/** @enum {number} */
var IgxListPanState = { NONE: 0, LEFT: 1, RIGHT: 2, };
export { IgxListPanState };
IgxListPanState[IgxListPanState.NONE] = "NONE";
IgxListPanState[IgxListPanState.LEFT] = "LEFT";
IgxListPanState[IgxListPanState.RIGHT] = "RIGHT";
var IgxEmptyListTemplateDirective = /** @class */ (function () {
    function IgxEmptyListTemplateDirective(template) {
        this.template = template;
    }
    IgxEmptyListTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxEmptyList]'
                },] },
    ];
    /** @nocollapse */
    IgxEmptyListTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxEmptyListTemplateDirective;
}());
export { IgxEmptyListTemplateDirective };
function IgxEmptyListTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxEmptyListTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxEmptyListTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxEmptyListTemplateDirective.prototype.template;
}
var IgxDataLoadingTemplateDirective = /** @class */ (function () {
    function IgxDataLoadingTemplateDirective(template) {
        this.template = template;
    }
    IgxDataLoadingTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxDataLoading]'
                },] },
    ];
    /** @nocollapse */
    IgxDataLoadingTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxDataLoadingTemplateDirective;
}());
export { IgxDataLoadingTemplateDirective };
function IgxDataLoadingTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxDataLoadingTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxDataLoadingTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxDataLoadingTemplateDirective.prototype.template;
}
var IgxListItemLeftPanningTemplateDirective = /** @class */ (function () {
    function IgxListItemLeftPanningTemplateDirective(template) {
        this.template = template;
    }
    IgxListItemLeftPanningTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxListItemLeftPanning]'
                },] },
    ];
    /** @nocollapse */
    IgxListItemLeftPanningTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxListItemLeftPanningTemplateDirective;
}());
export { IgxListItemLeftPanningTemplateDirective };
function IgxListItemLeftPanningTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxListItemLeftPanningTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxListItemLeftPanningTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxListItemLeftPanningTemplateDirective.prototype.template;
}
var IgxListItemRightPanningTemplateDirective = /** @class */ (function () {
    function IgxListItemRightPanningTemplateDirective(template) {
        this.template = template;
    }
    IgxListItemRightPanningTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxListItemRightPanning]'
                },] },
    ];
    /** @nocollapse */
    IgxListItemRightPanningTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
    ]; };
    return IgxListItemRightPanningTemplateDirective;
}());
export { IgxListItemRightPanningTemplateDirective };
function IgxListItemRightPanningTemplateDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxListItemRightPanningTemplateDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxListItemRightPanningTemplateDirective.ctorParameters;
    /** @type {?} */
    IgxListItemRightPanningTemplateDirective.prototype.template;
}
