// igniteui-angular/esm5/lib/data-operations/data-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { filteringStateDefaults } from './filtering-state.interface';
import { SortingStateDefaults } from './sorting-state.interface';
import { TreeGridSortingStrategy } from './sorting-strategy';
import { PagingError } from './paging-state.interface';
import { TransactionType } from '../services';
import { mergeObjects, cloneValue } from '../core/utils';
/** @enum {string} */
var DataType = {
    String: 'string',
    Number: 'number',
    Boolean: 'boolean',
    Date: 'date',
};
export { DataType };
/**
 * @hidden
 */
var /**
 * @hidden
 */
DataUtil = /** @class */ (function () {
    function DataUtil() {
    }
    /**
     * @param {?} target
     * @param {?} defaults
     * @return {?}
     */
    DataUtil.mergeDefaultProperties = /**
     * @param {?} target
     * @param {?} defaults
     * @return {?}
     */
    function (target, defaults) {
        if (!defaults) {
            return target;
        }
        if (!target) {
            target = Object.assign({}, defaults);
            return target;
        }
        Object
            .keys(defaults)
            .forEach(function (key) {
            if (target[key] === undefined && defaults[key] !== undefined) {
                target[key] = defaults[key];
            }
        });
        return target;
    };
    /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    DataUtil.sort = /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    function (data, state) {
        // set defaults
        DataUtil.mergeDefaultProperties(state, SortingStateDefaults);
        // apply default settings for each sorting expression(if not set)
        return state.strategy.sort(data, state.expressions);
    };
    /**
     * @param {?} hierarchicalData
     * @param {?} state
     * @param {?} parent
     * @return {?}
     */
    DataUtil.hierarchicalSort = /**
     * @param {?} hierarchicalData
     * @param {?} state
     * @param {?} parent
     * @return {?}
     */
    function (hierarchicalData, state, parent) {
        state.strategy = new TreeGridSortingStrategy();
        var /** @type {?} */ res = [];
        hierarchicalData.forEach(function (hr) {
            var /** @type {?} */ rec = DataUtil.cloneTreeGridRecord(hr);
            rec.parent = parent;
            if (rec.children) {
                rec.children = DataUtil.hierarchicalSort(rec.children, state, rec);
            }
            res.push(rec);
        });
        res = DataUtil.sort(res, state);
        return res;
    };
    /**
     * @param {?} hierarchicalRecord
     * @return {?}
     */
    DataUtil.cloneTreeGridRecord = /**
     * @param {?} hierarchicalRecord
     * @return {?}
     */
    function (hierarchicalRecord) {
        var /** @type {?} */ rec = {
            rowID: hierarchicalRecord.rowID,
            data: hierarchicalRecord.data,
            children: hierarchicalRecord.children,
            isFilteredOutParent: hierarchicalRecord.isFilteredOutParent,
            level: hierarchicalRecord.level,
            expanded: hierarchicalRecord.expanded,
            path: tslib_1.__spread(hierarchicalRecord.path)
        };
        return rec;
    };
    /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    DataUtil.group = /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    function (data, state) {
        // set defaults
        DataUtil.mergeDefaultProperties(state, SortingStateDefaults);
        // apply default settings for each grouping expression(if not set)
        return state.strategy.groupBy(data, state.expressions);
    };
    /**
     * @param {?} groupData
     * @param {?} state
     * @param {?=} groupsRecords
     * @return {?}
     */
    DataUtil.restoreGroups = /**
     * @param {?} groupData
     * @param {?} state
     * @param {?=} groupsRecords
     * @return {?}
     */
    function (groupData, state, groupsRecords) {
        if (groupsRecords === void 0) { groupsRecords = []; }
        DataUtil.mergeDefaultProperties(state, SortingStateDefaults);
        if (state.expressions.length === 0) {
            return groupData.data;
        }
        return this.restoreGroupsRecursive(groupData, 1, state.expressions.length, state.expansion, state.defaultExpanded, groupsRecords);
    };
    /**
     * @param {?} groupData
     * @param {?} level
     * @param {?} depth
     * @param {?} expansion
     * @param {?} defaultExpanded
     * @param {?} groupsRecords
     * @return {?}
     */
    DataUtil.restoreGroupsRecursive = /**
     * @param {?} groupData
     * @param {?} level
     * @param {?} depth
     * @param {?} expansion
     * @param {?} defaultExpanded
     * @param {?} groupsRecords
     * @return {?}
     */
    function (groupData, level, depth, expansion, defaultExpanded, groupsRecords) {
        var _this = this;
        var /** @type {?} */ i = 0;
        var /** @type {?} */ j;
        var /** @type {?} */ result = [];
        // empty the array without changing reference
        groupsRecords.splice(0, groupsRecords.length);
        if (level !== depth) {
            groupData.data = this.restoreGroupsRecursive(groupData, level + 1, depth, expansion, defaultExpanded, groupsRecords);
        }
        var _loop_1 = function () {
            var /** @type {?} */ g = level === depth ? groupData.metadata[i] :
                groupData.data[i].groupParent;
            for (j = i + 1; j < groupData.data.length; j++) {
                var /** @type {?} */ h = level === depth ? groupData.metadata[j] :
                    groupData.data[j].groupParent;
                if (h && g !== h && g.level === h.level) {
                    break;
                }
            }
            var /** @type {?} */ hierarchy = this_1.getHierarchy(g);
            var /** @type {?} */ expandState = expansion.find(function (state) {
                return _this.isHierarchyMatch(state.hierarchy || [{ fieldName: g.expression.fieldName, value: g.value }], hierarchy);
            });
            var /** @type {?} */ expanded = expandState ? expandState.expanded : defaultExpanded;
            result.push(g);
            groupsRecords.push(g);
            g['groups'] = groupData.data.slice(i, j).filter(function (e) {
                return e.records && e.records.length && e.level === g.level + 1;
            });
            while (groupsRecords.length) {
                if (groupsRecords[0].level + 1 > level) {
                    groupsRecords.shift();
                }
                else {
                    break;
                }
            }
            if (expanded) {
                result = result.concat(groupData.data.slice(i, j));
            }
            i = j;
        };
        var this_1 = this;
        while (i < groupData.data.length) {
            _loop_1();
        }
        return result;
    };
    /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    DataUtil.page = /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    function (data, state) {
        if (!state) {
            return data;
        }
        var /** @type {?} */ len = data.length;
        var /** @type {?} */ index = state.index;
        var /** @type {?} */ res = [];
        var /** @type {?} */ recordsPerPage = state.recordsPerPage;
        state.metadata = {
            countPages: 0,
            countRecords: data.length,
            error: PagingError.None
        };
        if (index < 0 || isNaN(index)) {
            state.metadata.error = PagingError.IncorrectPageIndex;
            return res;
        }
        if (recordsPerPage <= 0 || isNaN(recordsPerPage)) {
            state.metadata.error = PagingError.IncorrectRecordsPerPage;
            return res;
        }
        state.metadata.countPages = Math.ceil(len / recordsPerPage);
        if (!len) {
            return data;
        }
        if (index >= state.metadata.countPages) {
            state.metadata.error = PagingError.IncorrectPageIndex;
            return res;
        }
        return data.slice(index * recordsPerPage, (index + 1) * recordsPerPage);
    };
    /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    DataUtil.filter = /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    function (data, state) {
        // set defaults
        DataUtil.mergeDefaultProperties(state, filteringStateDefaults);
        if (!state.strategy) {
            return data;
        }
        return state.strategy.filter(data, state.expressionsTree);
    };
    /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    DataUtil.process = /**
     * @template T
     * @param {?} data
     * @param {?} state
     * @return {?}
     */
    function (data, state) {
        if (!state) {
            return data;
        }
        if (state.filtering) {
            data = DataUtil.filter(data, state.filtering);
        }
        if (state.sorting) {
            data = DataUtil.sort(data, state.sorting);
        }
        if (state.paging) {
            data = DataUtil.page(data, state.paging);
        }
        return data;
    };
    /**
     * @param {?} gRow
     * @return {?}
     */
    DataUtil.getHierarchy = /**
     * @param {?} gRow
     * @return {?}
     */
    function (gRow) {
        var /** @type {?} */ hierarchy = [];
        if (gRow !== undefined && gRow.expression) {
            hierarchy.push({ fieldName: gRow.expression.fieldName, value: gRow.value });
            while (gRow.groupParent) {
                gRow = gRow.groupParent;
                hierarchy.unshift({ fieldName: gRow.expression.fieldName, value: gRow.value });
            }
        }
        return hierarchy;
    };
    /**
     * @param {?} h1
     * @param {?} h2
     * @return {?}
     */
    DataUtil.isHierarchyMatch = /**
     * @param {?} h1
     * @param {?} h2
     * @return {?}
     */
    function (h1, h2) {
        if (h1.length !== h2.length) {
            return false;
        }
        return h1.every(function (level, index) {
            return level.fieldName === h2[index].fieldName && level.value === h2[index].value;
        });
    };
    /**
     * Merges all changes from provided transactions into provided data collection
     * @template T
     * @param {?} data Collection to merge
     * @param {?} transactions Transactions to merge into data
     * @param {?=} primaryKey Primary key of the collection, if any
     * @return {?}
     */
    DataUtil.mergeTransactions = /**
     * Merges all changes from provided transactions into provided data collection
     * @template T
     * @param {?} data Collection to merge
     * @param {?} transactions Transactions to merge into data
     * @param {?=} primaryKey Primary key of the collection, if any
     * @return {?}
     */
    function (data, transactions, primaryKey) {
        var _this = this;
        data.forEach(function (item, index) {
            var /** @type {?} */ rowId = primaryKey ? item[primaryKey] : item;
            var /** @type {?} */ transaction = transactions.find(function (t) { return t.id === rowId; });
            if (Array.isArray(item.children)) {
                _this.mergeTransactions(item.children, transactions, primaryKey);
            }
            if (transaction && transaction.type === TransactionType.UPDATE) {
                data[index] = transaction.newValue;
            }
        });
        data.push.apply(data, tslib_1.__spread(transactions
            .filter(function (t) { return t.type === TransactionType.ADD; })
            .map(function (t) { return t.newValue; })));
        return data;
    };
    /**
     * \@experimental \@hidden
     * @param {?} data
     * @param {?} transactions
     * @param {?} childDataKey
     * @param {?=} primaryKey
     * @param {?=} parentKey
     * @return {?}
     */
    DataUtil.mergeHierarchicalTransactions = /**
     * \@experimental \@hidden
     * @param {?} data
     * @param {?} transactions
     * @param {?} childDataKey
     * @param {?=} primaryKey
     * @param {?=} parentKey
     * @return {?}
     */
    function (data, transactions, childDataKey, primaryKey, parentKey) {
        var _loop_2 = function (index) {
            var /** @type {?} */ dataItem = data[index];
            var /** @type {?} */ rowId = primaryKey ? dataItem[primaryKey] : dataItem;
            var /** @type {?} */ updateTransaction = transactions.filter(function (t) { return t.type === TransactionType.UPDATE; }).find(function (t) { return t.id === rowId; });
            var /** @type {?} */ addedTransactions = transactions.filter(function (t) { return t.type === TransactionType.ADD; }).filter(function (t) { return t.parentId === rowId; });
            if (updateTransaction || addedTransactions.length > 0) {
                data[index] = mergeObjects(cloneValue(dataItem), updateTransaction && updateTransaction.newValue);
            }
            if (addedTransactions.length > 0) {
                if (!data[index][childDataKey]) {
                    data[index][childDataKey] = [];
                }
                try {
                    for (var addedTransactions_1 = tslib_1.__values(addedTransactions), addedTransactions_1_1 = addedTransactions_1.next(); !addedTransactions_1_1.done; addedTransactions_1_1 = addedTransactions_1.next()) {
                        var addedTransaction = addedTransactions_1_1.value;
                        data[index][childDataKey].push(addedTransaction.newValue);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (addedTransactions_1_1 && !addedTransactions_1_1.done && (_a = addedTransactions_1.return)) _a.call(addedTransactions_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            if (data[index][childDataKey]) {
                data[index][childDataKey] = this_2.mergeHierarchicalTransactions(data[index][childDataKey], transactions, childDataKey, primaryKey, rowId);
            }
            var e_1, _a;
        };
        var this_2 = this;
        for (var /** @type {?} */ index = 0; index < data.length; index++) {
            _loop_2(index);
        }
        return data;
    };
    return DataUtil;
}());
/**
 * @hidden
 */
export { DataUtil };
