// igniteui-angular/esm5/lib/data-operations/filtering-strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { FilteringLogic } from './filtering-expression.interface';
import { FilteringExpressionsTree } from './filtering-expressions-tree';
/**
 * @record
 */
export function IFilteringStrategy() { }
function IFilteringStrategy_tsickle_Closure_declarations() {
    /** @type {?} */
    IFilteringStrategy.prototype.filter;
}
/**
 * @abstract
 */
var /**
 * @abstract
 */
BaseFilteringStrategy = /** @class */ (function () {
    function BaseFilteringStrategy() {
    }
    /**
     * @param {?} rec
     * @param {?} expr
     * @return {?}
     */
    BaseFilteringStrategy.prototype.findMatchByExpression = /**
     * @param {?} rec
     * @param {?} expr
     * @return {?}
     */
    function (rec, expr) {
        var /** @type {?} */ cond = expr.condition;
        var /** @type {?} */ val = this.getFieldValue(rec, expr.fieldName);
        return cond.logic(val, expr.searchVal, expr.ignoreCase);
    };
    /**
     * @param {?} rec
     * @param {?} expressions
     * @return {?}
     */
    BaseFilteringStrategy.prototype.matchRecord = /**
     * @param {?} rec
     * @param {?} expressions
     * @return {?}
     */
    function (rec, expressions) {
        if (expressions) {
            if (expressions instanceof FilteringExpressionsTree) {
                var /** @type {?} */ expressionsTree = /** @type {?} */ (expressions);
                var /** @type {?} */ operator = /** @type {?} */ (expressionsTree.operator);
                var /** @type {?} */ match = void 0, /** @type {?} */ matchOperand = void 0, /** @type {?} */ operand = void 0;
                if (expressionsTree.filteringOperands) {
                    for (var /** @type {?} */ i = 0; i < expressionsTree.filteringOperands.length; i++) {
                        operand = expressionsTree.filteringOperands[i];
                        matchOperand = this.matchRecord(rec, operand);
                        if (match === undefined) {
                            match = matchOperand;
                        }
                        else if (operator === FilteringLogic.And) {
                            match = match && matchOperand;
                        }
                        else if (operator === FilteringLogic.Or) {
                            match = match || matchOperand;
                        }
                    }
                }
                return match === undefined ? true : match;
            }
            else {
                var /** @type {?} */ expression = /** @type {?} */ (expressions);
                return this.findMatchByExpression(rec, expression);
            }
        }
        return true;
    };
    return BaseFilteringStrategy;
}());
/**
 * @abstract
 */
export { BaseFilteringStrategy };
function BaseFilteringStrategy_tsickle_Closure_declarations() {
    /**
     * @abstract
     * @param {?} data
     * @param {?} expressionsTree
     * @return {?}
     */
    BaseFilteringStrategy.prototype.filter = function (data, expressionsTree) { };
    /**
     * @abstract
     * @param {?} rec
     * @param {?} fieldName
     * @return {?}
     */
    BaseFilteringStrategy.prototype.getFieldValue = function (rec, fieldName) { };
}
var FilteringStrategy = /** @class */ (function (_super) {
    tslib_1.__extends(FilteringStrategy, _super);
    function FilteringStrategy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @template T
     * @param {?} data
     * @param {?} expressionsTree
     * @return {?}
     */
    FilteringStrategy.prototype.filter = /**
     * @template T
     * @param {?} data
     * @param {?} expressionsTree
     * @return {?}
     */
    function (data, expressionsTree) {
        var /** @type {?} */ i;
        var /** @type {?} */ rec;
        var /** @type {?} */ len = data.length;
        var /** @type {?} */ res = [];
        if (!expressionsTree || !expressionsTree.filteringOperands || expressionsTree.filteringOperands.length === 0 || !len) {
            return data;
        }
        for (i = 0; i < len; i++) {
            rec = data[i];
            if (this.matchRecord(rec, expressionsTree)) {
                res.push(rec);
            }
        }
        return res;
    };
    /**
     * @param {?} rec
     * @param {?} fieldName
     * @return {?}
     */
    FilteringStrategy.prototype.getFieldValue = /**
     * @param {?} rec
     * @param {?} fieldName
     * @return {?}
     */
    function (rec, fieldName) {
        return rec[fieldName];
    };
    return FilteringStrategy;
}(BaseFilteringStrategy));
export { FilteringStrategy };
