// igniteui-angular/esm5/lib/calendar/calendar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { transition, trigger, useAnimation } from '@angular/animations';
import { Component, ContentChild, EventEmitter, forwardRef, HostBinding, HostListener, Input, Output, QueryList, ViewChildren } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { HAMMER_GESTURE_CONFIG, HammerGestureConfig } from '@angular/platform-browser';
import { fadeIn, scaleInCenter, slideInLeft, slideInRight } from '../animations/main';
import { Calendar, range } from './calendar';
import { IgxCalendarDateDirective, IgxCalendarHeaderTemplateDirective, IgxCalendarSubheaderTemplateDirective } from './calendar.directives';
import { DateRangeType } from '../core/dates/dateRange';
import { isDate } from 'util';
var /** @type {?} */ NEXT_ID = 0;
/** @enum {number} */
var CalendarView = {
    DEFAULT: 0,
    YEAR: 1,
    DECADE: 2,
};
export { CalendarView };
CalendarView[CalendarView.DEFAULT] = "DEFAULT";
CalendarView[CalendarView.YEAR] = "YEAR";
CalendarView[CalendarView.DECADE] = "DECADE";
/** @enum {string} */
var CalendarSelection = {
    SINGLE: 'single',
    MULTI: 'multi',
    RANGE: 'range',
};
export { CalendarSelection };
var CalendarHammerConfig = /** @class */ (function (_super) {
    tslib_1.__extends(CalendarHammerConfig, _super);
    function CalendarHammerConfig() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.overrides = {
            pan: { direction: Hammer.DIRECTION_VERTICAL, threshold: 1 }
        };
        return _this;
    }
    return CalendarHammerConfig;
}(HammerGestureConfig));
export { CalendarHammerConfig };
function CalendarHammerConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    CalendarHammerConfig.prototype.overrides;
}
/**
 * **Ignite UI for Angular Calendar** -
 * [Documentation](https://www.infragistics.com/products/ignite-ui-angular/angular/components/calendar.html)
 *
 * The Ignite UI Calendar provides an easy way to display a calendar and allow users to select dates using single, multiple
 * or range selection.
 *
 * Example:
 * ```html
 * <igx-calendar selection="range"></igx-calendar>
 * ```
 */
var IgxCalendarComponent = /** @class */ (function () {
    /**
     * @hidden
     */
    function IgxCalendarComponent() {
        /**
         * Sets/gets the `id` of the calendar.
         * If not set, the `id` will have value `"igx-calendar-0"`.
         * ```html
         * <igx-calendar id = "my-first-calendar"></igx-calendar>
         * ```
         * ```typescript
         * let calendarId =  this.calendar.id;
         * ```
         * \@memberof IgxCalendarComponent
         */
        this.id = "igx-calendar-" + NEXT_ID++;
        /**
         * Sets/gets whether the calendar header will be in vertical position.
         * Default value is `false`.
         * ```html
         * <igx-calendar [vertical] = "true"></igx-calendar>
         * ```
         * ```typescript
         * let isVertical = this.calendar.vertical;
         * ```
         */
        this.vertical = false;
        /**
         * Emits an event when a selection is made in the calendar.
         * Provides reference the `selectedDates` property in the `IgxCalendarComponent`.
         * ```html
         * <igx-calendar (onSelection) = "onSelection($event)"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         */
        this.onSelection = new EventEmitter();
        /**
         * The default `tabindex` attribute for the component.
         *
         * @hidden
         */
        this.tabindex = 0;
        /**
         * The default aria role attribute for the component.
         *
         * @hidden
         */
        this.role = 'grid';
        /**
         * The default aria lebelled by attribute for the component.
         *
         * @hidden
         */
        this.ariaLabelledBy = 'calendar';
        /**
         * @hidden
         */
        this._activeView = CalendarView.DEFAULT;
        /**
         * @hidden
         */
        this._selection = CalendarSelection.SINGLE;
        /**
         * @hidden
         */
        this.rangeStarted = false;
        /**
         * @hidden
         */
        this._monthAction = '';
        /**
         * @hidden
         */
        this._locale = 'en';
        /**
         * @hidden
         */
        this._formatOptions = {
            day: 'numeric',
            month: 'short',
            weekday: 'short',
            year: 'numeric'
        };
        /**
         * @hidden
         */
        this._formatViews = {
            day: false,
            month: true,
            year: false
        };
        /**
         * @hidden
         */
        this._disabledDates = null;
        /**
         * @hidden
         */
        this._specialDates = null;
        /**
         * @hidden
         */
        this._onTouchedCallback = function () { };
        /**
         * @hidden
         */
        this._onChangeCallback = function () { };
        this.calendarModel = new Calendar();
    }
    Object.defineProperty(IgxCalendarComponent.prototype, "weekStart", {
        get: /**
         * Gets the start day of the week.
         * Can return a numeric or an enum representation of the week day.
         * Defaults to `Sunday` / `0`.
         * ```typescript
         * let weekStart =  this.calendar.weekStart;
         * ```
         * \@memberof IgxCalendarComponent
         * @return {?}
         */
        function () {
            return this.calendarModel.firstWeekDay;
        },
        set: /**
         * Sets the start day of the week.
         * Can be assigned to a numeric value or to `WEEKDAYS` enum value.
         * ```html
         * <igx-calendar [weekStart] = "1"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.calendarModel.firstWeekDay = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "locale", {
        get: /**
         * Gets the `locale` of the calendar.
         * Default value is `"en"`.
         * ```typescript
         * let locale =  this.calendar.locale;
         * ```
         * \@memberof IgxCalendarComponent
         * @return {?}
         */
        function () {
            return this._locale;
        },
        set: /**
         * Sets the `locale` of the calendar.
         * Expects a valid BCP 47 language tag.
         * Default value is `"en"`.
         * ```html
         * <igx-calendar [locale] = "de"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._locale = value;
            this.initFormatters();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "selection", {
        get: /**
         *
         * Gets the selection type of the calendar.
         * Default value is `"single"`.
         * Changing the type of selection in the calendar resets the currently
         * selected values if any.
         * ```typescript
         * let selectionType =  this.calendar.selection;
         * ```
         * \@memberof IgxCalendarComponent
         * @return {?}
         */
        function () {
            return this._selection;
        },
        set: /**
         * Sets the selection type of the calendar.
         * ```html
         * <igx-calendar [selection] = "'multi'"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} value
         * @return {?}
         */
        function (value) {
            switch (value) {
                case 'single':
                    this.selectedDates = null;
                    break;
                case 'multi':
                case 'range':
                    this.selectedDates = [];
                    break;
                default:
                    throw new Error('Invalid selection value');
            }
            this._onChangeCallback(this.selectedDates);
            this.rangeStarted = false;
            this._selection = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "viewDate", {
        get: /**
         * Gets the date that is presented in the calendar.
         * By default it is the current date.
         * ```typescript
         * let date = this.calendar.viewDate;
         * ```
         * \@memberof IgxCalendarComponent
         * @return {?}
         */
        function () {
            return this._viewDate;
        },
        set: /**
         * Sets the date that will be presented in the default view when the calendar renders.
         * ```html
         * <igx-calendar viewDate = "15/06/2018"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._viewDate = this.getDateOnly(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "value", {
        get: /**
         * Gets the selected date(s) of the calendar.
         *
         * When the calendar selection is set to `single`, it returns
         * a single `Date` object.
         * Otherwise it is an array of `Date` objects.
         * ```typescript
         * let selectedDates =  this.calendar.value;
         * ```
         * \@memberof IgxCalendarComponent
         * @return {?}
         */
        function () {
            return this.selectedDates;
        },
        set: /**
         * Sets the selected date(s) of the calendar.
         *
         * When the calendar selection is set to `single`, it accepts
         * a single `Date` object.
         * Otherwise it is an array of `Date` objects.
         * ```typescript
         *  this.calendar.value =  new Date(`2016-06-12`);
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.selectDate(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "formatOptions", {
        get: /**
         * Gets the date format options of the calendar.
         * ```typescript
         * let dateFormatOptions = this.calendar.formatOptions.
         * ```
         * @return {?}
         */
        function () {
            return this._formatOptions;
        },
        set: /**
         * Sets the date format options of the calendar.
         * ```html
         * <igx-calendar> [formatOptions] = "{ day: '2-digit', month: 'short', weekday: 'long', year: 'numeric' }"</igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} formatOptions
         * @return {?}
         */
        function (formatOptions) {
            this._formatOptions = Object.assign(this._formatOptions, formatOptions);
            this.initFormatters();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "formatViews", {
        get: /**
         * Gets whether the `day`, `month` and `year` should be rendered
         * according to the locale and formatOptions, if any.
         * ```typescript
         * let formatViews = this.calendar.formatViews;
         * ```
         * @return {?}
         */
        function () {
            return this._formatViews;
        },
        set: /**
         * Gets whether the `day`, `month` and `year` should be rendered
         * according to the locale and formatOptions, if any.
         * ```html
         * <igx-calendar [formatViews] = "{ day: true, month: false, year: true }"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} formatViews
         * @return {?}
         */
        function (formatViews) {
            this._formatViews = Object.assign(this._formatViews, formatViews);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "disabledDates", {
        get: /**
         * Gets the disabled dates descriptors.
         * ```typescript
         * let disabledDates = this.calendar.disabledDates;
         * ```
         * @return {?}
         */
        function () {
            return this._disabledDates;
        },
        set: /**
         * Sets the disabled dates' descriptors.
         * ```typescript
         * \@ViewChild("MyCalendar")
         * public calendar: IgCalendarComponent;
         * ngOnInit(){
         *    this.calendar.disabledDates = [
         *      new DateRangeDescriptor(DateRangeType.Between, [new Date("2020-1-1"), new Date("2020-1-15")]),
         *      new DateRangeDescriptor(DateRangeType.Weekends)];
         * }
         * ```
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabledDates = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "specialDates", {
        get: /**
         * Gets the special dates descriptors.
         * ```typescript
         * let specialDates = this.calendar.specialDates;
         * ```
         * @return {?}
         */
        function () {
            return this._specialDates;
        },
        set: /**
         * Sets the special dates' descriptors.
         * ```typescript
         * \@ViewChild("MyCalendar")
         * public calendar: IgCalendarComponent;
         * ngOnInit(){
         *    this.calendar.specialDates = [
         *      new DateRangeDescriptor(DateRangeType.Between, [new Date("2020-1-1"), new Date("2020-1-15")]),
         *      new DateRangeDescriptor(DateRangeType.Weekends)];
         * }
         * ```
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._specialDates = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "styleClass", {
        get: /**
         * The default css class applied to the component.
         *
         * @hidden
         * @return {?}
         */
        function () {
            if (this.vertical) {
                return 'igx-calendar--vertical';
            }
            return 'igx-calendar';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "months", {
        /**
         * Returns an array of date objects which are then used to
         * properly render the month names.
         *
         * Used in the template of the component
         *
         * @hidden
         */
        get: /**
         * Returns an array of date objects which are then used to
         * properly render the month names.
         *
         * Used in the template of the component
         *
         * @hidden
         * @return {?}
         */
        function () {
            var /** @type {?} */ start = new Date(this._viewDate.getFullYear(), 0, 1);
            var /** @type {?} */ result = [];
            for (var /** @type {?} */ i = 0; i < 12; i++) {
                result.push(start);
                start = this.calendarModel.timedelta(start, 'month', 1);
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "decade", {
        /**
         * Returns an array of date objects which are then used to properly
         * render the years.
         *
         * Used in the template of the component.
         *
         * @hidden
         */
        get: /**
         * Returns an array of date objects which are then used to properly
         * render the years.
         *
         * Used in the template of the component.
         *
         * @hidden
         * @return {?}
         */
        function () {
            var /** @type {?} */ result = [];
            var /** @type {?} */ start = this._viewDate.getFullYear() - 3;
            var /** @type {?} */ end = this._viewDate.getFullYear() + 4;
            try {
                for (var _a = tslib_1.__values(range(start, end)), _b = _a.next(); !_b.done; _b = _a.next()) {
                    var year = _b.value;
                    result.push(new Date(year, this._viewDate.getMonth(), this._viewDate.getDate()));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return result;
            var e_1, _c;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "isDefaultView", {
        get: /**
         * @return {?}
         */
        function () {
            return this._activeView === CalendarView.DEFAULT;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "isYearView", {
        get: /**
         * @return {?}
         */
        function () {
            return this._activeView === CalendarView.YEAR;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "isDecadeView", {
        get: /**
         * @return {?}
         */
        function () {
            return this._activeView === CalendarView.DECADE;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "activeView", {
        /**
         * Gets the current active view of the calendar.
         * ```typescript
         * let activeView =  this.calendar.activeView;
         * ```
         */
        get: /**
         * Gets the current active view of the calendar.
         * ```typescript
         * let activeView =  this.calendar.activeView;
         * ```
         * @return {?}
         */
        function () {
            return this._activeView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "monthAction", {
        /**
         * @hidden
         */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this._monthAction;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "headerTemplate", {
        /**
         * Gets the header template.
         * ```typescript
         * let headerTemplate =  this.calendar.headerTeamplate;
         * ```
         * @memberof IgxCalendarComponent
         */
        get: /**
         * Gets the header template.
         * ```typescript
         * let headerTemplate =  this.calendar.headerTeamplate;
         * ```
         * \@memberof IgxCalendarComponent
         * @return {?}
         */
        function () {
            if (this.headerTemplateDirective) {
                return this.headerTemplateDirective.template;
            }
            return null;
        },
        /**
         * Sets the header template.
         * ```html
         * <igx-calendar headerTemplateDirective = "igxCalendarHeader"></igx-calendar>
         * ```
         * @memberof IgxCalendarComponent
         */
        set: /**
         * Sets the header template.
         * ```html
         * <igx-calendar headerTemplateDirective = "igxCalendarHeader"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} directive
         * @return {?}
         */
        function (directive) {
            this.headerTemplateDirective = directive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "subheaderTemplate", {
        /**
         * Gets the subheader template.
         * ```typescript
         * let subheaderTemplate = this.calendar.subheaderTemplate;
         * ```
         */
        get: /**
         * Gets the subheader template.
         * ```typescript
         * let subheaderTemplate = this.calendar.subheaderTemplate;
         * ```
         * @return {?}
         */
        function () {
            if (this.subheaderTemplateDirective) {
                return this.subheaderTemplateDirective.template;
            }
            return null;
        },
        /**
         * Sets the subheader template.
         * ```html
         * <igx-calendar subheaderTemplate = "igxCalendarSubheader"></igx-calendar>
         * ```
         * @memberof IgxCalendarComponent
         */
        set: /**
         * Sets the subheader template.
         * ```html
         * <igx-calendar subheaderTemplate = "igxCalendarSubheader"></igx-calendar>
         * ```
         * \@memberof IgxCalendarComponent
         * @param {?} directive
         * @return {?}
         */
        function (directive) {
            this.subheaderTemplateDirective = directive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "headerContext", {
        /**
         * Gets the context for the template marked with the `igxCalendarHeader` directive.
         * ```typescript
         * let headerContext =  this.calendar.headerContext;
         * ```
         */
        get: /**
         * Gets the context for the template marked with the `igxCalendarHeader` directive.
         * ```typescript
         * let headerContext =  this.calendar.headerContext;
         * ```
         * @return {?}
         */
        function () {
            var /** @type {?} */ date = this.headerDate;
            return this.generateContext(date);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "context", {
        /**
         * Gets the context for the template marked with either `igxCalendarSubHeaderMonth`
         * or `igxCalendarSubHeaderYear` directive.
         * ```typescript
         * let context =  this.calendar.context;
         * ```
         */
        get: /**
         * Gets the context for the template marked with either `igxCalendarSubHeaderMonth`
         * or `igxCalendarSubHeaderYear` directive.
         * ```typescript
         * let context =  this.calendar.context;
         * ```
         * @return {?}
         */
        function () {
            var /** @type {?} */ date = this._viewDate;
            return this.generateContext(date);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxCalendarComponent.prototype, "headerDate", {
        /**
         * @hidden
         */
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this.selectedDates ? this.selectedDates : new Date();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.ngOnInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        var /** @type {?} */ today = new Date();
        this.calendarModel.firstWeekDay = this.weekStart;
        this._viewDate = this._viewDate ? this._viewDate : today;
        this.initFormatters();
    };
    /**
     * Resets the formatters when locale or formatOptions are changed
     *
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.initFormatters = /**
     * Resets the formatters when locale or formatOptions are changed
     *
     * @hidden
     * @return {?}
     */
    function () {
        this.formatterMonth = new Intl.DateTimeFormat(this._locale, { month: this._formatOptions.month });
        this.formatterDay = new Intl.DateTimeFormat(this._locale, { day: this._formatOptions.day });
        this.formatterYear = new Intl.DateTimeFormat(this._locale, { year: this._formatOptions.year });
        this.formatterMonthday = new Intl.DateTimeFormat(this._locale, { month: this._formatOptions.month, day: this._formatOptions.day });
        this.formatterWeekday = new Intl.DateTimeFormat(this._locale, { weekday: this._formatOptions.weekday });
    };
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    IgxCalendarComponent.prototype.registerOnChange = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChangeCallback = fn;
    };
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    IgxCalendarComponent.prototype.registerOnTouched = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onTouchedCallback = fn;
    };
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.writeValue = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.selectedDates = value;
    };
    /**
     * Returns the locale representation of the month in the month view if enabled,
     * otherwise returns the default `Date.getMonth()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.formattedMonth = /**
     * Returns the locale representation of the month in the month view if enabled,
     * otherwise returns the default `Date.getMonth()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this._formatViews.month) {
            return this.formatterMonth.format(value);
        }
        return "" + value.getMonth();
    };
    /**
     * Returns the locale representation of the date in the default view if enabled,
     * otherwise returns the default `Date.getDate()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.formattedDate = /**
     * Returns the locale representation of the date in the default view if enabled,
     * otherwise returns the default `Date.getDate()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this._formatViews.day) {
            return this.formatterDay.format(value);
        }
        return "" + value.getDate();
    };
    /**
     * Returns the locale representation of the year in the year view if enabled,
     * otherwise returns the default `Date.getFullYear()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.formattedYear = /**
     * Returns the locale representation of the year in the year view if enabled,
     * otherwise returns the default `Date.getFullYear()` value.
     *
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this._formatViews.year) {
            return this.formatterYear.format(value);
        }
        return "" + value.getFullYear();
    };
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.isCurrentMonth = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.viewDate.getMonth() === value.getMonth();
    };
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.isCurrentYear = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.viewDate.getFullYear() === value.getFullYear();
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.previousMonth = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._viewDate = this.calendarModel.timedelta(this._viewDate, 'month', -1);
        this._monthAction = 'prev';
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.nextMonth = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._viewDate = this.calendarModel.timedelta(this._viewDate, 'month', 1);
        this._monthAction = 'next';
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.previousYear = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._viewDate = this.calendarModel.timedelta(this._viewDate, 'year', -1);
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.nextYear = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._viewDate = this.calendarModel.timedelta(this._viewDate, 'year', 1);
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.getFormattedDate = /**
     * @hidden
     * @return {?}
     */
    function () {
        var /** @type {?} */ date = this.headerDate;
        return {
            monthday: this.formatterMonthday.format(date),
            weekday: this.formatterWeekday.format(date),
        };
    };
    /**
     * @hidden
     * @param {?} instance
     * @return {?}
     */
    IgxCalendarComponent.prototype.childClicked = /**
     * @hidden
     * @param {?} instance
     * @return {?}
     */
    function (instance) {
        if (instance.isPrevMonth) {
            this.previousMonth();
        }
        if (instance.isNextMonth) {
            this.nextMonth();
        }
        this.selectDateFromClient(instance.date);
        this.onSelection.emit(this.selectedDates);
    };
    /**
     * @param {?} event
     * @param {?} isLast
     * @return {?}
     */
    IgxCalendarComponent.prototype.animationDone = /**
     * @param {?} event
     * @param {?} isLast
     * @return {?}
     */
    function (event, isLast) {
        if (isLast) {
            var /** @type {?} */ date_1 = this.dates.find(function (d) { return d.selected; });
            if (date_1) {
                setTimeout(function () { return date_1.nativeElement.focus(); }, parseInt(slideInRight.options.params["duration"], 10));
            }
        }
    };
    /**
     * Selects date(s) (based on the selection type).
     * ```typescript
     * this.calendar.selectDate(new Date(`2018-06-12`));
     * ```
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.selectDate = /**
     * Selects date(s) (based on the selection type).
     * ```typescript
     * this.calendar.selectDate(new Date(`2018-06-12`));
     * ```
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value === null || value === undefined || (Array.isArray(value) && value.length === 0)) {
            throw new Error('Date or array should be set for the selectDate method.');
        }
        switch (this.selection) {
            case 'single':
                this.selectSingle(/** @type {?} */ (value));
                break;
            case 'multi':
                this.selectMultiple(value);
                break;
            case 'range':
                this.selectRange(value);
                break;
        }
    };
    /**
     * Deselects date(s) (based on the selection type).
     * ```typescript
     * this.calendar.deselectDate(new Date(`2018-06-12`));
     * ````
     * @param {?=} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.deselectDate = /**
     * Deselects date(s) (based on the selection type).
     * ```typescript
     * this.calendar.deselectDate(new Date(`2018-06-12`));
     * ````
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        if (this.selectedDates === null || this.selectedDates === []) {
            return;
        }
        if (value === null || value === undefined) {
            this.selectedDates = this.selection === 'single' ? null : [];
            this._onChangeCallback(this.selectedDates);
            return;
        }
        switch (this.selection) {
            case 'single':
                this.deselectSingle(/** @type {?} */ (value));
                break;
            case 'multi':
                this.deselectMultiple(/** @type {?} */ (value));
                break;
            case 'range':
                this.deselectRange(/** @type {?} */ (value));
                break;
        }
    };
    /**
     * Checks whether a date is disabled.
     * ```typescript
     * this.calendar.isDateDisabled(new Date(`2018-06-12`));
     * ```
     * @hidden
     * @param {?} date
     * @return {?}
     */
    IgxCalendarComponent.prototype.isDateDisabled = /**
     * Checks whether a date is disabled.
     * ```typescript
     * this.calendar.isDateDisabled(new Date(`2018-06-12`));
     * ```
     * @hidden
     * @param {?} date
     * @return {?}
     */
    function (date) {
        if (this.disabledDates === null) {
            return false;
        }
        return this.isDateInRanges(date, this.disabledDates);
    };
    /**
     * Checks whether a date is special.
     * ```typescript
     * this.calendar.isDateSpecial(new Date(`2018-06-12`));
     * ```
     * @hidden
     * @param {?} date
     * @return {?}
     */
    IgxCalendarComponent.prototype.isDateSpecial = /**
     * Checks whether a date is special.
     * ```typescript
     * this.calendar.isDateSpecial(new Date(`2018-06-12`));
     * ```
     * @hidden
     * @param {?} date
     * @return {?}
     */
    function (date) {
        if (this.specialDates === null) {
            return false;
        }
        return this.isDateInRanges(date, this.specialDates);
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.generateWeekHeader = /**
     * @hidden
     * @return {?}
     */
    function () {
        var /** @type {?} */ dayNames = [];
        var /** @type {?} */ rv = this.calendarModel.monthdatescalendar(this.viewDate.getFullYear(), this.viewDate.getMonth())[0];
        try {
            for (var rv_1 = tslib_1.__values(rv), rv_1_1 = rv_1.next(); !rv_1_1.done; rv_1_1 = rv_1.next()) {
                var day = rv_1_1.value;
                dayNames.push(this.formatterWeekday.format(day.date));
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (rv_1_1 && !rv_1_1.done && (_a = rv_1.return)) _a.call(rv_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return dayNames;
        var e_2, _a;
    };
    Object.defineProperty(IgxCalendarComponent.prototype, "getCalendarMonth", {
        get: /**
         * @hidden
         * @return {?}
         */
        function () {
            return this.calendarModel.monthdatescalendar(this.viewDate.getFullYear(), this.viewDate.getMonth(), true);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.changeYear = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this._viewDate = new Date(event.getFullYear(), this._viewDate.getMonth());
        this._activeView = CalendarView.DEFAULT;
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.changeMonth = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this._viewDate = new Date(this._viewDate.getFullYear(), event.getMonth());
        this._activeView = CalendarView.DEFAULT;
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.activeViewYear = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._activeView = CalendarView.YEAR;
    };
    /**
     * @hidden
     * @return {?}
     */
    IgxCalendarComponent.prototype.activeViewDecade = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._activeView = CalendarView.DECADE;
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onScroll = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopPropagation();
        var /** @type {?} */ delta = event.deltaY < 0 ? -1 : 1;
        this.generateYearRange(delta);
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onPan = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var /** @type {?} */ delta = event.deltaY < 0 ? 1 : -1;
        this.generateYearRange(delta);
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownPageUp = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        this.previousMonth();
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownPageDown = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        this.nextMonth();
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownShiftPageUp = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        this.previousYear();
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownShiftPageDown = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        this.nextYear();
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownArrowUp = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ node = this.dates.find(function (date) { return date.nativeElement === event.target; });
        if (!node) {
            return;
        }
        var /** @type {?} */ dates = this.dates.toArray();
        for (var /** @type {?} */ index = dates.indexOf(node); index - 7 > -1; index -= 7) {
            var /** @type {?} */ date = dates[index - 7];
            if (!date.isDisabled) {
                date.nativeElement.focus();
                break;
            }
        }
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownArrowDown = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ node = this.dates.find(function (date) { return date.nativeElement === event.target; });
        if (!node) {
            return;
        }
        var /** @type {?} */ dates = this.dates.toArray();
        for (var /** @type {?} */ index = dates.indexOf(node); index + 7 < this.dates.length; index += 7) {
            var /** @type {?} */ date = dates[index + 7];
            if (!date.isDisabled) {
                date.nativeElement.focus();
                break;
            }
        }
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownArrowLeft = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ node = this.dates.find(function (date) { return date.nativeElement === event.target; });
        if (!node) {
            return;
        }
        var /** @type {?} */ dates = this.dates.toArray();
        for (var /** @type {?} */ index = dates.indexOf(node); index > 0; index--) {
            var /** @type {?} */ date = dates[index - 1];
            if (!date.isDisabled) {
                date.nativeElement.focus();
                break;
            }
        }
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownArrowRight = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ node = this.dates.find(function (date) { return date.nativeElement === event.target; });
        if (!node) {
            return;
        }
        var /** @type {?} */ dates = this.dates.toArray();
        for (var /** @type {?} */ index = dates.indexOf(node); index < this.dates.length - 1; index++) {
            var /** @type {?} */ date = dates[index + 1];
            if (!date.isDisabled) {
                date.nativeElement.focus();
                break;
            }
        }
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownHome = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ dates = this.dates.filter(function (d) { return d.isCurrentMonth; });
        for (var /** @type {?} */ i = 0; i < dates.length; i++) {
            if (!dates[i].isDisabled) {
                dates[i].nativeElement.focus();
                break;
            }
        }
    };
    /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    IgxCalendarComponent.prototype.onKeydownEnd = /**
     * @hidden
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        var /** @type {?} */ dates = this.dates.filter(function (d) { return d.isCurrentMonth; });
        for (var /** @type {?} */ i = dates.length - 1; i >= 0; i--) {
            if (!dates[i].isDisabled) {
                dates[i].nativeElement.focus();
                break;
            }
        }
    };
    /**
     * @hidden
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    IgxCalendarComponent.prototype.dateTracker = /**
     * @hidden
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    function (index, item) {
        return item.date.getMonth() + "--" + item.date.getDate();
    };
    /**
     * @hidden
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    IgxCalendarComponent.prototype.rowTracker = /**
     * @hidden
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    function (index, item) {
        return "" + item[index].date.getMonth() + item[index].date.getDate();
    };
    /**
     * Performs a single selection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.selectSingle = /**
     * Performs a single selection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.selectedDates = this.getDateOnly(value);
        this._onChangeCallback(this.selectedDates);
    };
    /**
     * Performs a multiple selection
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.selectMultiple = /**
     * Performs a multiple selection
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        if (Array.isArray(value)) {
            this.selectedDates = this.selectedDates.concat(value.map(function (v) { return _this.getDateOnly(v); }));
        }
        else {
            var /** @type {?} */ valueDateOnly_1 = this.getDateOnly(value);
            if (this.selectedDates.every(function (date) { return date.getTime() !== valueDateOnly_1.getTime(); })) {
                this.selectedDates.push(valueDateOnly_1);
            }
            else {
                this.selectedDates = this.selectedDates.filter(function (date) { return date.getTime() !== valueDateOnly_1.getTime(); });
            }
        }
        this._onChangeCallback(this.selectedDates);
    };
    /**
     * @hidden
     * @param {?} value
     * @param {?=} excludeDisabledDates
     * @return {?}
     */
    IgxCalendarComponent.prototype.selectRange = /**
     * @hidden
     * @param {?} value
     * @param {?=} excludeDisabledDates
     * @return {?}
     */
    function (value, excludeDisabledDates) {
        var _this = this;
        if (excludeDisabledDates === void 0) { excludeDisabledDates = false; }
        var /** @type {?} */ start;
        var /** @type {?} */ end;
        if (Array.isArray(value)) {
            this.rangeStarted = false;
            value.sort(function (a, b) { return a.valueOf() - b.valueOf(); });
            start = this.getDateOnly(value[0]);
            end = this.getDateOnly(value[value.length - 1]);
            this.selectedDates = tslib_1.__spread([start], this.generateDateRange(start, end));
        }
        else {
            if (!this.rangeStarted) {
                this.rangeStarted = true;
                this.selectedDates = [value];
            }
            else {
                this.rangeStarted = false;
                if (this.selectedDates[0].getTime() === value.getTime()) {
                    this.selectedDates = [];
                    this._onChangeCallback(this.selectedDates);
                    return;
                }
                this.selectedDates.push(value);
                this.selectedDates.sort(function (a, b) { return a.valueOf() - b.valueOf(); });
                start = this.selectedDates.shift();
                end = this.selectedDates.pop();
                this.selectedDates = tslib_1.__spread([start], this.generateDateRange(start, end));
            }
        }
        if (excludeDisabledDates) {
            this.selectedDates = this.selectedDates.filter(function (d) { return !_this.isDateDisabled(d); });
        }
        this._onChangeCallback(this.selectedDates);
    };
    /**
     * Performs a single deselection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.deselectSingle = /**
     * Performs a single deselection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.selectedDates !== null &&
            this.getDateOnlyInMs(/** @type {?} */ (value)) === this.getDateOnlyInMs(this.selectedDates)) {
            this.selectedDates = null;
            this._onChangeCallback(this.selectedDates);
        }
    };
    /**
     * Performs a multiple deselection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.deselectMultiple = /**
     * Performs a multiple deselection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        value = value.filter(function (v) { return v !== null; });
        var /** @type {?} */ selectedDatesCount = this.selectedDates.length;
        var /** @type {?} */ datesInMsToDeselect = new Set(value.map(function (v) { return _this.getDateOnlyInMs(v); }));
        for (var /** @type {?} */ i = this.selectedDates.length - 1; i >= 0; i--) {
            if (datesInMsToDeselect.has(this.getDateOnlyInMs(this.selectedDates[i]))) {
                this.selectedDates.splice(i, 1);
            }
        }
        if (this.selectedDates.length !== selectedDatesCount) {
            this._onChangeCallback(this.selectedDates);
        }
    };
    /**
     * Performs a range deselection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.deselectRange = /**
     * Performs a range deselection.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        value = value.filter(function (v) { return v !== null && isDate(v); });
        if (value.length < 1) {
            return;
        }
        value.sort(function (a, b) { return a.valueOf() - b.valueOf(); });
        var /** @type {?} */ valueStart = this.getDateOnlyInMs(value[0]);
        var /** @type {?} */ valueEnd = this.getDateOnlyInMs(value[value.length - 1]);
        this.selectedDates.sort(function (a, b) { return a.valueOf() - b.valueOf(); });
        var /** @type {?} */ selectedDatesStart = this.getDateOnlyInMs(this.selectedDates[0]);
        var /** @type {?} */ selectedDatesEnd = this.getDateOnlyInMs(this.selectedDates[this.selectedDates.length - 1]);
        if (!(valueEnd < selectedDatesStart) && !(valueStart > selectedDatesEnd)) {
            this.selectedDates = [];
            this._onChangeCallback(this.selectedDates);
        }
    };
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.selectDateFromClient = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        switch (this.selection) {
            case 'single':
            case 'multi':
                if (!this.isDateDisabled(value)) {
                    this.selectDate(value);
                }
                break;
            case 'range':
                this.selectRange(value, true);
                break;
        }
    };
    /**
     * @hidden
     * @param {?} date
     * @param {?} ranges
     * @return {?}
     */
    IgxCalendarComponent.prototype.isDateInRanges = /**
     * @hidden
     * @param {?} date
     * @param {?} ranges
     * @return {?}
     */
    function (date, ranges) {
        date = new Date(date.getFullYear(), date.getMonth(), date.getDate());
        var /** @type {?} */ dateInMs = date.getTime();
        try {
            for (var ranges_1 = tslib_1.__values(ranges), ranges_1_1 = ranges_1.next(); !ranges_1_1.done; ranges_1_1 = ranges_1.next()) {
                var descriptor = ranges_1_1.value;
                var /** @type {?} */ dRanges = descriptor.dateRange ? descriptor.dateRange.map(function (r) { return new Date(r.getFullYear(), r.getMonth(), r.getDate()); }) : undefined;
                switch (descriptor.type) {
                    case (DateRangeType.After):
                        if (dateInMs > dRanges[0].getTime()) {
                            return true;
                        }
                        break;
                    case (DateRangeType.Before):
                        if (dateInMs < dRanges[0].getTime()) {
                            return true;
                        }
                        break;
                    case (DateRangeType.Between):
                        var /** @type {?} */ dRange = dRanges.map(function (d) { return d.getTime(); });
                        var /** @type {?} */ min = Math.min(dRange[0], dRange[1]);
                        var /** @type {?} */ max = Math.max(dRange[0], dRange[1]);
                        if (dateInMs >= min && dateInMs <= max) {
                            return true;
                        }
                        break;
                    case (DateRangeType.Specific):
                        var /** @type {?} */ datesInMs = dRanges.map(function (d) { return d.getTime(); });
                        try {
                            for (var datesInMs_1 = tslib_1.__values(datesInMs), datesInMs_1_1 = datesInMs_1.next(); !datesInMs_1_1.done; datesInMs_1_1 = datesInMs_1.next()) {
                                var specificDateInMs = datesInMs_1_1.value;
                                if (dateInMs === specificDateInMs) {
                                    return true;
                                }
                            }
                        }
                        catch (e_3_1) { e_3 = { error: e_3_1 }; }
                        finally {
                            try {
                                if (datesInMs_1_1 && !datesInMs_1_1.done && (_a = datesInMs_1.return)) _a.call(datesInMs_1);
                            }
                            finally { if (e_3) throw e_3.error; }
                        }
                        break;
                    case (DateRangeType.Weekdays):
                        var /** @type {?} */ day = date.getDay();
                        if (day % 6 !== 0) {
                            return true;
                        }
                        break;
                    case (DateRangeType.Weekends):
                        var /** @type {?} */ weekday = date.getDay();
                        if (weekday % 6 === 0) {
                            return true;
                        }
                        break;
                    default:
                        return false;
                }
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (ranges_1_1 && !ranges_1_1.done && (_b = ranges_1.return)) _b.call(ranges_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return false;
        var e_4, _b, e_3, _a;
    };
    /**
     * Helper method building and returning the context object inside
     * the calendar templates.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    IgxCalendarComponent.prototype.generateContext = /**
     * Helper method building and returning the context object inside
     * the calendar templates.
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        var /** @type {?} */ formatObject = tslib_1.__assign({ monthView: function () { return _this.activeViewYear(); }, yearView: function () { return _this.activeViewDecade(); } }, this.calendarModel.formatToParts(value, this.locale, this._formatOptions, ['era', 'year', 'month', 'day', 'weekday']));
        return { $implicit: formatObject };
    };
    /**
     * @hidden
     * @param {?} start
     * @param {?} end
     * @return {?}
     */
    IgxCalendarComponent.prototype.generateDateRange = /**
     * @hidden
     * @param {?} start
     * @param {?} end
     * @return {?}
     */
    function (start, end) {
        var /** @type {?} */ result = [];
        start = this.getDateOnly(start);
        end = this.getDateOnly(end);
        while (start.getTime() !== end.getTime()) {
            start = this.calendarModel.timedelta(start, 'day', 1);
            result.push(start);
        }
        return result;
    };
    /**
     * @hidden
     * @param {?} delta
     * @return {?}
     */
    IgxCalendarComponent.prototype.generateYearRange = /**
     * @hidden
     * @param {?} delta
     * @return {?}
     */
    function (delta) {
        var /** @type {?} */ currentYear = new Date().getFullYear();
        if ((delta > 0 && this._viewDate.getFullYear() - currentYear >= 95) ||
            (delta < 0 && currentYear - this._viewDate.getFullYear() >= 95)) {
            return;
        }
        this._viewDate = this.calendarModel.timedelta(this._viewDate, 'year', delta);
    };
    /**
     * @hidden
     * @param {?} date
     * @return {?}
     */
    IgxCalendarComponent.prototype.getDateOnlyInMs = /**
     * @hidden
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.getDateOnly(date).getTime();
    };
    /**
     * @hidden
     * @param {?} date
     * @return {?}
     */
    IgxCalendarComponent.prototype.getDateOnly = /**
     * @hidden
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return new Date(date.getFullYear(), date.getMonth(), date.getDate());
    };
    IgxCalendarComponent.decorators = [
        { type: Component, args: [{
                    animations: [
                        trigger('animateView', [
                            transition('void => 0', useAnimation(fadeIn)),
                            transition('void => *', useAnimation(scaleInCenter, {
                                params: {
                                    duration: '.2s',
                                    fromScale: .9
                                }
                            }))
                        ]),
                        trigger('animateChange', [
                            transition('* => prev', useAnimation(slideInLeft, {
                                params: {
                                    fromPosition: 'translateX(-30%)'
                                }
                            })),
                            transition('* => next', useAnimation(slideInRight, {
                                params: {
                                    fromPosition: 'translateX(30%)'
                                }
                            }))
                        ])
                    ],
                    providers: [
                        {
                            multi: true,
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: IgxCalendarComponent
                        },
                        {
                            provide: HAMMER_GESTURE_CONFIG,
                            useClass: CalendarHammerConfig
                        }
                    ],
                    selector: 'igx-calendar',
                    template: "<ng-template let-result #defaultHeader>\n    <span class=\"date-text\">{{ getFormattedDate().weekday }},&nbsp;</span>\n    <span class=\"date-text\">{{ getFormattedDate().monthday }}</span>\n</ng-template>\n\n<ng-template let-result #defaultMonth>\n    <span (click)=\"activeViewYear()\" class=\"date__el\">\n        {{ formattedMonth(viewDate) }}\n    </span>\n    <span (click)=\"activeViewDecade()\" class=\"date__el\">\n        {{ formattedYear(viewDate) }}\n    </span>\n</ng-template>\n\n<div *ngIf=\"selection === 'single'\" class=\"igx-calendar__header\">\n    <h5 class=\"igx-calendar__header-year\">{{ formattedYear(headerDate) }}</h5>\n    <h2 class=\"igx-calendar__header-date\">\n        <ng-container *ngTemplateOutlet=\"headerTemplate ? headerTemplate : defaultHeader; context: headerContext\">\n        </ng-container>\n    </h2>\n</div>\n\n<div *ngIf=\"isDefaultView\" class=\"igx-calendar__body\" [@animateView]=\"activeView\" (swiperight)=\"previousMonth()\" (swipeleft)=\"nextMonth()\">\n    <div class=\"igx-calendar__body-picker\">\n        <div class=\"prev\" (click)=\"previousMonth()\">\n            <igx-icon fontSet=\"material\">keyboard_arrow_left</igx-icon>\n        </div>\n        <div class=\"date\">\n            <ng-container *ngTemplateOutlet=\"subheaderTemplate ? subheaderTemplate : defaultMonth; context: context\">\n            </ng-container>\n        </div>\n        <div class=\"next\" (click)=\"nextMonth()\">\n            <igx-icon fontSet=\"material\">keyboard_arrow_right</igx-icon>\n        </div>\n    </div>\n\n    <div class=\"igx-calendar__body-row\">\n        <span *ngFor=\"let dayName of generateWeekHeader()\" class=\"igx-calendar__label\">\n            {{ dayName | titlecase }}\n        </span>\n    </div>\n\n    <div *ngFor=\"let week of getCalendarMonth; last as isLast; index as i; trackBy: rowTracker\" class=\"igx-calendar__body-row\" [@animateChange]=\"monthAction\"\n        (@animateChange.done)=\"animationDone($event, isLast)\">\n        <span [igxCalendarDate]=\"day\" (onDateSelection)=\"childClicked($event)\" *ngFor=\"let day of week; trackBy: dateTracker\">\n            {{ formattedDate(day.date) }}\n        </span>\n    </div>\n</div>\n\n<div *ngIf=\"isYearView\" class=\"igx-calendar__body\" [@animateView]=\"activeView\">\n    <div class=\"igx-calendar__body-row--wrap\">\n        <div (onMonthSelection)=\"changeMonth($event)\" [igxCalendarMonth]=\"month\" [index]=\"i\" *ngFor=\"let month of months; index as i;\">\n            {{ formattedMonth(month) | titlecase }}\n        </div>\n    </div>\n</div>\n\n<div *ngIf=\"isDecadeView\" class=\"igx-calendar__body\" [@animateView]=\"activeView\">\n    <div class=\"igx-calendar__body-column\" (wheel)=\"onScroll($event)\" (pan)=\"onPan($event)\">\n        <span (onYearSelection)=\"changeYear($event)\" [igxCalendarYear]=\"year\" *ngFor=\"let year of decade\">\n            {{ formattedYear(year) }}\n        </span>\n    </div>\n</div>\n"
                },] },
    ];
    /** @nocollapse */
    IgxCalendarComponent.ctorParameters = function () { return []; };
    IgxCalendarComponent.propDecorators = {
        "id": [{ type: HostBinding, args: ['attr.id',] }, { type: Input },],
        "weekStart": [{ type: Input },],
        "locale": [{ type: Input },],
        "selection": [{ type: Input },],
        "viewDate": [{ type: Input },],
        "value": [{ type: Input },],
        "formatOptions": [{ type: Input },],
        "formatViews": [{ type: Input },],
        "vertical": [{ type: Input },],
        "onSelection": [{ type: Output },],
        "dates": [{ type: ViewChildren, args: [forwardRef(function () { return IgxCalendarDateDirective; }), { read: IgxCalendarDateDirective },] },],
        "tabindex": [{ type: HostBinding, args: ['attr.tabindex',] },],
        "role": [{ type: HostBinding, args: ['attr.role',] },],
        "ariaLabelledBy": [{ type: HostBinding, args: ['attr.aria-labelledby',] },],
        "styleClass": [{ type: HostBinding, args: ['class',] },],
        "headerTemplateDirective": [{ type: ContentChild, args: [forwardRef(function () { return IgxCalendarHeaderTemplateDirective; }), { read: IgxCalendarHeaderTemplateDirective },] },],
        "subheaderTemplateDirective": [{ type: ContentChild, args: [forwardRef(function () { return IgxCalendarSubheaderTemplateDirective; }), { read: IgxCalendarSubheaderTemplateDirective },] },],
        "onKeydownPageUp": [{ type: HostListener, args: ['keydown.pageup', ['$event'],] },],
        "onKeydownPageDown": [{ type: HostListener, args: ['keydown.pagedown', ['$event'],] },],
        "onKeydownShiftPageUp": [{ type: HostListener, args: ['keydown.shift.pageup', ['$event'],] },],
        "onKeydownShiftPageDown": [{ type: HostListener, args: ['keydown.shift.pagedown', ['$event'],] },],
        "onKeydownArrowUp": [{ type: HostListener, args: ['keydown.arrowup', ['$event'],] },],
        "onKeydownArrowDown": [{ type: HostListener, args: ['keydown.arrowdown', ['$event'],] },],
        "onKeydownArrowLeft": [{ type: HostListener, args: ['keydown.arrowleft', ['$event'],] },],
        "onKeydownArrowRight": [{ type: HostListener, args: ['keydown.arrowright', ['$event'],] },],
        "onKeydownHome": [{ type: HostListener, args: ['keydown.home', ['$event'],] },],
        "onKeydownEnd": [{ type: HostListener, args: ['keydown.end', ['$event'],] },],
    };
    return IgxCalendarComponent;
}());
export { IgxCalendarComponent };
function IgxCalendarComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    IgxCalendarComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    IgxCalendarComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    IgxCalendarComponent.propDecorators;
    /**
     * Sets/gets the `id` of the calendar.
     * If not set, the `id` will have value `"igx-calendar-0"`.
     * ```html
     * <igx-calendar id = "my-first-calendar"></igx-calendar>
     * ```
     * ```typescript
     * let calendarId =  this.calendar.id;
     * ```
     * \@memberof IgxCalendarComponent
     * @type {?}
     */
    IgxCalendarComponent.prototype.id;
    /**
     * Sets/gets whether the calendar header will be in vertical position.
     * Default value is `false`.
     * ```html
     * <igx-calendar [vertical] = "true"></igx-calendar>
     * ```
     * ```typescript
     * let isVertical = this.calendar.vertical;
     * ```
     * @type {?}
     */
    IgxCalendarComponent.prototype.vertical;
    /**
     * Emits an event when a selection is made in the calendar.
     * Provides reference the `selectedDates` property in the `IgxCalendarComponent`.
     * ```html
     * <igx-calendar (onSelection) = "onSelection($event)"></igx-calendar>
     * ```
     * \@memberof IgxCalendarComponent
     * @type {?}
     */
    IgxCalendarComponent.prototype.onSelection;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.dates;
    /**
     * The default `tabindex` attribute for the component.
     *
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.tabindex;
    /**
     * The default aria role attribute for the component.
     *
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.role;
    /**
     * The default aria lebelled by attribute for the component.
     *
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.ariaLabelledBy;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.headerTemplateDirective;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.subheaderTemplateDirective;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._viewDate;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.calendarModel;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._activeView;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.selectedDates;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._selection;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.rangeStarted;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._monthAction;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._locale;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._formatOptions;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._formatViews;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._disabledDates;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.formatterMonth;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.formatterDay;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.formatterYear;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.formatterMonthday;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype.formatterWeekday;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._specialDates;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._onTouchedCallback;
    /**
     * @hidden
     * @type {?}
     */
    IgxCalendarComponent.prototype._onChangeCallback;
}
